{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "During startup - Warning messages:\n",
      "1: package ‘utils’ in options(\"defaultPackages\") was not found \n",
      "2: package ‘stats’ in options(\"defaultPackages\") was not found \n",
      "R[write to console]: Error in .Primitive(\"as.environment\")(\"package:utils\") : \n",
      "  no item called \"package:utils\" on the search list\n",
      "\n"
     ]
    },
    {
     "ename": "RRuntimeError",
     "evalue": "Error in .Primitive(\"as.environment\")(\"package:utils\") : \n  no item called \"package:utils\" on the search list\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[1;32m/home/icb/raphael.kfuri-rubens/git/hlca-v2/notebooks/core_datasets/01_wrangling_validation_assembling/Nawijn_unpubl/Nawijn_unpubl_wrangling.ipynb Cell 1\u001b[0m line \u001b[0;36m4\n\u001b[1;32m      <a href='vscode-notebook-cell://localhost:8081/home/icb/raphael.kfuri-rubens/git/hlca-v2/notebooks/core_datasets/01_wrangling_validation_assembling/Nawijn_unpubl/Nawijn_unpubl_wrangling.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mpandas\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mpd\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell://localhost:8081/home/icb/raphael.kfuri-rubens/git/hlca-v2/notebooks/core_datasets/01_wrangling_validation_assembling/Nawijn_unpubl/Nawijn_unpubl_wrangling.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39manndata\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mad\u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell://localhost:8081/home/icb/raphael.kfuri-rubens/git/hlca-v2/notebooks/core_datasets/01_wrangling_validation_assembling/Nawijn_unpubl/Nawijn_unpubl_wrangling.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=3'>4</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39manndata2ri\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell://localhost:8081/home/icb/raphael.kfuri-rubens/git/hlca-v2/notebooks/core_datasets/01_wrangling_validation_assembling/Nawijn_unpubl/Nawijn_unpubl_wrangling.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=4'>5</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mos\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell://localhost:8081/home/icb/raphael.kfuri-rubens/git/hlca-v2/notebooks/core_datasets/01_wrangling_validation_assembling/Nawijn_unpubl/Nawijn_unpubl_wrangling.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=5'>6</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mos\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mpath\u001b[39;00m \u001b[39mimport\u001b[39;00m join\n",
      "File \u001b[0;32m~/miniforge-pypy3/envs/hlcav2-env/lib/python3.10/site-packages/anndata2ri/__init__.py:20\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpathlib\u001b[39;00m \u001b[39mimport\u001b[39;00m Path\n\u001b[1;32m     18\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mtyping\u001b[39;00m \u001b[39mimport\u001b[39;00m TYPE_CHECKING\n\u001b[0;32m---> 20\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m \u001b[39mimport\u001b[39;00m _py2r, _r2py  \u001b[39m# noqa: F401\u001b[39;00m\n\u001b[1;32m     21\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_conv\u001b[39;00m \u001b[39mimport\u001b[39;00m activate, converter, deactivate\n\u001b[1;32m     24\u001b[0m \u001b[39mif\u001b[39;00m TYPE_CHECKING:\n",
      "File \u001b[0;32m~/miniforge-pypy3/envs/hlcav2-env/lib/python3.10/site-packages/anndata2ri/_py2r.py:9\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mnumpy\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mnp\u001b[39;00m\n\u001b[1;32m      8\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39manndata\u001b[39;00m \u001b[39mimport\u001b[39;00m AnnData\n\u001b[0;32m----> 9\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m \u001b[39mimport\u001b[39;00m conversion, default_converter, pandas2ri\n\u001b[1;32m     10\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mconversion\u001b[39;00m \u001b[39mimport\u001b[39;00m localconverter\n\u001b[1;32m     11\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mvectors\u001b[39;00m \u001b[39mimport\u001b[39;00m ListVector\n",
      "File \u001b[0;32m~/miniforge-pypy3/envs/hlcav2-env/lib/python3.10/site-packages/rpy2/robjects/__init__.py:22\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrlike\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcontainer\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mrlc\u001b[39;00m\n\u001b[1;32m     21\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobject\u001b[39;00m \u001b[39mimport\u001b[39;00m RObjectMixin, RObject\n\u001b[0;32m---> 22\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mfunctions\u001b[39;00m\n\u001b[1;32m     23\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39menvironments\u001b[39;00m \u001b[39mimport\u001b[39;00m (Environment,\n\u001b[1;32m     24\u001b[0m                                         local_context)\n\u001b[1;32m     25\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mmethods\u001b[39;00m \u001b[39mimport\u001b[39;00m methods_env\n",
      "File \u001b[0;32m~/miniforge-pypy3/envs/hlcav2-env/lib/python3.10/site-packages/rpy2/robjects/functions.py:14\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m \u001b[39mimport\u001b[39;00m help\n\u001b[1;32m     13\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m \u001b[39mimport\u001b[39;00m conversion\n\u001b[0;32m---> 14\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mvectors\u001b[39;00m \u001b[39mimport\u001b[39;00m Vector\n\u001b[1;32m     16\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrpy2\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrobjects\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mpackages_utils\u001b[39;00m \u001b[39mimport\u001b[39;00m (default_symbol_r2python,\n\u001b[1;32m     17\u001b[0m                                           default_symbol_resolve,\n\u001b[1;32m     18\u001b[0m                                           _map_symbols,\n\u001b[1;32m     19\u001b[0m                                           _fix_map_symbols)\n\u001b[1;32m     22\u001b[0m baseenv_ri \u001b[39m=\u001b[39m rinterface\u001b[39m.\u001b[39mbaseenv\n",
      "File \u001b[0;32m~/miniforge-pypy3/envs/hlcav2-env/lib/python3.10/site-packages/rpy2/robjects/vectors.py:37\u001b[0m\n\u001b[1;32m     35\u001b[0m r_concat \u001b[39m=\u001b[39m baseenv_ri[\u001b[39m'\u001b[39m\u001b[39mc\u001b[39m\u001b[39m'\u001b[39m]\n\u001b[1;32m     36\u001b[0m as_character \u001b[39m=\u001b[39m baseenv_ri[\u001b[39m'\u001b[39m\u001b[39mas.character\u001b[39m\u001b[39m'\u001b[39m]\n\u001b[0;32m---> 37\u001b[0m utils_ri \u001b[39m=\u001b[39m baseenv_ri[\u001b[39m'\u001b[39;49m\u001b[39mas.environment\u001b[39;49m\u001b[39m'\u001b[39;49m](\n\u001b[1;32m     38\u001b[0m     rinterface\u001b[39m.\u001b[39;49mStrSexpVector((\u001b[39m\"\u001b[39;49m\u001b[39mpackage:utils\u001b[39;49m\u001b[39m\"\u001b[39;49m, ))\n\u001b[1;32m     39\u001b[0m )\n\u001b[1;32m     41\u001b[0m \u001b[39m# The default timezone can be used for time or datetime objects.\u001b[39;00m\n\u001b[1;32m     42\u001b[0m default_timezone \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniforge-pypy3/envs/hlcav2-env/lib/python3.10/site-packages/rpy2/rinterface_lib/conversion.py:45\u001b[0m, in \u001b[0;36m_cdata_res_to_rinterface.<locals>._\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_\u001b[39m(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[0;32m---> 45\u001b[0m     cdata \u001b[39m=\u001b[39m function(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m     46\u001b[0m     \u001b[39m# TODO: test cdata is of the expected CType\u001b[39;00m\n\u001b[1;32m     47\u001b[0m     \u001b[39mreturn\u001b[39;00m _cdata_to_rinterface(cdata)\n",
      "File \u001b[0;32m~/miniforge-pypy3/envs/hlcav2-env/lib/python3.10/site-packages/rpy2/rinterface.py:869\u001b[0m, in \u001b[0;36mSexpClosure.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    862\u001b[0m     res \u001b[39m=\u001b[39m rmemory\u001b[39m.\u001b[39mprotect(\n\u001b[1;32m    863\u001b[0m         openrlib\u001b[39m.\u001b[39mrlib\u001b[39m.\u001b[39mR_tryEval(\n\u001b[1;32m    864\u001b[0m             call_r,\n\u001b[1;32m    865\u001b[0m             call_context\u001b[39m.\u001b[39m__sexp__\u001b[39m.\u001b[39m_cdata,\n\u001b[1;32m    866\u001b[0m             error_occured)\n\u001b[1;32m    867\u001b[0m     )\n\u001b[1;32m    868\u001b[0m     \u001b[39mif\u001b[39;00m error_occured[\u001b[39m0\u001b[39m]:\n\u001b[0;32m--> 869\u001b[0m         \u001b[39mraise\u001b[39;00m embedded\u001b[39m.\u001b[39mRRuntimeError(_rinterface\u001b[39m.\u001b[39m_geterrmessage())\n\u001b[1;32m    870\u001b[0m \u001b[39mreturn\u001b[39;00m res\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in .Primitive(\"as.environment\")(\"package:utils\") : \n  no item called \"package:utils\" on the search list\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import anndata2ri\n",
    "import os\n",
    "from os.path import join\n",
    "import scipy.sparse as sp\n",
    "import numpy as np\n",
    "\n",
    "from hlca_v2.ingestion_utils import get_gspread_df, ValidationWorkflow, AnnDataMerger\n",
    "\n",
    "GSPREAD_JSON = \"/home/icb/raphael.kfuri-rubens/data/hlca_v2/google_sheets_api/hlca-v2-8d5fea10d8f3.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_ID = \"Nawijn_unpubl\"\n",
    "H5AD_PATH = f\"/home/icb/raphael.kfuri-rubens/data/hlca_v2/{DATASET_ID}/{DATASET_ID}_2024_ensembleIDs.h5ad\"\n",
    "OUTPUT_PATH_PREREVISION = '/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_prerevision'\n",
    "OUTPUT_PATH_POSTREVISION = '/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_postrevision'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Name constants\n",
    "AUTHOR_CELL_TYPE_L0 = 'author_cell_type_level_0'\n",
    "AUTHOR_CELL_TYPE_L1 = 'author_cell_type_level_1'\n",
    "AUTHOR_CELL_TYPE_L2 = 'author_cell_type_level_2'\n",
    "AUTHOR_CELL_TYPE_L3 = 'author_cell_type_level_3'\n",
    "\n",
    "CELL_TYPE_ONTOLOGY_ID_L3 = 'cell_type_ontology_term_id_level_3'\n",
    "\n",
    "# Finest grained annotation will be generic dataset cell type\n",
    "AUTHOR_CELL_TYPE = 'author_cell_type'\n",
    "CELL_TYPE_ONTOLOGY_ID = 'cell_type_ontology_term_id'\n",
    "MARKER_GENES = 'author_cell_type_markers'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(H5AD_PATH)\n",
    "obs = get_gspread_df(GSPREAD_JSON, DATASET_ID, \"tier_1\", \"obs\")\n",
    "uns = get_gspread_df(GSPREAD_JSON, DATASET_ID, \"tier_1\", \"uns\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validate obs and uns from Tier 1 Metadata Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation workflow successful. Returning UNS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation workflow successful. Returning UNS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>study_PI</th>\n",
       "      <th>batch_condition</th>\n",
       "      <th>default_embedding</th>\n",
       "      <th>unpublished</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>All--cells from adult human lung from low-air...</td>\n",
       "      <td>Martijn Nawijn</td>\n",
       "      <td>donorID</td>\n",
       "      <td>X_scANVI</td>\n",
       "      <td>consented for release, protected under embargo</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               title        study_PI  \\\n",
       "0   All--cells from adult human lung from low-air...  Martijn Nawijn   \n",
       "\n",
       "  batch_condition default_embedding  \\\n",
       "0         donorID          X_scANVI   \n",
       "\n",
       "                                      unpublished comments  \n",
       "0  consented for release, protected under embargo       -   "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = uns,\n",
    "    axis = 'uns'\n",
    ")\n",
    "\n",
    "validated_uns = val_workflow.init_workflow()\n",
    "validated_uns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation workflow successful. Returning OBS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation workflow successful. Returning OBS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>protocol_URL</th>\n",
       "      <th>institute</th>\n",
       "      <th>sample_collection_site</th>\n",
       "      <th>sample_collection_relative_time_point</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>library_ID_repository</th>\n",
       "      <th>author_batch_notes</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sequenced_fragment</th>\n",
       "      <th>sequencing_platform</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>reference_genome</th>\n",
       "      <th>gene_annotation_version</th>\n",
       "      <th>alignment_software</th>\n",
       "      <th>intron_inclusion</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S02</td>\n",
       "      <td>GRO-02</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S2_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S03</td>\n",
       "      <td>GRO-03</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S3_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S04</td>\n",
       "      <td>GRO-04</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S4_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S05</td>\n",
       "      <td>GRO-05</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S5_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S06</td>\n",
       "      <td>GRO-06</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S07_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S07</td>\n",
       "      <td>GRO-07</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S7_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>S08</td>\n",
       "      <td>GRO-08</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S8_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S09</td>\n",
       "      <td>GRO-09</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S9_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>S10</td>\n",
       "      <td>GRO-10</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S10_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>S11</td>\n",
       "      <td>GRO-11</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S11_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>S12</td>\n",
       "      <td>GRO-12</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S12_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>S13</td>\n",
       "      <td>GRO-13</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S13_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>S14</td>\n",
       "      <td>GRO-14</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S14_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>S15</td>\n",
       "      <td>GRO-15</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S15_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>S16</td>\n",
       "      <td>GRO-16</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S16_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>S17</td>\n",
       "      <td>GRO-17</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S17_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>S18</td>\n",
       "      <td>GRO-18</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S18_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>S19</td>\n",
       "      <td>GRO-19</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S19_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>S20</td>\n",
       "      <td>GRO-20</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S20_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>S21</td>\n",
       "      <td>GRO-21</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S21_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>S22</td>\n",
       "      <td>GRO-22</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S22_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>S23</td>\n",
       "      <td>GRO-23</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S23_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>S24</td>\n",
       "      <td>GRO-24</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S24_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>S25</td>\n",
       "      <td>GRO-25</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S25_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>S26</td>\n",
       "      <td>GRO-26</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S26_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>S27</td>\n",
       "      <td>GRO-27</td>\n",
       "      <td>N/A</td>\n",
       "      <td>University Medical Center Groningen</td>\n",
       "      <td>UMCG_EXPIRE</td>\n",
       "      <td>S27_day1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Batch run by different personnel on different ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_ID donor_id protocol_URL                            institute  \\\n",
       "0        S02   GRO-02          N/A  University Medical Center Groningen   \n",
       "1        S03   GRO-03          N/A  University Medical Center Groningen   \n",
       "2        S04   GRO-04          N/A  University Medical Center Groningen   \n",
       "3        S05   GRO-05          N/A  University Medical Center Groningen   \n",
       "4        S06   GRO-06          N/A  University Medical Center Groningen   \n",
       "5        S07   GRO-07          N/A  University Medical Center Groningen   \n",
       "6        S08   GRO-08          N/A  University Medical Center Groningen   \n",
       "7        S09   GRO-09          N/A  University Medical Center Groningen   \n",
       "8        S10   GRO-10          N/A  University Medical Center Groningen   \n",
       "9        S11   GRO-11          N/A  University Medical Center Groningen   \n",
       "10       S12   GRO-12          N/A  University Medical Center Groningen   \n",
       "11       S13   GRO-13          N/A  University Medical Center Groningen   \n",
       "12       S14   GRO-14          N/A  University Medical Center Groningen   \n",
       "13       S15   GRO-15          N/A  University Medical Center Groningen   \n",
       "14       S16   GRO-16          N/A  University Medical Center Groningen   \n",
       "15       S17   GRO-17          N/A  University Medical Center Groningen   \n",
       "16       S18   GRO-18          N/A  University Medical Center Groningen   \n",
       "17       S19   GRO-19          N/A  University Medical Center Groningen   \n",
       "18       S20   GRO-20          N/A  University Medical Center Groningen   \n",
       "19       S21   GRO-21          N/A  University Medical Center Groningen   \n",
       "20       S22   GRO-22          N/A  University Medical Center Groningen   \n",
       "21       S23   GRO-23          N/A  University Medical Center Groningen   \n",
       "22       S24   GRO-24          N/A  University Medical Center Groningen   \n",
       "23       S25   GRO-25          N/A  University Medical Center Groningen   \n",
       "24       S26   GRO-26          N/A  University Medical Center Groningen   \n",
       "25       S27   GRO-27          N/A  University Medical Center Groningen   \n",
       "\n",
       "   sample_collection_site sample_collection_relative_time_point library_ID  \\\n",
       "0             UMCG_EXPIRE                               S2_day1        N/A   \n",
       "1             UMCG_EXPIRE                               S3_day1        N/A   \n",
       "2             UMCG_EXPIRE                               S4_day1        N/A   \n",
       "3             UMCG_EXPIRE                               S5_day1        N/A   \n",
       "4             UMCG_EXPIRE                              S07_day1        N/A   \n",
       "5             UMCG_EXPIRE                               S7_day1        N/A   \n",
       "6             UMCG_EXPIRE                               S8_day1        N/A   \n",
       "7             UMCG_EXPIRE                               S9_day1        N/A   \n",
       "8             UMCG_EXPIRE                              S10_day1        N/A   \n",
       "9             UMCG_EXPIRE                              S11_day1        N/A   \n",
       "10            UMCG_EXPIRE                              S12_day1        N/A   \n",
       "11            UMCG_EXPIRE                              S13_day1        N/A   \n",
       "12            UMCG_EXPIRE                              S14_day1        N/A   \n",
       "13            UMCG_EXPIRE                              S15_day1        N/A   \n",
       "14            UMCG_EXPIRE                              S16_day1        N/A   \n",
       "15            UMCG_EXPIRE                              S17_day1        N/A   \n",
       "16            UMCG_EXPIRE                              S18_day1        N/A   \n",
       "17            UMCG_EXPIRE                              S19_day1        N/A   \n",
       "18            UMCG_EXPIRE                              S20_day1        N/A   \n",
       "19            UMCG_EXPIRE                              S21_day1        N/A   \n",
       "20            UMCG_EXPIRE                              S22_day1        N/A   \n",
       "21            UMCG_EXPIRE                              S23_day1        N/A   \n",
       "22            UMCG_EXPIRE                              S24_day1        N/A   \n",
       "23            UMCG_EXPIRE                              S25_day1        N/A   \n",
       "24            UMCG_EXPIRE                              S26_day1        N/A   \n",
       "25            UMCG_EXPIRE                              S27_day1        N/A   \n",
       "\n",
       "   library_ID_repository                                 author_batch_notes  \\\n",
       "0                    N/A  Batch run by different personnel on different ...   \n",
       "1                    N/A  Batch run by different personnel on different ...   \n",
       "2                    N/A  Batch run by different personnel on different ...   \n",
       "3                    N/A  Batch run by different personnel on different ...   \n",
       "4                    N/A  Batch run by different personnel on different ...   \n",
       "5                    N/A  Batch run by different personnel on different ...   \n",
       "6                    N/A  Batch run by different personnel on different ...   \n",
       "7                    N/A  Batch run by different personnel on different ...   \n",
       "8                    N/A  Batch run by different personnel on different ...   \n",
       "9                    N/A  Batch run by different personnel on different ...   \n",
       "10                   N/A  Batch run by different personnel on different ...   \n",
       "11                   N/A  Batch run by different personnel on different ...   \n",
       "12                   N/A  Batch run by different personnel on different ...   \n",
       "13                   N/A  Batch run by different personnel on different ...   \n",
       "14                   N/A  Batch run by different personnel on different ...   \n",
       "15                   N/A  Batch run by different personnel on different ...   \n",
       "16                   N/A  Batch run by different personnel on different ...   \n",
       "17                   N/A  Batch run by different personnel on different ...   \n",
       "18                   N/A  Batch run by different personnel on different ...   \n",
       "19                   N/A  Batch run by different personnel on different ...   \n",
       "20                   N/A  Batch run by different personnel on different ...   \n",
       "21                   N/A  Batch run by different personnel on different ...   \n",
       "22                   N/A  Batch run by different personnel on different ...   \n",
       "23                   N/A  Batch run by different personnel on different ...   \n",
       "24                   N/A  Batch run by different personnel on different ...   \n",
       "25                   N/A  Batch run by different personnel on different ...   \n",
       "\n",
       "   organism_ontology_term_id  ... sequenced_fragment sequencing_platform  \\\n",
       "0             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "1             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "2             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "3             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "4             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "5             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "6             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "7             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "8             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "9             NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "10            NCBITaxon:9606  ...        3 prime tag         EFO_0008563   \n",
       "11            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "12            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "13            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "14            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "15            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "16            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "17            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "18            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "19            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "20            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "21            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "22            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "23            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "24            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "25            NCBITaxon:9606  ...        3 prime tag         EFO_0008637   \n",
       "\n",
       "   is_primary_data reference_genome gene_annotation_version  \\\n",
       "0             true           GRCh38                     v98   \n",
       "1             true           GRCh38                     v98   \n",
       "2             true           GRCh38                     v98   \n",
       "3             true           GRCh38                     v98   \n",
       "4             true           GRCh38                     v98   \n",
       "5             true           GRCh38                     v98   \n",
       "6             true           GRCh38                     v98   \n",
       "7             true           GRCh38                     v98   \n",
       "8             true           GRCh38                     v98   \n",
       "9             true           GRCh38                     v98   \n",
       "10            true           GRCh38                     v98   \n",
       "11            true           GRCh38                     v98   \n",
       "12            true           GRCh38                     v98   \n",
       "13            true           GRCh38                     v98   \n",
       "14            true           GRCh38                     v98   \n",
       "15            true           GRCh38                     v98   \n",
       "16            true           GRCh38                     v98   \n",
       "17            true           GRCh38                     v98   \n",
       "18            true           GRCh38                     v98   \n",
       "19            true           GRCh38                     v98   \n",
       "20            true           GRCh38                     v98   \n",
       "21            true           GRCh38                     v98   \n",
       "22            true           GRCh38                     v98   \n",
       "23            true           GRCh38                     v98   \n",
       "24            true           GRCh38                     v98   \n",
       "25            true           GRCh38                     v98   \n",
       "\n",
       "   alignment_software intron_inclusion disease_ontology_term_id  \\\n",
       "0    cellranger-6.1.2               no            PATO:0000461    \n",
       "1    cellranger-6.1.2               no             PATO:0000461   \n",
       "2    cellranger-6.1.2               no             PATO:0000461   \n",
       "3    cellranger-6.1.2               no             PATO:0000461   \n",
       "4    cellranger-6.1.2               no             PATO:0000461   \n",
       "5    cellranger-6.1.2               no             PATO:0000461   \n",
       "6    cellranger-6.1.2               no             PATO:0000461   \n",
       "7    cellranger-6.1.2               no             PATO:0000461   \n",
       "8    cellranger-6.1.2               no             PATO:0000461   \n",
       "9    cellranger-6.1.2               no             PATO:0000461   \n",
       "10   cellranger-6.1.2               no             PATO:0000461   \n",
       "11   cellranger-6.1.2               no             PATO:0000461   \n",
       "12   cellranger-6.1.2               no             PATO:0000461   \n",
       "13   cellranger-6.1.2               no             PATO:0000461   \n",
       "14   cellranger-6.1.2               no             PATO:0000461   \n",
       "15   cellranger-6.1.2               no             PATO:0000461   \n",
       "16   cellranger-6.1.2               no             PATO:0000461   \n",
       "17   cellranger-6.1.2               no             PATO:0000461   \n",
       "18   cellranger-6.1.2               no             PATO:0000461   \n",
       "19   cellranger-6.1.2               no             PATO:0000461   \n",
       "20   cellranger-6.1.2               no             PATO:0000461   \n",
       "21   cellranger-6.1.2               no             PATO:0000461   \n",
       "22   cellranger-6.1.2               no             PATO:0000461   \n",
       "23   cellranger-6.1.2               no             PATO:0000461   \n",
       "24   cellranger-6.1.2               no             PATO:0000461   \n",
       "25   cellranger-6.1.2               no             PATO:0000461   \n",
       "\n",
       "   self_reported_ethnicity_ontology_term_id development_stage_ontology_term_id  \n",
       "0                                   unknown                     HsapDv:0000241  \n",
       "1                                   unknown                     HsapDv:0000240  \n",
       "2                                   unknown                     HsapDv:0000240  \n",
       "3                                   unknown                     HsapDv:0000239  \n",
       "4                                   unknown                     HsapDv:0000240  \n",
       "5                                   unknown                     HsapDv:0000241  \n",
       "6                                   unknown                     HsapDv:0000240  \n",
       "7                                   unknown                     HsapDv:0000240  \n",
       "8                                   unknown                     HsapDv:0000240  \n",
       "9                                   unknown                     HsapDv:0000240  \n",
       "10                                  unknown                     HsapDv:0000240  \n",
       "11                                  unknown                     HsapDv:0000239  \n",
       "12                                  unknown                     HsapDv:0000240  \n",
       "13                                  unknown                     HsapDv:0000241  \n",
       "14                                  unknown                     HsapDv:0000239  \n",
       "15                                  unknown                     HsapDv:0000239  \n",
       "16                                  unknown                     HsapDv:0000241  \n",
       "17                                  unknown                     HsapDv:0000240  \n",
       "18                                  unknown                     HsapDv:0000240  \n",
       "19                                  unknown                     HsapDv:0000240  \n",
       "20                                  unknown                     HsapDv:0000237  \n",
       "21                                  unknown                     HsapDv:0000240  \n",
       "22                                  unknown                     HsapDv:0000241  \n",
       "23                                  unknown                     HsapDv:0000240  \n",
       "24                                  unknown                     HsapDv:0000240  \n",
       "25                                  unknown                     HsapDv:0000239  \n",
       "\n",
       "[26 rows x 37 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = obs,\n",
    "    axis = 'obs'\n",
    ")\n",
    "\n",
    "validated_obs = val_workflow.init_workflow()\n",
    "validated_obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validate obs and uns from adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 147242 × 36601\n",
       "    obs: 'donorID', 'sampleID', 'sex', '10x_v2_or_v3', 'Cell_onto(CL)', 'condition', 'sample_type', 'doublet_score', 'Mylevel1_annot', 'Mylevel2_annot', 'Mylevel3_annot', 'Mylevel4_annot', 'n_genes', 'n_counts', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt'\n",
       "    var: 'gene_names', 'ensembleID'\n",
       "    uns: 'Mylevel1_annot_colors', 'Mylevel4_annot_colors', '_scvi_manager_uuid', '_scvi_uuid', 'hvg', 'log1p', 'pca', 'sampleID_colors', 'sample_type_colors', 'scANVI', 'umap'\n",
       "    obsm: 'Unintegrated', 'X-Scanorama', 'X_Harmony', 'X_pca', 'X_scANVI', 'X_scGen', 'X_scVI', 'X_umap'\n",
       "    obsp: 'scANVI_connectivities', 'scANVI_distances'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 147242 × 36601\n",
       "    obs: 'donorID', 'sampleID', 'sex', '10x_v2_or_v3', 'Cell_onto(CL)', 'condition', 'sample_type', 'doublet_score', 'Mylevel1_annot', 'Mylevel2_annot', 'Mylevel3_annot', 'Mylevel4_annot', 'n_genes', 'n_counts', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt'\n",
       "    var: 'gene_names', 'ensembleID'\n",
       "    uns: 'Mylevel1_annot_colors', 'Mylevel4_annot_colors', '_scvi_manager_uuid', '_scvi_uuid', 'hvg', 'log1p', 'pca', 'sampleID_colors', 'sample_type_colors', 'scANVI', 'umap', 'title', 'study_PI', 'batch_condition', 'default_embedding', 'unpublished', 'comments'\n",
       "    obsm: 'Unintegrated', 'X-Scanorama', 'X_Harmony', 'X_pca', 'X_scANVI', 'X_scGen', 'X_scVI', 'X_umap'\n",
       "    obsp: 'scANVI_connectivities', 'scANVI_distances'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merger = AnnDataMerger(\n",
    "    adata = adata,\n",
    "    uns_df = uns\n",
    ")\n",
    "\n",
    "adata = merger.add_uns_metadata()\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_ids = set(obs['sample_ID'].value_counts().index)\n",
    "adata_ids = set(adata.obs['sampleID'].value_counts().index)\n",
    "symmetric_difference = obs_ids.symmetric_difference(adata_ids)\n",
    "symmetric_difference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donorID</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>sex</th>\n",
       "      <th>10x_v2_or_v3</th>\n",
       "      <th>Cell_onto(CL)</th>\n",
       "      <th>condition</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>doublet_score</th>\n",
       "      <th>Mylevel1_annot</th>\n",
       "      <th>Mylevel2_annot</th>\n",
       "      <th>...</th>\n",
       "      <th>sequencing_platform</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>reference_genome</th>\n",
       "      <th>gene_annotation_version</th>\n",
       "      <th>alignment_software</th>\n",
       "      <th>intron_inclusion</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "      <th>sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGAGTTGC-1_GRO-12</th>\n",
       "      <td>GRO-12</td>\n",
       "      <td>S12</td>\n",
       "      <td>Male</td>\n",
       "      <td>V2</td>\n",
       "      <td>CL:0005012</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.072319</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "      <td>S12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGATGAGGCGCTCT-1_GRO-12</th>\n",
       "      <td>GRO-12</td>\n",
       "      <td>S12</td>\n",
       "      <td>Male</td>\n",
       "      <td>V2</td>\n",
       "      <td>CL:0005012</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.089035</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "      <td>S12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGCAACAGTACACT-1_GRO-12</th>\n",
       "      <td>GRO-12</td>\n",
       "      <td>S12</td>\n",
       "      <td>Male</td>\n",
       "      <td>V2</td>\n",
       "      <td>CL:0005012</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.127660</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "      <td>S12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGCAATCCTAGTGA-1_GRO-12</th>\n",
       "      <td>GRO-12</td>\n",
       "      <td>S12</td>\n",
       "      <td>Male</td>\n",
       "      <td>V2</td>\n",
       "      <td>CL:0005012</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.127660</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "      <td>S12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGTAGGTTGTCTTT-1_GRO-12</th>\n",
       "      <td>GRO-12</td>\n",
       "      <td>S12</td>\n",
       "      <td>Male</td>\n",
       "      <td>V2</td>\n",
       "      <td>CL:0005012</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.096457</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "      <td>S12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGTGTTCAAATACGA-1_GRO-27</th>\n",
       "      <td>GRO-27</td>\n",
       "      <td>S27</td>\n",
       "      <td>female</td>\n",
       "      <td>V3.1</td>\n",
       "      <td>CL:0000165</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.103030</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "      <td>S27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATCTCTAGTACAATAG-1_GRO-27</th>\n",
       "      <td>GRO-27</td>\n",
       "      <td>S27</td>\n",
       "      <td>female</td>\n",
       "      <td>V3.1</td>\n",
       "      <td>CL:0000165</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.142140</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "      <td>S27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATGAAAGCAGGTGACA-1_GRO-27</th>\n",
       "      <td>GRO-27</td>\n",
       "      <td>S27</td>\n",
       "      <td>female</td>\n",
       "      <td>V3.1</td>\n",
       "      <td>CL:0000165</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.126348</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "      <td>S27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGATGAAGACCAGAC-1_GRO-27</th>\n",
       "      <td>GRO-27</td>\n",
       "      <td>S27</td>\n",
       "      <td>female</td>\n",
       "      <td>V3.1</td>\n",
       "      <td>CL:0000165</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.121622</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "      <td>S27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGCGATAGTAATGATG-1_GRO-27</th>\n",
       "      <td>GRO-27</td>\n",
       "      <td>S27</td>\n",
       "      <td>female</td>\n",
       "      <td>V3.1</td>\n",
       "      <td>CL:0000165</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>br_biopsy</td>\n",
       "      <td>0.104905</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO_0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v98</td>\n",
       "      <td>cellranger-6.1.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "      <td>S27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>147242 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          donorID sampleID     sex 10x_v2_or_v3 Cell_onto(CL)  \\\n",
       "AAACCTGAGGAGTTGC-1_GRO-12  GRO-12      S12    Male           V2    CL:0005012   \n",
       "AAAGATGAGGCGCTCT-1_GRO-12  GRO-12      S12    Male           V2    CL:0005012   \n",
       "AAAGCAACAGTACACT-1_GRO-12  GRO-12      S12    Male           V2    CL:0005012   \n",
       "AAAGCAATCCTAGTGA-1_GRO-12  GRO-12      S12    Male           V2    CL:0005012   \n",
       "AAAGTAGGTTGTCTTT-1_GRO-12  GRO-12      S12    Male           V2    CL:0005012   \n",
       "...                           ...      ...     ...          ...           ...   \n",
       "AGGTGTTCAAATACGA-1_GRO-27  GRO-27      S27  female         V3.1    CL:0000165   \n",
       "ATCTCTAGTACAATAG-1_GRO-27  GRO-27      S27  female         V3.1    CL:0000165   \n",
       "ATGAAAGCAGGTGACA-1_GRO-27  GRO-27      S27  female         V3.1    CL:0000165   \n",
       "GGGATGAAGACCAGAC-1_GRO-27  GRO-27      S27  female         V3.1    CL:0000165   \n",
       "TGCGATAGTAATGATG-1_GRO-27  GRO-27      S27  female         V3.1    CL:0000165   \n",
       "\n",
       "                          condition sample_type  doublet_score Mylevel1_annot  \\\n",
       "AAACCTGAGGAGTTGC-1_GRO-12   Healthy   br_biopsy       0.072319     Epithelial   \n",
       "AAAGATGAGGCGCTCT-1_GRO-12   Healthy   br_biopsy       0.089035     Epithelial   \n",
       "AAAGCAACAGTACACT-1_GRO-12   Healthy   br_biopsy       0.127660     Epithelial   \n",
       "AAAGCAATCCTAGTGA-1_GRO-12   Healthy   br_biopsy       0.127660     Epithelial   \n",
       "AAAGTAGGTTGTCTTT-1_GRO-12   Healthy   br_biopsy       0.096457     Epithelial   \n",
       "...                             ...         ...            ...            ...   \n",
       "AGGTGTTCAAATACGA-1_GRO-27   Healthy   br_biopsy       0.103030     Epithelial   \n",
       "ATCTCTAGTACAATAG-1_GRO-27   Healthy   br_biopsy       0.142140     Epithelial   \n",
       "ATGAAAGCAGGTGACA-1_GRO-27   Healthy   br_biopsy       0.126348     Epithelial   \n",
       "GGGATGAAGACCAGAC-1_GRO-27   Healthy   br_biopsy       0.121622     Epithelial   \n",
       "TGCGATAGTAATGATG-1_GRO-27   Healthy   br_biopsy       0.104905     Epithelial   \n",
       "\n",
       "                              Mylevel2_annot  ... sequencing_platform  \\\n",
       "AAACCTGAGGAGTTGC-1_GRO-12  Airway epithelium  ...         EFO_0008563   \n",
       "AAAGATGAGGCGCTCT-1_GRO-12  Airway epithelium  ...         EFO_0008563   \n",
       "AAAGCAACAGTACACT-1_GRO-12  Airway epithelium  ...         EFO_0008563   \n",
       "AAAGCAATCCTAGTGA-1_GRO-12  Airway epithelium  ...         EFO_0008563   \n",
       "AAAGTAGGTTGTCTTT-1_GRO-12  Airway epithelium  ...         EFO_0008563   \n",
       "...                                      ...  ...                 ...   \n",
       "AGGTGTTCAAATACGA-1_GRO-27  Airway epithelium  ...         EFO_0008637   \n",
       "ATCTCTAGTACAATAG-1_GRO-27  Airway epithelium  ...         EFO_0008637   \n",
       "ATGAAAGCAGGTGACA-1_GRO-27  Airway epithelium  ...         EFO_0008637   \n",
       "GGGATGAAGACCAGAC-1_GRO-27  Airway epithelium  ...         EFO_0008637   \n",
       "TGCGATAGTAATGATG-1_GRO-27  Airway epithelium  ...         EFO_0008637   \n",
       "\n",
       "                          is_primary_data  reference_genome  \\\n",
       "AAACCTGAGGAGTTGC-1_GRO-12            true            GRCh38   \n",
       "AAAGATGAGGCGCTCT-1_GRO-12            true            GRCh38   \n",
       "AAAGCAACAGTACACT-1_GRO-12            true            GRCh38   \n",
       "AAAGCAATCCTAGTGA-1_GRO-12            true            GRCh38   \n",
       "AAAGTAGGTTGTCTTT-1_GRO-12            true            GRCh38   \n",
       "...                                   ...               ...   \n",
       "AGGTGTTCAAATACGA-1_GRO-27            true            GRCh38   \n",
       "ATCTCTAGTACAATAG-1_GRO-27            true            GRCh38   \n",
       "ATGAAAGCAGGTGACA-1_GRO-27            true            GRCh38   \n",
       "GGGATGAAGACCAGAC-1_GRO-27            true            GRCh38   \n",
       "TGCGATAGTAATGATG-1_GRO-27            true            GRCh38   \n",
       "\n",
       "                           gene_annotation_version  alignment_software  \\\n",
       "AAACCTGAGGAGTTGC-1_GRO-12                      v98    cellranger-6.1.2   \n",
       "AAAGATGAGGCGCTCT-1_GRO-12                      v98    cellranger-6.1.2   \n",
       "AAAGCAACAGTACACT-1_GRO-12                      v98    cellranger-6.1.2   \n",
       "AAAGCAATCCTAGTGA-1_GRO-12                      v98    cellranger-6.1.2   \n",
       "AAAGTAGGTTGTCTTT-1_GRO-12                      v98    cellranger-6.1.2   \n",
       "...                                            ...                 ...   \n",
       "AGGTGTTCAAATACGA-1_GRO-27                      v98    cellranger-6.1.2   \n",
       "ATCTCTAGTACAATAG-1_GRO-27                      v98    cellranger-6.1.2   \n",
       "ATGAAAGCAGGTGACA-1_GRO-27                      v98    cellranger-6.1.2   \n",
       "GGGATGAAGACCAGAC-1_GRO-27                      v98    cellranger-6.1.2   \n",
       "TGCGATAGTAATGATG-1_GRO-27                      v98    cellranger-6.1.2   \n",
       "\n",
       "                           intron_inclusion  disease_ontology_term_id  \\\n",
       "AAACCTGAGGAGTTGC-1_GRO-12                no              PATO:0000461   \n",
       "AAAGATGAGGCGCTCT-1_GRO-12                no              PATO:0000461   \n",
       "AAAGCAACAGTACACT-1_GRO-12                no              PATO:0000461   \n",
       "AAAGCAATCCTAGTGA-1_GRO-12                no              PATO:0000461   \n",
       "AAAGTAGGTTGTCTTT-1_GRO-12                no              PATO:0000461   \n",
       "...                                     ...                       ...   \n",
       "AGGTGTTCAAATACGA-1_GRO-27                no              PATO:0000461   \n",
       "ATCTCTAGTACAATAG-1_GRO-27                no              PATO:0000461   \n",
       "ATGAAAGCAGGTGACA-1_GRO-27                no              PATO:0000461   \n",
       "GGGATGAAGACCAGAC-1_GRO-27                no              PATO:0000461   \n",
       "TGCGATAGTAATGATG-1_GRO-27                no              PATO:0000461   \n",
       "\n",
       "                           self_reported_ethnicity_ontology_term_id  \\\n",
       "AAACCTGAGGAGTTGC-1_GRO-12                                   unknown   \n",
       "AAAGATGAGGCGCTCT-1_GRO-12                                   unknown   \n",
       "AAAGCAACAGTACACT-1_GRO-12                                   unknown   \n",
       "AAAGCAATCCTAGTGA-1_GRO-12                                   unknown   \n",
       "AAAGTAGGTTGTCTTT-1_GRO-12                                   unknown   \n",
       "...                                                             ...   \n",
       "AGGTGTTCAAATACGA-1_GRO-27                                   unknown   \n",
       "ATCTCTAGTACAATAG-1_GRO-27                                   unknown   \n",
       "ATGAAAGCAGGTGACA-1_GRO-27                                   unknown   \n",
       "GGGATGAAGACCAGAC-1_GRO-27                                   unknown   \n",
       "TGCGATAGTAATGATG-1_GRO-27                                   unknown   \n",
       "\n",
       "                          development_stage_ontology_term_id sample_ID  \n",
       "AAACCTGAGGAGTTGC-1_GRO-12                     HsapDv:0000240       S12  \n",
       "AAAGATGAGGCGCTCT-1_GRO-12                     HsapDv:0000240       S12  \n",
       "AAAGCAACAGTACACT-1_GRO-12                     HsapDv:0000240       S12  \n",
       "AAAGCAATCCTAGTGA-1_GRO-12                     HsapDv:0000240       S12  \n",
       "AAAGTAGGTTGTCTTT-1_GRO-12                     HsapDv:0000240       S12  \n",
       "...                                                      ...       ...  \n",
       "AGGTGTTCAAATACGA-1_GRO-27                     HsapDv:0000239       S27  \n",
       "ATCTCTAGTACAATAG-1_GRO-27                     HsapDv:0000239       S27  \n",
       "ATGAAAGCAGGTGACA-1_GRO-27                     HsapDv:0000239       S27  \n",
       "GGGATGAAGACCAGAC-1_GRO-27                     HsapDv:0000239       S27  \n",
       "TGCGATAGTAATGATG-1_GRO-27                     HsapDv:0000239       S27  \n",
       "\n",
       "[147242 rows x 55 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge obs and uns\n",
    "merger = AnnDataMerger(\n",
    "    adata = adata,\n",
    "    obs_df = obs\n",
    ")\n",
    "\n",
    "adata = merger.add_obs_metadata(\n",
    "    adata_col = 'sampleID',\n",
    "    df_col = 'sample_ID',\n",
    "    skip = None,\n",
    ")\n",
    "\n",
    "adata.obs['sample_ID'] = adata.obs['sampleID']\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add author cell type markers to UNS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['donorID', 'sampleID', 'sex', '10x_v2_or_v3', 'Cell_onto(CL)',\n",
       "       'condition', 'sample_type', 'doublet_score', 'Mylevel1_annot',\n",
       "       'Mylevel2_annot', 'Mylevel3_annot', 'Mylevel4_annot', 'n_genes',\n",
       "       'n_counts', 'n_genes_by_counts', 'total_counts', 'total_counts_mt',\n",
       "       'pct_counts_mt', 'donor_id', 'protocol_URL', 'institute',\n",
       "       'sample_collection_site', 'sample_collection_relative_time_point',\n",
       "       'library_ID', 'library_ID_repository', 'author_batch_notes',\n",
       "       'organism_ontology_term_id', 'manner_of_death', 'sample_source',\n",
       "       'sex_ontology_term_id', 'sample_collection_method', 'tissue_type',\n",
       "       'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text',\n",
       "       'sample_preservation_method', 'suspension_type', 'cell_enrichment',\n",
       "       'cell_viability_percentage', 'cell_number_loaded',\n",
       "       'sample_collection_year', 'assay_ontology_term_id',\n",
       "       'library_preparation_batch', 'library_sequencing_run',\n",
       "       'sequenced_fragment', 'sequencing_platform', 'is_primary_data',\n",
       "       'reference_genome', 'gene_annotation_version', 'alignment_software',\n",
       "       'intron_inclusion', 'disease_ontology_term_id',\n",
       "       'self_reported_ethnicity_ontology_term_id',\n",
       "       'development_stage_ontology_term_id', 'sample_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes_dict = {\n",
    "    \"Club\": ['CYP2F1','SCGB3A1','BPIFB1'],\n",
    "    \"Goblet\": ['ALDH1A3','CEACAM5','MUC5B','MUC5AC'],\n",
    "    \"Basal\": [\"KRT5\",\"S100A2\"],\n",
    "    \"Basal resting cells\": [\"KRT15\",\"KRT17\",\"TP63\"],\n",
    "    \"Cycling Basal\": [\"TOP2A\",\"MKI67\"],\n",
    "    \"Suprabasal\": [\"SERPINB3\"],\n",
    "    \"Hillock-like\": [\"KRT13\",\"KRT4\",\"DSG3\",\"TNS4\",'KRT6A'],\n",
    "    \"Precursor hillock-like\": [\"KRT13\",\"DSG3\",\"TNS4\",'KRT6A','KRT14'],\n",
    "    \"RBCs\": [\"HBB\", \"HBA1\",\"HBA2\"],\n",
    "    #\"Dying cells\": [\"MALAT1\",\"NEAT1\",\"pct_counts_mt\"],\n",
    "    \"Deuterosomal\": [\"CDC20B\",\"E2F7\"],\n",
    "    \"Multiciliated\": [\"FOXJ1\"],\n",
    "    \"Multiciliated (HLCA)\": [\"RSPH1\",\"C20orf85\"],\n",
    "    \"SMG\": [\"LYZ\",\"AZGP1\"],\n",
    "    #\"SMG - Duct\": [\"MGST1\"],\n",
    "    \"SMG - Serous\": [\"S100A1\",'PRB3'],\n",
    "    \"SMG - Mucous\": [\"BPIFB2\",\"MUC5B\",\"NKX3-1\"],\n",
    "    \"Rare\": [\"FOXI1\",\"CFTR\",\"BSND\",\"IGF1\",'CLCNKB', \"SLC6A17\",\"DAB1\"],\n",
    "    \"RBCs\": [\"HBB\", \"HBA1\",\"HBA2\"],\n",
    "    \"B cells\": [\"MS4A1\"],\n",
    "    #\"Plasma cells\": [\"TNFRSF17\"],\n",
    "    #\"T cells proliferating\": [\"CENPW\", \"TK1\", \"MKI67\"],\n",
    "    \"CD4 T cells\": [\"CD40LG\", \"CD4\"], # poss. lowly expressed, non-unique, non-universal\n",
    "    \"CD8 T cells\": [\"CD8A\", \"CD8B\", \"TRGC2\"], # poss lowly expressed, non unique\n",
    "    \"NK cells\": [\"GNLY\", \"SPON2\", \"FCGR3A\"],\n",
    "    \"Plasmacytoid DCs\": [\"SCT\",  \"LILRA4\"],\n",
    "    \"DC1\": [\"XCR1\", \"CLEC9A\", \"CLNK\"],\n",
    "    \"DC2\": [\"CLEC10A\", \"CD1E\"],\n",
    "    \"Migratory DCs\": [\"LAD1\", \"CCL19\"],\n",
    "    \"Alveolar macrophages\": [ \"MARCO\", \"FABP4\"],\n",
    "    \"Alveolar mph CCL3+\": [\"CCL20\"],\n",
    "    \"Alveolar Mph MT-positive\": [\"MT1M\", \"CCL18\"],\n",
    "    \"Alveolar Mph proliferating\": [\"TYMS\", \"CENPW\"],\n",
    "    \"Interstitial Mph perivascular\": [\"F13A1\"],\n",
    "    \"Monocyte derived Mph\": [\"SPP1\"],\n",
    "    \"Classical monocytes\": [\"S100A12\", \"FCN1\", \"RNASE2\"],\n",
    "    \"Non-classical monocyte\": [\"LILRA5\", \"FCN1\"],\n",
    "    \"Mast cell\": [\"MS4A2\", \"SLC18A2\", \"RGS13\"],\n",
    "    \"Hematopoietic stem cell\": [\"CRHBP\", \"SERPINE2\"]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>author_cell_type</th>\n",
       "      <th>author_cell_type_markers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Club</td>\n",
       "      <td>CYP2F1; SCGB3A1; BPIFB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Goblet</td>\n",
       "      <td>ALDH1A3; CEACAM5; MUC5B; MUC5AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Basal</td>\n",
       "      <td>KRT5; S100A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Basal resting cells</td>\n",
       "      <td>KRT15; KRT17; TP63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cycling Basal</td>\n",
       "      <td>TOP2A; MKI67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Suprabasal</td>\n",
       "      <td>SERPINB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Hillock-like</td>\n",
       "      <td>KRT13; KRT4; DSG3; TNS4; KRT6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Precursor hillock-like</td>\n",
       "      <td>KRT13; DSG3; TNS4; KRT6A; KRT14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RBCs</td>\n",
       "      <td>HBB; HBA1; HBA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Deuterosomal</td>\n",
       "      <td>CDC20B; E2F7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Multiciliated</td>\n",
       "      <td>FOXJ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Multiciliated (HLCA)</td>\n",
       "      <td>RSPH1; C20orf85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SMG</td>\n",
       "      <td>LYZ; AZGP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SMG - Serous</td>\n",
       "      <td>S100A1; PRB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SMG - Mucous</td>\n",
       "      <td>BPIFB2; MUC5B; NKX3-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Rare</td>\n",
       "      <td>FOXI1; CFTR; BSND; IGF1; CLCNKB; SLC6A17; DAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>B cells</td>\n",
       "      <td>MS4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CD4 T cells</td>\n",
       "      <td>CD40LG; CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CD8 T cells</td>\n",
       "      <td>CD8A; CD8B; TRGC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NK cells</td>\n",
       "      <td>GNLY; SPON2; FCGR3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Plasmacytoid DCs</td>\n",
       "      <td>SCT; LILRA4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>DC1</td>\n",
       "      <td>XCR1; CLEC9A; CLNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>DC2</td>\n",
       "      <td>CLEC10A; CD1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Migratory DCs</td>\n",
       "      <td>LAD1; CCL19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Alveolar macrophages</td>\n",
       "      <td>MARCO; FABP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Alveolar mph CCL3+</td>\n",
       "      <td>CCL20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Alveolar Mph MT-positive</td>\n",
       "      <td>MT1M; CCL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Alveolar Mph proliferating</td>\n",
       "      <td>TYMS; CENPW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Interstitial Mph perivascular</td>\n",
       "      <td>F13A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Monocyte derived Mph</td>\n",
       "      <td>SPP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Classical monocytes</td>\n",
       "      <td>S100A12; FCN1; RNASE2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Non-classical monocyte</td>\n",
       "      <td>LILRA5; FCN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Mast cell</td>\n",
       "      <td>MS4A2; SLC18A2; RGS13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Hematopoietic stem cell</td>\n",
       "      <td>CRHBP; SERPINE2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 author_cell_type  \\\n",
       "0                            Club   \n",
       "1                          Goblet   \n",
       "2                           Basal   \n",
       "3             Basal resting cells   \n",
       "4                   Cycling Basal   \n",
       "5                      Suprabasal   \n",
       "6                    Hillock-like   \n",
       "7          Precursor hillock-like   \n",
       "8                            RBCs   \n",
       "9                    Deuterosomal   \n",
       "10                  Multiciliated   \n",
       "11           Multiciliated (HLCA)   \n",
       "12                            SMG   \n",
       "13                   SMG - Serous   \n",
       "14                   SMG - Mucous   \n",
       "15                           Rare   \n",
       "16                        B cells   \n",
       "17                    CD4 T cells   \n",
       "18                    CD8 T cells   \n",
       "19                       NK cells   \n",
       "20               Plasmacytoid DCs   \n",
       "21                            DC1   \n",
       "22                            DC2   \n",
       "23                  Migratory DCs   \n",
       "24           Alveolar macrophages   \n",
       "25             Alveolar mph CCL3+   \n",
       "26       Alveolar Mph MT-positive   \n",
       "27     Alveolar Mph proliferating   \n",
       "28  Interstitial Mph perivascular   \n",
       "29           Monocyte derived Mph   \n",
       "30            Classical monocytes   \n",
       "31         Non-classical monocyte   \n",
       "32                      Mast cell   \n",
       "33        Hematopoietic stem cell   \n",
       "\n",
       "                          author_cell_type_markers  \n",
       "0                          CYP2F1; SCGB3A1; BPIFB1  \n",
       "1                  ALDH1A3; CEACAM5; MUC5B; MUC5AC  \n",
       "2                                     KRT5; S100A2  \n",
       "3                               KRT15; KRT17; TP63  \n",
       "4                                     TOP2A; MKI67  \n",
       "5                                         SERPINB3  \n",
       "6                   KRT13; KRT4; DSG3; TNS4; KRT6A  \n",
       "7                  KRT13; DSG3; TNS4; KRT6A; KRT14  \n",
       "8                                  HBB; HBA1; HBA2  \n",
       "9                                     CDC20B; E2F7  \n",
       "10                                           FOXJ1  \n",
       "11                                 RSPH1; C20orf85  \n",
       "12                                      LYZ; AZGP1  \n",
       "13                                    S100A1; PRB3  \n",
       "14                           BPIFB2; MUC5B; NKX3-1  \n",
       "15  FOXI1; CFTR; BSND; IGF1; CLCNKB; SLC6A17; DAB1  \n",
       "16                                           MS4A1  \n",
       "17                                     CD40LG; CD4  \n",
       "18                               CD8A; CD8B; TRGC2  \n",
       "19                             GNLY; SPON2; FCGR3A  \n",
       "20                                     SCT; LILRA4  \n",
       "21                              XCR1; CLEC9A; CLNK  \n",
       "22                                   CLEC10A; CD1E  \n",
       "23                                     LAD1; CCL19  \n",
       "24                                    MARCO; FABP4  \n",
       "25                                           CCL20  \n",
       "26                                     MT1M; CCL18  \n",
       "27                                     TYMS; CENPW  \n",
       "28                                           F13A1  \n",
       "29                                            SPP1  \n",
       "30                           S100A12; FCN1; RNASE2  \n",
       "31                                    LILRA5; FCN1  \n",
       "32                           MS4A2; SLC18A2; RGS13  \n",
       "33                                 CRHBP; SERPINE2  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert marker_genes_dict to df\n",
    "marker_genes_df = pd.DataFrame(marker_genes_dict.items(), columns=[AUTHOR_CELL_TYPE, MARKER_GENES])\n",
    "marker_genes_df[MARKER_GENES] = marker_genes_df[MARKER_GENES].apply(lambda x: ', '.join(x))\n",
    "\n",
    "# replace , by ;\n",
    "marker_genes_df[MARKER_GENES] = marker_genes_df[MARKER_GENES].str.replace(',', ';')\n",
    "marker_genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns[MARKER_GENES] = marker_genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check author cell type annotations and Cell Ontology IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['donorID', 'sampleID', 'sex', '10x_v2_or_v3', 'Cell_onto(CL)',\n",
       "       'condition', 'sample_type', 'doublet_score', 'Mylevel1_annot',\n",
       "       'Mylevel2_annot', 'Mylevel3_annot', 'Mylevel4_annot', 'n_genes',\n",
       "       'n_counts', 'n_genes_by_counts', 'total_counts', 'total_counts_mt',\n",
       "       'pct_counts_mt', 'donor_id', 'protocol_URL', 'institute',\n",
       "       'sample_collection_site', 'sample_collection_relative_time_point',\n",
       "       'library_ID', 'library_ID_repository', 'author_batch_notes',\n",
       "       'organism_ontology_term_id', 'manner_of_death', 'sample_source',\n",
       "       'sex_ontology_term_id', 'sample_collection_method', 'tissue_type',\n",
       "       'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text',\n",
       "       'sample_preservation_method', 'suspension_type', 'cell_enrichment',\n",
       "       'cell_viability_percentage', 'cell_number_loaded',\n",
       "       'sample_collection_year', 'assay_ontology_term_id',\n",
       "       'library_preparation_batch', 'library_sequencing_run',\n",
       "       'sequenced_fragment', 'sequencing_platform', 'is_primary_data',\n",
       "       'reference_genome', 'gene_annotation_version', 'alignment_software',\n",
       "       'intron_inclusion', 'disease_ontology_term_id',\n",
       "       'self_reported_ethnicity_ontology_term_id',\n",
       "       'development_stage_ontology_term_id', 'sample_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mylevel1_annot</th>\n",
       "      <th>Mylevel2_annot</th>\n",
       "      <th>Mylevel3_annot</th>\n",
       "      <th>Mylevel4_annot</th>\n",
       "      <th>Cell_onto(CL)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Multiciliated lineage</td>\n",
       "      <td>Multiciliated cells</td>\n",
       "      <td>CL:0005012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Suprabasal cells</td>\n",
       "      <td>CL:4033048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Dendritic cells</td>\n",
       "      <td>Monocytes-derived DCs</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Multiciliated lineage</td>\n",
       "      <td>Mucous ciliated cells</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Secretory</td>\n",
       "      <td>Club cells</td>\n",
       "      <td>CL:0000158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>Monocyte derived macrophages</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Monocytes</td>\n",
       "      <td>Classical monocytes</td>\n",
       "      <td>CL:0000860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>T cell lineage</td>\n",
       "      <td>CD4 T cells</td>\n",
       "      <td>CL:4033038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Cycling cells</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Dendritic cells</td>\n",
       "      <td>DC2</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Stroma</td>\n",
       "      <td>Fibroblast lineage</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>3_Fibroblasts</td>\n",
       "      <td>CL:0002553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Submucosal Gland</td>\n",
       "      <td>Submucosal Secretory</td>\n",
       "      <td>SMG - Serous cells</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Blood vessels</td>\n",
       "      <td>EC venous</td>\n",
       "      <td>3_EC venous</td>\n",
       "      <td>CL:0002543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>T cell lineage</td>\n",
       "      <td>CD8 T cells</td>\n",
       "      <td>CL:4033039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>B cell lineage</td>\n",
       "      <td>Plasma cells</td>\n",
       "      <td>CL:0000786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Mast cells</td>\n",
       "      <td>3_Mast_cells</td>\n",
       "      <td>CL:0000097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>3_Macrophages</td>\n",
       "      <td>CL:1001603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Blood vessels</td>\n",
       "      <td>EC arterial</td>\n",
       "      <td>3_EC arterial</td>\n",
       "      <td>CL:1000413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>Innate Lymphoid cells NK</td>\n",
       "      <td>NK cells</td>\n",
       "      <td>CL:0000623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal resting cells</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Secretory</td>\n",
       "      <td>Progenitor hillock-like cells</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>B cell lineage</td>\n",
       "      <td>B cells</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Stroma</td>\n",
       "      <td>Smooth muscle</td>\n",
       "      <td>2_Smooth muscle</td>\n",
       "      <td>2_Smooth muscle</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Secretory</td>\n",
       "      <td>Goblet cells</td>\n",
       "      <td>CL:0000160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Secretory</td>\n",
       "      <td>Hillock-like cells</td>\n",
       "      <td>CL:4030024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Submucosal Gland</td>\n",
       "      <td>Submucosal Secretory</td>\n",
       "      <td>SMG - Duct cells</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Monocytes</td>\n",
       "      <td>Non-classical monocytes</td>\n",
       "      <td>CL:0000875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Lymphatic EC</td>\n",
       "      <td>Lymphatic EC mature</td>\n",
       "      <td>3_Lymphatic EC mature</td>\n",
       "      <td>CL:0009115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Dendritic cells</td>\n",
       "      <td>Plasmacytoid DCs</td>\n",
       "      <td>CL:0001058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Multiciliated lineage</td>\n",
       "      <td>Deuterosomal cells</td>\n",
       "      <td>CL:4033044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Rare</td>\n",
       "      <td>Ionocytes</td>\n",
       "      <td>CL:0005006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Dendritic cells</td>\n",
       "      <td>DC1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Rare</td>\n",
       "      <td>Tuft cells</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Submucosal Gland</td>\n",
       "      <td>Submucosal Secretory</td>\n",
       "      <td>SMG - Mucous cells</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Rare</td>\n",
       "      <td>Neuroendocrine cells</td>\n",
       "      <td>CL:0000165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mylevel1_annot      Mylevel2_annot            Mylevel3_annot  \\\n",
       "0      Epithelial   Airway epithelium     Multiciliated lineage   \n",
       "1      Epithelial   Airway epithelium                     Basal   \n",
       "2          Immune             Myeloid           Dendritic cells   \n",
       "3      Epithelial   Airway epithelium     Multiciliated lineage   \n",
       "4      Epithelial   Airway epithelium                 Secretory   \n",
       "5          Immune             Myeloid               Macrophages   \n",
       "6          Immune             Myeloid                 Monocytes   \n",
       "7          Immune            Lymphoid            T cell lineage   \n",
       "8      Epithelial   Airway epithelium                     Basal   \n",
       "9          Immune             Myeloid           Dendritic cells   \n",
       "10         Stroma  Fibroblast lineage               Fibroblasts   \n",
       "11     Epithelial    Submucosal Gland      Submucosal Secretory   \n",
       "12    Endothelial       Blood vessels                 EC venous   \n",
       "13         Immune            Lymphoid            T cell lineage   \n",
       "14         Immune            Lymphoid            B cell lineage   \n",
       "15         Immune             Myeloid                Mast cells   \n",
       "16         Immune             Myeloid               Macrophages   \n",
       "17    Endothelial       Blood vessels               EC arterial   \n",
       "18         Immune            Lymphoid  Innate Lymphoid cells NK   \n",
       "19     Epithelial   Airway epithelium                     Basal   \n",
       "20     Epithelial   Airway epithelium                 Secretory   \n",
       "21         Immune            Lymphoid            B cell lineage   \n",
       "22         Stroma       Smooth muscle           2_Smooth muscle   \n",
       "23     Epithelial   Airway epithelium                 Secretory   \n",
       "24     Epithelial   Airway epithelium                 Secretory   \n",
       "25     Epithelial    Submucosal Gland      Submucosal Secretory   \n",
       "26         Immune             Myeloid                 Monocytes   \n",
       "27    Endothelial        Lymphatic EC       Lymphatic EC mature   \n",
       "28         Immune             Myeloid           Dendritic cells   \n",
       "29     Epithelial   Airway epithelium     Multiciliated lineage   \n",
       "30     Epithelial   Airway epithelium                      Rare   \n",
       "31         Immune             Myeloid           Dendritic cells   \n",
       "32     Epithelial   Airway epithelium                      Rare   \n",
       "33     Epithelial    Submucosal Gland      Submucosal Secretory   \n",
       "34     Epithelial   Airway epithelium                      Rare   \n",
       "\n",
       "                   Mylevel4_annot Cell_onto(CL)  \n",
       "0             Multiciliated cells    CL:0005012  \n",
       "1                Suprabasal cells    CL:4033048  \n",
       "2           Monocytes-derived DCs          none  \n",
       "3           Mucous ciliated cells          none  \n",
       "4                      Club cells    CL:0000158  \n",
       "5    Monocyte derived macrophages          none  \n",
       "6             Classical monocytes    CL:0000860  \n",
       "7                     CD4 T cells    CL:4033038  \n",
       "8                   Cycling cells          none  \n",
       "9                             DC2          none  \n",
       "10                  3_Fibroblasts    CL:0002553  \n",
       "11             SMG - Serous cells          none  \n",
       "12                    3_EC venous    CL:0002543  \n",
       "13                    CD8 T cells    CL:4033039  \n",
       "14                   Plasma cells    CL:0000786  \n",
       "15                   3_Mast_cells    CL:0000097  \n",
       "16                  3_Macrophages    CL:1001603  \n",
       "17                  3_EC arterial    CL:1000413  \n",
       "18                       NK cells    CL:0000623  \n",
       "19            Basal resting cells          none  \n",
       "20  Progenitor hillock-like cells          none  \n",
       "21                        B cells          none  \n",
       "22                2_Smooth muscle          none  \n",
       "23                   Goblet cells    CL:0000160  \n",
       "24             Hillock-like cells    CL:4030024  \n",
       "25               SMG - Duct cells          none  \n",
       "26        Non-classical monocytes    CL:0000875  \n",
       "27          3_Lymphatic EC mature    CL:0009115  \n",
       "28               Plasmacytoid DCs    CL:0001058  \n",
       "29             Deuterosomal cells    CL:4033044  \n",
       "30                      Ionocytes    CL:0005006  \n",
       "31                            DC1          none  \n",
       "32                     Tuft cells          none  \n",
       "33             SMG - Mucous cells          none  \n",
       "34           Neuroendocrine cells    CL:0000165  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations = adata.obs[['Mylevel1_annot','Mylevel2_annot', 'Mylevel3_annot', 'Mylevel4_annot', 'Cell_onto(CL)']].drop_duplicates(subset='Mylevel4_annot').reset_index(drop=True)\n",
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mylevel1_annot</th>\n",
       "      <th>Mylevel2_annot</th>\n",
       "      <th>Mylevel3_annot</th>\n",
       "      <th>Mylevel4_annot</th>\n",
       "      <th>Cell_onto(CL)</th>\n",
       "      <th>author_cell_type_markers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Multiciliated lineage</td>\n",
       "      <td>Multiciliated cells</td>\n",
       "      <td>CL:0005012</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Suprabasal cells</td>\n",
       "      <td>CL:4033048</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Dendritic cells</td>\n",
       "      <td>Monocytes-derived DCs</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Multiciliated lineage</td>\n",
       "      <td>Mucous ciliated cells</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Secretory</td>\n",
       "      <td>Club cells</td>\n",
       "      <td>CL:0000158</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>Monocyte derived macrophages</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Monocytes</td>\n",
       "      <td>Classical monocytes</td>\n",
       "      <td>CL:0000860</td>\n",
       "      <td>S100A12; FCN1; RNASE2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>T cell lineage</td>\n",
       "      <td>CD4 T cells</td>\n",
       "      <td>CL:4033038</td>\n",
       "      <td>CD40LG; CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Cycling cells</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Dendritic cells</td>\n",
       "      <td>DC2</td>\n",
       "      <td>none</td>\n",
       "      <td>CLEC10A; CD1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Stroma</td>\n",
       "      <td>Fibroblast lineage</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>3_Fibroblasts</td>\n",
       "      <td>CL:0002553</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Submucosal Gland</td>\n",
       "      <td>Submucosal Secretory</td>\n",
       "      <td>SMG - Serous cells</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Blood vessels</td>\n",
       "      <td>EC venous</td>\n",
       "      <td>3_EC venous</td>\n",
       "      <td>CL:0002543</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>T cell lineage</td>\n",
       "      <td>CD8 T cells</td>\n",
       "      <td>CL:4033039</td>\n",
       "      <td>CD8A; CD8B; TRGC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>B cell lineage</td>\n",
       "      <td>Plasma cells</td>\n",
       "      <td>CL:0000786</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Mast cells</td>\n",
       "      <td>3_Mast_cells</td>\n",
       "      <td>CL:0000097</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>3_Macrophages</td>\n",
       "      <td>CL:1001603</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Blood vessels</td>\n",
       "      <td>EC arterial</td>\n",
       "      <td>3_EC arterial</td>\n",
       "      <td>CL:1000413</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>Innate Lymphoid cells NK</td>\n",
       "      <td>NK cells</td>\n",
       "      <td>CL:0000623</td>\n",
       "      <td>GNLY; SPON2; FCGR3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal resting cells</td>\n",
       "      <td>none</td>\n",
       "      <td>KRT15; KRT17; TP63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Secretory</td>\n",
       "      <td>Progenitor hillock-like cells</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>B cell lineage</td>\n",
       "      <td>B cells</td>\n",
       "      <td>none</td>\n",
       "      <td>MS4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Stroma</td>\n",
       "      <td>Smooth muscle</td>\n",
       "      <td>2_Smooth muscle</td>\n",
       "      <td>2_Smooth muscle</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Secretory</td>\n",
       "      <td>Goblet cells</td>\n",
       "      <td>CL:0000160</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Secretory</td>\n",
       "      <td>Hillock-like cells</td>\n",
       "      <td>CL:4030024</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Submucosal Gland</td>\n",
       "      <td>Submucosal Secretory</td>\n",
       "      <td>SMG - Duct cells</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Monocytes</td>\n",
       "      <td>Non-classical monocytes</td>\n",
       "      <td>CL:0000875</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Lymphatic EC</td>\n",
       "      <td>Lymphatic EC mature</td>\n",
       "      <td>3_Lymphatic EC mature</td>\n",
       "      <td>CL:0009115</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Dendritic cells</td>\n",
       "      <td>Plasmacytoid DCs</td>\n",
       "      <td>CL:0001058</td>\n",
       "      <td>SCT; LILRA4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Multiciliated lineage</td>\n",
       "      <td>Deuterosomal cells</td>\n",
       "      <td>CL:4033044</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Rare</td>\n",
       "      <td>Ionocytes</td>\n",
       "      <td>CL:0005006</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Immune</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Dendritic cells</td>\n",
       "      <td>DC1</td>\n",
       "      <td>none</td>\n",
       "      <td>XCR1; CLEC9A; CLNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Rare</td>\n",
       "      <td>Tuft cells</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Submucosal Gland</td>\n",
       "      <td>Submucosal Secretory</td>\n",
       "      <td>SMG - Mucous cells</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Airway epithelium</td>\n",
       "      <td>Rare</td>\n",
       "      <td>Neuroendocrine cells</td>\n",
       "      <td>CL:0000165</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mylevel1_annot      Mylevel2_annot            Mylevel3_annot  \\\n",
       "0      Epithelial   Airway epithelium     Multiciliated lineage   \n",
       "1      Epithelial   Airway epithelium                     Basal   \n",
       "2          Immune             Myeloid           Dendritic cells   \n",
       "3      Epithelial   Airway epithelium     Multiciliated lineage   \n",
       "4      Epithelial   Airway epithelium                 Secretory   \n",
       "5          Immune             Myeloid               Macrophages   \n",
       "6          Immune             Myeloid                 Monocytes   \n",
       "7          Immune            Lymphoid            T cell lineage   \n",
       "8      Epithelial   Airway epithelium                     Basal   \n",
       "9          Immune             Myeloid           Dendritic cells   \n",
       "10         Stroma  Fibroblast lineage               Fibroblasts   \n",
       "11     Epithelial    Submucosal Gland      Submucosal Secretory   \n",
       "12    Endothelial       Blood vessels                 EC venous   \n",
       "13         Immune            Lymphoid            T cell lineage   \n",
       "14         Immune            Lymphoid            B cell lineage   \n",
       "15         Immune             Myeloid                Mast cells   \n",
       "16         Immune             Myeloid               Macrophages   \n",
       "17    Endothelial       Blood vessels               EC arterial   \n",
       "18         Immune            Lymphoid  Innate Lymphoid cells NK   \n",
       "19     Epithelial   Airway epithelium                     Basal   \n",
       "20     Epithelial   Airway epithelium                 Secretory   \n",
       "21         Immune            Lymphoid            B cell lineage   \n",
       "22         Stroma       Smooth muscle           2_Smooth muscle   \n",
       "23     Epithelial   Airway epithelium                 Secretory   \n",
       "24     Epithelial   Airway epithelium                 Secretory   \n",
       "25     Epithelial    Submucosal Gland      Submucosal Secretory   \n",
       "26         Immune             Myeloid                 Monocytes   \n",
       "27    Endothelial        Lymphatic EC       Lymphatic EC mature   \n",
       "28         Immune             Myeloid           Dendritic cells   \n",
       "29     Epithelial   Airway epithelium     Multiciliated lineage   \n",
       "30     Epithelial   Airway epithelium                      Rare   \n",
       "31         Immune             Myeloid           Dendritic cells   \n",
       "32     Epithelial   Airway epithelium                      Rare   \n",
       "33     Epithelial    Submucosal Gland      Submucosal Secretory   \n",
       "34     Epithelial   Airway epithelium                      Rare   \n",
       "\n",
       "                   Mylevel4_annot Cell_onto(CL) author_cell_type_markers  \n",
       "0             Multiciliated cells    CL:0005012                      NaN  \n",
       "1                Suprabasal cells    CL:4033048                      NaN  \n",
       "2           Monocytes-derived DCs          none                      NaN  \n",
       "3           Mucous ciliated cells          none                      NaN  \n",
       "4                      Club cells    CL:0000158                      NaN  \n",
       "5    Monocyte derived macrophages          none                      NaN  \n",
       "6             Classical monocytes    CL:0000860    S100A12; FCN1; RNASE2  \n",
       "7                     CD4 T cells    CL:4033038              CD40LG; CD4  \n",
       "8                   Cycling cells          none                      NaN  \n",
       "9                             DC2          none            CLEC10A; CD1E  \n",
       "10                  3_Fibroblasts    CL:0002553                      NaN  \n",
       "11             SMG - Serous cells          none                      NaN  \n",
       "12                    3_EC venous    CL:0002543                      NaN  \n",
       "13                    CD8 T cells    CL:4033039        CD8A; CD8B; TRGC2  \n",
       "14                   Plasma cells    CL:0000786                      NaN  \n",
       "15                   3_Mast_cells    CL:0000097                      NaN  \n",
       "16                  3_Macrophages    CL:1001603                      NaN  \n",
       "17                  3_EC arterial    CL:1000413                      NaN  \n",
       "18                       NK cells    CL:0000623      GNLY; SPON2; FCGR3A  \n",
       "19            Basal resting cells          none       KRT15; KRT17; TP63  \n",
       "20  Progenitor hillock-like cells          none                      NaN  \n",
       "21                        B cells          none                    MS4A1  \n",
       "22                2_Smooth muscle          none                      NaN  \n",
       "23                   Goblet cells    CL:0000160                      NaN  \n",
       "24             Hillock-like cells    CL:4030024                      NaN  \n",
       "25               SMG - Duct cells          none                      NaN  \n",
       "26        Non-classical monocytes    CL:0000875                      NaN  \n",
       "27          3_Lymphatic EC mature    CL:0009115                      NaN  \n",
       "28               Plasmacytoid DCs    CL:0001058              SCT; LILRA4  \n",
       "29             Deuterosomal cells    CL:4033044                      NaN  \n",
       "30                      Ionocytes    CL:0005006                      NaN  \n",
       "31                            DC1          none       XCR1; CLEC9A; CLNK  \n",
       "32                     Tuft cells          none                      NaN  \n",
       "33             SMG - Mucous cells          none                      NaN  \n",
       "34           Neuroendocrine cells    CL:0000165                      NaN  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations[MARKER_GENES] = annotations['Mylevel4_annot'].map(marker_genes_df.set_index(AUTHOR_CELL_TYPE)[MARKER_GENES])\n",
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Basal resting cells              31227\n",
       "Suprabasal cells                 24875\n",
       "Multiciliated cells              22147\n",
       "CD8 T cells                      20258\n",
       "Goblet cells                      7152\n",
       "3_EC venous                       6328\n",
       "Club cells                        4889\n",
       "NK cells                          4026\n",
       "Mucous ciliated cells             3529\n",
       "3_Fibroblasts                     3144\n",
       "B cells                           2688\n",
       "CD4 T cells                       2507\n",
       "DC2                               1919\n",
       "Classical monocytes               1744\n",
       "Monocyte derived macrophages      1400\n",
       "3_Mast_cells                      1396\n",
       "3_EC arterial                     1023\n",
       "2_Smooth muscle                    932\n",
       "Progenitor hillock-like cells      929\n",
       "Plasma cells                       882\n",
       "Monocytes-derived DCs              804\n",
       "Cycling cells                      754\n",
       "SMG - Serous cells                 682\n",
       "SMG - Duct cells                   422\n",
       "Hillock-like cells                 297\n",
       "3_Macrophages                      215\n",
       "Ionocytes                          204\n",
       "Non-classical monocytes            184\n",
       "Plasmacytoid DCs                   153\n",
       "Deuterosomal cells                 144\n",
       "DC1                                124\n",
       "SMG - Mucous cells                 124\n",
       "3_Lymphatic EC mature               88\n",
       "Tuft cells                          34\n",
       "Neuroendocrine cells                18\n",
       "Name: author_cell_type, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[AUTHOR_CELL_TYPE] = adata.obs['Mylevel4_annot']\n",
    "adata.obs[AUTHOR_CELL_TYPE].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "none          45568\n",
       "CL:4033048    24875\n",
       "CL:0005012    22147\n",
       "CL:4033039    20258\n",
       "CL:0000160     7152\n",
       "CL:0002543     6328\n",
       "CL:0000158     4889\n",
       "CL:0000623     4026\n",
       "CL:0002553     3144\n",
       "CL:4033038     2507\n",
       "CL:0000860     1744\n",
       "CL:0000097     1396\n",
       "CL:1000413     1023\n",
       "CL:0000786      882\n",
       "CL:4030024      297\n",
       "CL:1001603      215\n",
       "CL:0005006      204\n",
       "CL:0000875      184\n",
       "CL:0001058      153\n",
       "CL:4033044      144\n",
       "CL:0009115       88\n",
       "CL:0000165       18\n",
       "Name: cell_type_ontology_term_id, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[CELL_TYPE_ONTOLOGY_ID] = adata.obs['Cell_onto(CL)']\n",
    "adata.obs[CELL_TYPE_ONTOLOGY_ID].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[AUTHOR_CELL_TYPE_L0] = adata.obs['Mylevel1_annot']\n",
    "adata.obs[AUTHOR_CELL_TYPE_L1] = adata.obs['Mylevel2_annot']\n",
    "adata.obs[AUTHOR_CELL_TYPE_L2] = adata.obs['Mylevel3_annot']\n",
    "adata.obs[AUTHOR_CELL_TYPE_L3] = adata.obs['Mylevel4_annot']\n",
    "\n",
    "adata.obs[CELL_TYPE_ONTOLOGY_ID_L3] = adata.obs['Cell_onto(CL)']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check whether ENSEMBL IDs in var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_names</th>\n",
       "      <th>ensembleID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-mir-1253</th>\n",
       "      <td>hsa-mir-1253</td>\n",
       "      <td>ENSG00000272920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-mir-423</th>\n",
       "      <td>hsa-mir-423</td>\n",
       "      <td>ENSG00000266919</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_names       ensembleID\n",
       "A1BG                  A1BG  ENSG00000121410\n",
       "A1BG-AS1          A1BG-AS1  ENSG00000268895\n",
       "A1CF                  A1CF  ENSG00000148584\n",
       "A2M                    A2M  ENSG00000175899\n",
       "A2M-AS1            A2M-AS1  ENSG00000245105\n",
       "...                    ...              ...\n",
       "ZYG11B              ZYG11B  ENSG00000162378\n",
       "ZYX                    ZYX  ENSG00000159840\n",
       "ZZEF1                ZZEF1  ENSG00000074755\n",
       "hsa-mir-1253  hsa-mir-1253  ENSG00000272920\n",
       "hsa-mir-423    hsa-mir-423  ENSG00000266919\n",
       "\n",
       "[36601 rows x 2 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000121410</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268895</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000148584</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000175899</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000245105</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000162378</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000159840</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000074755</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000272920</th>\n",
       "      <td>hsa-mir-1253</td>\n",
       "      <td>ENSG00000272920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000266919</th>\n",
       "      <td>hsa-mir-423</td>\n",
       "      <td>ENSG00000266919</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_symbol       ensembl_id\n",
       "index                                         \n",
       "ENSG00000121410          A1BG  ENSG00000121410\n",
       "ENSG00000268895      A1BG-AS1  ENSG00000268895\n",
       "ENSG00000148584          A1CF  ENSG00000148584\n",
       "ENSG00000175899           A2M  ENSG00000175899\n",
       "ENSG00000245105       A2M-AS1  ENSG00000245105\n",
       "...                       ...              ...\n",
       "ENSG00000162378        ZYG11B  ENSG00000162378\n",
       "ENSG00000159840           ZYX  ENSG00000159840\n",
       "ENSG00000074755         ZZEF1  ENSG00000074755\n",
       "ENSG00000272920  hsa-mir-1253  ENSG00000272920\n",
       "ENSG00000266919   hsa-mir-423  ENSG00000266919\n",
       "\n",
       "[36601 rows x 2 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.index = adata.var['ensembleID']\n",
    "adata.var.index.name = 'index'\n",
    "adata.var.rename(columns={'gene_names': 'gene_symbol', 'ensembleID': 'ensembl_id'}, inplace=True)\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27821.0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.toarray().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.X.astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw = adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<147242x36601 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 256943696 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<147242x36601 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 256943696 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27821"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.toarray().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27821"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.X.toarray().max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validation result\n",
    "\n",
    "### UNS Validation\n",
    "- OK: Tier 1 UNS Google Sheet\n",
    "- OK: Tier 1 UNS AnnData Object  same as Google Sheet\n",
    "\n",
    "### OBS Validation\n",
    "- OK: Tier 1 OBS Google Sheet\n",
    "- OK: Tier 1 OBS AnnData Object same as Google Sheet\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Submission Status\n",
    "\n",
    "- CHECK: Raw counts in X and in raw\n",
    "- CHECK: Tier 1 Metadata in OBS\n",
    "- CHECK: Cell Ontology IDs in OBS\n",
    "- CHECK: Author cell type in OBS\n",
    "- CHECK: Marker genes in UNS (however ontologies and marker genes could use some refinement)\n",
    "- CHECK: ENSEMBL IDs and gene symbols in var\n",
    "\n",
    "# Revisions:\n",
    "\n",
    "DONE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(join(OUTPUT_PATH_PREREVISION, f\"{DATASET_ID}.h5ad\"))\n",
    "adata.write_h5ad(join(OUTPUT_PATH_POSTREVISION, f\"{DATASET_ID}.h5ad\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CellxGene ingestion prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/icb/raphael.kfuri-rubens/miniconda3/envs/hlcav2-env/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/home/icb/raphael.kfuri-rubens/miniconda3/envs/hlcav2-env/lib/python3.10/site-packages/anndata/_core/anndata.py:1830: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read_h5ad(join(OUTPUT_PATH_POSTREVISION, f\"{DATASET_ID}.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check whether all obs index are unique\n",
    "len(adata.obs) == len(adata.obs.index.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var.index = adata.var.index.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(adata.var) == len(adata.var.index.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(join(OUTPUT_PATH_POSTREVISION, f\"{DATASET_ID}.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(join(OUTPUT_PATH_POSTREVISION, f\"{DATASET_ID}.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 147242 × 36601\n",
       "    obs: 'sample_ID', 'donor_id', 'protocol_URL', 'institute', 'sample_collection_site', 'sample_collection_relative_time_point', 'library_ID', 'library_ID_repository', 'author_batch_notes', 'organism_ontology_term_id', 'manner_of_death', 'sample_source', 'sex_ontology_term_id', 'sample_collection_method', 'tissue_type', 'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text', 'sample_preservation_method', 'suspension_type', 'cell_enrichment', 'cell_viability_percentage', 'cell_number_loaded', 'sample_collection_year', 'assay_ontology_term_id', 'library_preparation_batch', 'library_sequencing_run', 'sequenced_fragment', 'sequencing_platform', 'is_primary_data', 'reference_genome', 'gene_annotation_version', 'alignment_software', 'intron_inclusion', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'author_cell_type', 'cell_type_ontology_term_id', 'author_cell_type_markers', 'author_cell_type_description'\n",
       "    var: 'gene_symbol'\n",
       "    uns: 'Mylevel1_annot_colors', 'Mylevel4_annot_colors', '_scvi_manager_uuid', '_scvi_uuid', 'author_cell_type_markers', 'batch_condition', 'comments', 'default_embedding', 'hvg', 'log1p', 'pca', 'sampleID_colors', 'sample_type_colors', 'scANVI', 'study_PI', 'title', 'umap', 'unpublished'\n",
       "    obsm: 'Unintegrated', 'X-Scanorama', 'X_Harmony', 'X_pca', 'X_scANVI', 'X_scGen', 'X_scVI', 'X_umap'\n",
       "    obsp: 'scANVI_connectivities', 'scANVI_distances'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(\"/ictstr01/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_postrevision_reduced/Nawijn_unpubl.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 147242 × 36601\n",
       "    obs: 'sample_ID', 'donor_id', 'protocol_URL', 'institute', 'sample_collection_site', 'sample_collection_relative_time_point', 'library_ID', 'library_ID_repository', 'author_batch_notes', 'organism_ontology_term_id', 'manner_of_death', 'sample_source', 'sex_ontology_term_id', 'sample_collection_method', 'tissue_type', 'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text', 'sample_preservation_method', 'suspension_type', 'cell_enrichment', 'cell_viability_percentage', 'cell_number_loaded', 'sample_collection_year', 'assay_ontology_term_id', 'library_preparation_batch', 'library_sequencing_run', 'sequenced_fragment', 'sequencing_platform', 'is_primary_data', 'reference_genome', 'gene_annotation_version', 'alignment_software', 'intron_inclusion', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'author_cell_type', 'cell_type_ontology_term_id', 'author_cell_type_markers', 'author_cell_type_description'\n",
       "    var: 'gene_symbol'\n",
       "    uns: 'author_cell_type_markers', 'batch_condition', 'comments', 'default_embedding', 'study_PI', 'title', 'unpublished'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "remove_uns = ['Mylevel1_annot_colors', 'Mylevel4_annot_colors', '_scvi_manager_uuid', '_scvi_uuid', 'hvg', 'log1p', 'pca', 'sampleID_colors', 'sample_type_colors', 'scANVI', 'umap']\n",
    "\n",
    "for key in remove_uns:\n",
    "    if key in adata.uns:\n",
    "        del adata.uns[key]\n",
    "\n",
    "del adata.obsm\n",
    "\n",
    "del adata.obsp\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"/ictstr01/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_postrevision_reduced/Nawijn_unpubl.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hlcav2-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
