{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "package ‘methods’ was built under R version 4.3.3 \n",
      "During startup - Warning messages:\n",
      "1: package ‘datasets’ was built under R version 4.3.3 \n",
      "2: package ‘utils’ was built under R version 4.3.3 \n",
      "3: package ‘grDevices’ was built under R version 4.3.3 \n",
      "4: package ‘graphics’ was built under R version 4.3.3 \n",
      "5: package ‘stats’ was built under R version 4.3.3 \n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import anndata2ri\n",
    "import os\n",
    "from os.path import join\n",
    "import scipy.sparse as sp\n",
    "import numpy as np\n",
    "\n",
    "from hlca_v2.ingestion_utils import get_gspread_df, ValidationWorkflow, AnnDataMerger\n",
    "\n",
    "GSPREAD_JSON = \"/home/icb/raphael.kfuri-rubens/data/hlca_v2/google_sheets_api/hlca-v2-8d5fea10d8f3.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_ID = \"Meyer_01_publ\"\n",
    "H5AD_PATH = f\"/home/icb/raphael.kfuri-rubens/data/hlca_v2/{DATASET_ID}/{DATASET_ID}.h5ad\"\n",
    "OUTPUT_PATH_PREREVISION = '/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_prerevision'\n",
    "OUTPUT_PATH_POSTREVISION = '/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_postrevision'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Name constants\n",
    "AUTHOR_CELL_TYPE_L0 = 'author_cell_type_level_0'\n",
    "AUTHOR_CELL_TYPE_L1 = 'author_cell_type_level_1'\n",
    "AUTHOR_CELL_TYPE_L2 = 'author_cell_type_level_2'\n",
    "\n",
    "CELL_TYPE_ONTOLOGY_ID_L0 = 'cell_type_ontology_term_id_level_0'\n",
    "CELL_TYPE_ONTOLOGY_ID_L1 = 'cell_type_ontology_term_id_level_1'\n",
    "CELL_TYPE_ONTOLOGY_ID_L2 = 'cell_type_ontology_term_id_level_2'\n",
    "\n",
    "MARKER_GENES_L0 = 'author_cell_type_markers_level_0'\n",
    "MARKER_GENES_L1 = 'author_cell_type_markers_level_1'\n",
    "MARKER_GENES_L2 = 'author_cell_type_markers_level_2'\n",
    "\n",
    "# Finest grained annotation will be generic dataset cell type\n",
    "AUTHOR_CELL_TYPE = 'author_cell_type'\n",
    "CELL_TYPE_ONTOLOGY_ID = 'cell_type_ontology_term_id'\n",
    "MARKER_GENES = 'author_cell_type_markers'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(H5AD_PATH)\n",
    "obs = get_gspread_df(GSPREAD_JSON, DATASET_ID, \"tier_1\", \"obs\")\n",
    "uns = get_gspread_df(GSPREAD_JSON, DATASET_ID, \"tier_1\", \"uns\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_ID', 'donor_id', 'protocol_URL', 'institute',\n",
       "       'sample_collection_site', 'sample_collection_relative_time_point',\n",
       "       'library_ID', 'library_ID_repository', 'author_batch_notes',\n",
       "       'organism_ontology_term_id', 'manner_of_death', 'sample_source',\n",
       "       'sex_ontology_term_id', 'sample_collection_method', 'tissue_type',\n",
       "       'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text',\n",
       "       'sample_preservation_method', 'suspension_type', 'cell_enrichment',\n",
       "       'cell_viability_percentage', 'cell_number_loaded',\n",
       "       'sample_collection_year', 'assay_ontology_term_id',\n",
       "       'library_preparation_batch', 'library_sequencing_run',\n",
       "       'sequenced_fragment', 'sequencing_platform', 'is_primary_data',\n",
       "       'reference_genome', 'gene_annotation_version', 'alignment_software',\n",
       "       'intron_inclusion', 'disease_ontology_term_id',\n",
       "       'self_reported_ethnicity_ontology_term_id',\n",
       "       'development_stage_ontology_term_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validate obs and uns from Tier 1 Metadata Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation workflow successful. Returning UNS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation workflow successful. Returning UNS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>study_PI</th>\n",
       "      <th>batch_condition</th>\n",
       "      <th>default_embedding</th>\n",
       "      <th>unpublished</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cells collection from A spatially resolved atl...</td>\n",
       "      <td>Kerstin, B, Meyer</td>\n",
       "      <td>-</td>\n",
       "      <td>X_umap</td>\n",
       "      <td>published</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               title           study_PI  \\\n",
       "0  Cells collection from A spatially resolved atl...  Kerstin, B, Meyer   \n",
       "\n",
       "  batch_condition default_embedding unpublished comments  \n",
       "0               -            X_umap   published        -  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = uns,\n",
    "    axis = 'uns'\n",
    ")\n",
    "\n",
    "validated_uns = val_workflow.init_workflow()\n",
    "validated_uns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation workflow successful. Returning OBS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation workflow successful. Returning OBS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>protocol_URL</th>\n",
       "      <th>institute</th>\n",
       "      <th>sample_collection_site</th>\n",
       "      <th>sample_collection_relative_time_point</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>library_ID_repository</th>\n",
       "      <th>author_batch_notes</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sequenced_fragment</th>\n",
       "      <th>sequencing_platform</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>reference_genome</th>\n",
       "      <th>gene_annotation_version</th>\n",
       "      <th>alignment_software</th>\n",
       "      <th>intron_inclusion</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5841STDY7991475</td>\n",
       "      <td>A37</td>\n",
       "      <td></td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37-LNG-1-SC-45P-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>5 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5841STDY7991476</td>\n",
       "      <td>A37</td>\n",
       "      <td></td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37-LNG-3-SC-45P-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>5 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5841STDY7991477</td>\n",
       "      <td>A37</td>\n",
       "      <td></td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37-LNG-4-SC-45P-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>5 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5841STDY7991478</td>\n",
       "      <td>A37</td>\n",
       "      <td></td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37-LNG-4-SC-45P-2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>5 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5841STDY7991479</td>\n",
       "      <td>A37</td>\n",
       "      <td></td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37-LNG-5-SC-45P-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A37_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>5 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>WTDAtest7888001</td>\n",
       "      <td>A32</td>\n",
       "      <td></td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32-LNG-2-SC-45N-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>WTDAtest7888002</td>\n",
       "      <td>A32</td>\n",
       "      <td></td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32-LNG-2-SC-45P-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>WTDAtest7985096</td>\n",
       "      <td>A26</td>\n",
       "      <td>https://www.protocols.io/view/nuclei-isolation...</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WTDAtest7985096</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A26_nuclei</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>WTDAtest8433409</td>\n",
       "      <td>A48</td>\n",
       "      <td>https://www.protocols.io/view/nuclei-isolation...</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WTDAtest8433409</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A48_nuclei</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>WTDAtest8550127</td>\n",
       "      <td>A48</td>\n",
       "      <td>https://www.protocols.io/view/single-cell-and-...</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WTDAtest8550127</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A48_nuclei</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample_ID donor_id  \\\n",
       "0   5841STDY7991475      A37   \n",
       "1   5841STDY7991476      A37   \n",
       "2   5841STDY7991477      A37   \n",
       "3   5841STDY7991478      A37   \n",
       "4   5841STDY7991479      A37   \n",
       "..              ...      ...   \n",
       "70  WTDAtest7888001      A32   \n",
       "71  WTDAtest7888002      A32   \n",
       "72  WTDAtest7985096      A26   \n",
       "73  WTDAtest8433409      A48   \n",
       "74  WTDAtest8550127      A48   \n",
       "\n",
       "                                         protocol_URL  \\\n",
       "0                                                       \n",
       "1                                                       \n",
       "2                                                       \n",
       "3                                                       \n",
       "4                                                       \n",
       "..                                                ...   \n",
       "70                                                      \n",
       "71                                                      \n",
       "72  https://www.protocols.io/view/nuclei-isolation...   \n",
       "73  https://www.protocols.io/view/nuclei-isolation...   \n",
       "74  https://www.protocols.io/view/single-cell-and-...   \n",
       "\n",
       "                    institute sample_collection_site  \\\n",
       "0   Wellcome Sanger Institute              Cambridge   \n",
       "1   Wellcome Sanger Institute              Cambridge   \n",
       "2   Wellcome Sanger Institute              Cambridge   \n",
       "3   Wellcome Sanger Institute              Cambridge   \n",
       "4   Wellcome Sanger Institute              Cambridge   \n",
       "..                        ...                    ...   \n",
       "70  Wellcome Sanger Institute              Cambridge   \n",
       "71  Wellcome Sanger Institute              Cambridge   \n",
       "72  Wellcome Sanger Institute              Cambridge   \n",
       "73  Wellcome Sanger Institute              Cambridge   \n",
       "74  Wellcome Sanger Institute              Cambridge   \n",
       "\n",
       "   sample_collection_relative_time_point          library_ID  \\\n",
       "0                                    N/A  A37-LNG-1-SC-45P-1   \n",
       "1                                    N/A  A37-LNG-3-SC-45P-1   \n",
       "2                                    N/A  A37-LNG-4-SC-45P-1   \n",
       "3                                    N/A  A37-LNG-4-SC-45P-2   \n",
       "4                                    N/A  A37-LNG-5-SC-45P-1   \n",
       "..                                   ...                 ...   \n",
       "70                                   N/A  A32-LNG-2-SC-45N-1   \n",
       "71                                   N/A  A32-LNG-2-SC-45P-1   \n",
       "72                                   N/A     WTDAtest7985096   \n",
       "73                                   N/A     WTDAtest8433409   \n",
       "74                                   N/A     WTDAtest8550127   \n",
       "\n",
       "   library_ID_repository author_batch_notes organism_ontology_term_id  ...  \\\n",
       "0                    N/A          A37_cells            NCBITaxon:9606  ...   \n",
       "1                    N/A          A37_cells            NCBITaxon:9606  ...   \n",
       "2                    N/A          A37_cells            NCBITaxon:9606  ...   \n",
       "3                    N/A          A37_cells            NCBITaxon:9606  ...   \n",
       "4                    N/A          A37_cells            NCBITaxon:9606  ...   \n",
       "..                   ...                ...                       ...  ...   \n",
       "70                   N/A          A32_cells            NCBITaxon:9606  ...   \n",
       "71                   N/A          A32_cells            NCBITaxon:9606  ...   \n",
       "72                   N/A         A26_nuclei            NCBITaxon:9606  ...   \n",
       "73                   N/A         A48_nuclei            NCBITaxon:9606  ...   \n",
       "74                   N/A         A48_nuclei            NCBITaxon:9606  ...   \n",
       "\n",
       "   sequenced_fragment sequencing_platform is_primary_data reference_genome  \\\n",
       "0         5 prime tag         EFO:0008637            true           GRCh38   \n",
       "1         5 prime tag         EFO:0008637            true           GRCh38   \n",
       "2         5 prime tag         EFO:0008637            true           GRCh38   \n",
       "3         5 prime tag         EFO:0008637            true           GRCh38   \n",
       "4         5 prime tag         EFO:0008637            true           GRCh38   \n",
       "..                ...                 ...             ...              ...   \n",
       "70        3 prime tag         EFO:0008563            true           GRCh38   \n",
       "71        3 prime tag         EFO:0008563            true           GRCh38   \n",
       "72                 na         EFO:0008637            true           GRCh38   \n",
       "73                 na         EFO:0008637            true           GRCh38   \n",
       "74                 na         EFO:0008637            true           GRCh38   \n",
       "\n",
       "   gene_annotation_version alignment_software intron_inclusion  \\\n",
       "0               Ensembl 93  cell ranger 3.0.2               no   \n",
       "1               Ensembl 93  cell ranger 3.0.2               no   \n",
       "2               Ensembl 93  cell ranger 3.0.2               no   \n",
       "3               Ensembl 93  cell ranger 3.0.2               no   \n",
       "4               Ensembl 93  cell ranger 3.0.2               no   \n",
       "..                     ...                ...              ...   \n",
       "70              Ensembl 93  cell ranger 3.0.2               no   \n",
       "71              Ensembl 93  cell ranger 3.0.2               no   \n",
       "72              Ensembl 93  cell ranger 3.0.2              yes   \n",
       "73              Ensembl 93  cell ranger 3.0.2              yes   \n",
       "74              Ensembl 93  cell ranger 3.0.2              yes   \n",
       "\n",
       "   disease_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0              PATO:0000461                                  unknown   \n",
       "1              PATO:0000461                                  unknown   \n",
       "2              PATO:0000461                                  unknown   \n",
       "3              PATO:0000461                                  unknown   \n",
       "4              PATO:0000461                                  unknown   \n",
       "..                      ...                                      ...   \n",
       "70             PATO:0000461                                  unknown   \n",
       "71             PATO:0000461                                  unknown   \n",
       "72             PATO:0000461                                  unknown   \n",
       "73             PATO:0000461                                  unknown   \n",
       "74             PATO:0000461                                  unknown   \n",
       "\n",
       "   development_stage_ontology_term_id  \n",
       "0                      HsapDv:0000240  \n",
       "1                      HsapDv:0000240  \n",
       "2                      HsapDv:0000240  \n",
       "3                      HsapDv:0000240  \n",
       "4                      HsapDv:0000240  \n",
       "..                                ...  \n",
       "70                     HsapDv:0000237  \n",
       "71                     HsapDv:0000237  \n",
       "72                     HsapDv:0000242  \n",
       "73                     HsapDv:0000241  \n",
       "74                     HsapDv:0000241  \n",
       "\n",
       "[75 rows x 37 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = obs,\n",
    "    axis = 'obs'\n",
    ")\n",
    "\n",
    "validated_obs = val_workflow.init_workflow()\n",
    "validated_obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validate obs and uns from adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation workflow successful. Returning UNS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation workflow successful. Returning UNS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>study_PI</th>\n",
       "      <th>unpublished</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cells collection from A spatially resolved atl...</td>\n",
       "      <td>Kerstin B. Meyer</td>\n",
       "      <td>published</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               title          study_PI  \\\n",
       "0  Cells collection from A spatially resolved atl...  Kerstin B. Meyer   \n",
       "\n",
       "  unpublished  \n",
       "0   published  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = adata,\n",
    "    axis = 'uns'\n",
    ")\n",
    "\n",
    "validated_uns = val_workflow.init_workflow()\n",
    "validated_uns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation <span style=\"color: #800080; text-decoration-color: #800080; font-weight: bold\">error</span><span style=\"font-weight: bold\">(</span>s<span style=\"font-weight: bold\">)</span> encountered.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation \u001b[1;35merror\u001b[0m\u001b[1m(\u001b[0ms\u001b[1m)\u001b[0m encountered.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">{</span>\n",
       "<span style=\"color: #7fbf7f; text-decoration-color: #7fbf7f\">│   </span><span style=\"color: #008000; text-decoration-color: #008000\">'sample_preservation_method'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"Must be a valid method such as 'fresh', 'frozen at -80C', etc. Invalid entries found.\"</span>\n",
       "<span style=\"font-weight: bold\">}</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1m{\u001b[0m\n",
       "\u001b[2;32m│   \u001b[0m\u001b[32m'sample_preservation_method'\u001b[0m: \u001b[32m\"Must be a valid method such as 'fresh', 'frozen at -80C', etc. Invalid entries found.\"\u001b[0m\n",
       "\u001b[1m}\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Error containing field: sample_preservation_method\n",
       "Error message: Must be a valid method such as <span style=\"color: #008000; text-decoration-color: #008000\">'fresh'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'frozen at -80C'</span>, etc. Invalid entries found.\n",
       "Value counts:\n",
       "fresh     <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">129340</span>\n",
       "frozen     <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">63768</span>\n",
       "Name: sample_preservation_method, dtype: int64\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Error containing field: sample_preservation_method\n",
       "Error message: Must be a valid method such as \u001b[32m'fresh'\u001b[0m, \u001b[32m'frozen at -80C'\u001b[0m, etc. Invalid entries found.\n",
       "Value counts:\n",
       "fresh     \u001b[1;36m129340\u001b[0m\n",
       "frozen     \u001b[1;36m63768\u001b[0m\n",
       "Name: sample_preservation_method, dtype: int64\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Returning error-containing OBS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Returning error-containing OBS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>protocol_URL</th>\n",
       "      <th>institute</th>\n",
       "      <th>sample_collection_site</th>\n",
       "      <th>sample_collection_relative_time_point</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>library_ID_repository</th>\n",
       "      <th>author_batch_notes</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sequenced_fragment</th>\n",
       "      <th>sequencing_platform</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>reference_genome</th>\n",
       "      <th>gene_annotation_version</th>\n",
       "      <th>alignment_software</th>\n",
       "      <th>intron_inclusion</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WTDAtest7887999-AAACCTGGTGTGAATA</th>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>A32</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32-LNG-1-SC-45N-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WTDAtest7887999-AAACGGGCAACCGCCA</th>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>A32</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32-LNG-1-SC-45N-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WTDAtest7887999-AACCATGGTACAGCAG</th>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>A32</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32-LNG-1-SC-45N-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WTDAtest7887999-AACGTTGGTGTCCTCT</th>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>A32</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32-LNG-1-SC-45N-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WTDAtest7887999-AACTCCCTCCTAGTGA</th>\n",
       "      <td>WTDAtest7887999</td>\n",
       "      <td>A32</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32-LNG-1-SC-45N-1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A32_cells</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WSSS_A_LNG8757929-TTTGGTTGTAATGCGG</th>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>A42</td>\n",
       "      <td>https://www.protocols.io/view/single-cell-and-...</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT</th>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>A42</td>\n",
       "      <td>https://www.protocols.io/view/single-cell-and-...</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA</th>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>A42</td>\n",
       "      <td>https://www.protocols.io/view/single-cell-and-...</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA</th>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>A42</td>\n",
       "      <td>https://www.protocols.io/view/single-cell-and-...</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC</th>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>A42</td>\n",
       "      <td>https://www.protocols.io/view/single-cell-and-...</td>\n",
       "      <td>Wellcome Sanger Institute</td>\n",
       "      <td>Cambridge</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WSSS_A_LNG8757929</td>\n",
       "      <td>N/A</td>\n",
       "      <td>A42_nuclei</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>Ensembl 93</td>\n",
       "      <td>cell ranger 3.0.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193108 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            sample_ID donor_id  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA      WTDAtest7887999      A32   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA      WTDAtest7887999      A32   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG      WTDAtest7887999      A32   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT      WTDAtest7887999      A32   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA      WTDAtest7887999      A32   \n",
       "...                                               ...      ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG  WSSS_A_LNG8757929      A42   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT  WSSS_A_LNG8757929      A42   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA  WSSS_A_LNG8757929      A42   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA  WSSS_A_LNG8757929      A42   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC  WSSS_A_LNG8757929      A42   \n",
       "\n",
       "                                                                         protocol_URL  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA                                                  N/A   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA                                                  N/A   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG                                                  N/A   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT                                                  N/A   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA                                                  N/A   \n",
       "...                                                                               ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG  https://www.protocols.io/view/single-cell-and-...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT  https://www.protocols.io/view/single-cell-and-...   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA  https://www.protocols.io/view/single-cell-and-...   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA  https://www.protocols.io/view/single-cell-and-...   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC  https://www.protocols.io/view/single-cell-and-...   \n",
       "\n",
       "                                                    institute  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA    Wellcome Sanger Institute   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA    Wellcome Sanger Institute   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG    Wellcome Sanger Institute   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT    Wellcome Sanger Institute   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA    Wellcome Sanger Institute   \n",
       "...                                                       ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG  Wellcome Sanger Institute   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT  Wellcome Sanger Institute   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA  Wellcome Sanger Institute   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA  Wellcome Sanger Institute   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC  Wellcome Sanger Institute   \n",
       "\n",
       "                                   sample_collection_site  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA                Cambridge   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA                Cambridge   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG                Cambridge   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT                Cambridge   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA                Cambridge   \n",
       "...                                                   ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG              Cambridge   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT              Cambridge   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA              Cambridge   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA              Cambridge   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC              Cambridge   \n",
       "\n",
       "                                   sample_collection_relative_time_point  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA                                     N/A   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA                                     N/A   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG                                     N/A   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT                                     N/A   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA                                     N/A   \n",
       "...                                                                  ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG                                   N/A   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT                                   N/A   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA                                   N/A   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA                                   N/A   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC                                   N/A   \n",
       "\n",
       "                                            library_ID library_ID_repository  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA    A32-LNG-1-SC-45N-1                   N/A   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA    A32-LNG-1-SC-45N-1                   N/A   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG    A32-LNG-1-SC-45N-1                   N/A   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT    A32-LNG-1-SC-45N-1                   N/A   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA    A32-LNG-1-SC-45N-1                   N/A   \n",
       "...                                                ...                   ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG   WSSS_A_LNG8757929                   N/A   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT   WSSS_A_LNG8757929                   N/A   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA   WSSS_A_LNG8757929                   N/A   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA   WSSS_A_LNG8757929                   N/A   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC   WSSS_A_LNG8757929                   N/A   \n",
       "\n",
       "                                   author_batch_notes  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA            A32_cells   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA            A32_cells   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG            A32_cells   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT            A32_cells   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA            A32_cells   \n",
       "...                                               ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG         A42_nuclei   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT         A42_nuclei   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA         A42_nuclei   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA         A42_nuclei   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC         A42_nuclei   \n",
       "\n",
       "                                   organism_ontology_term_id  ...  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA              NCBITaxon:9606  ...   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA              NCBITaxon:9606  ...   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG              NCBITaxon:9606  ...   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT              NCBITaxon:9606  ...   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA              NCBITaxon:9606  ...   \n",
       "...                                                      ...  ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG            NCBITaxon:9606  ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT            NCBITaxon:9606  ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA            NCBITaxon:9606  ...   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA            NCBITaxon:9606  ...   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC            NCBITaxon:9606  ...   \n",
       "\n",
       "                                   sequenced_fragment sequencing_platform  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA          3 prime tag         EFO:0008563   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA          3 prime tag         EFO:0008563   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG          3 prime tag         EFO:0008563   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT          3 prime tag         EFO:0008563   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA          3 prime tag         EFO:0008563   \n",
       "...                                               ...                 ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG                 na         EFO:0008637   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT                 na         EFO:0008637   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA                 na         EFO:0008637   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA                 na         EFO:0008637   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC                 na         EFO:0008637   \n",
       "\n",
       "                                   is_primary_data reference_genome  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA              true           GRCh38   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA              true           GRCh38   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG              true           GRCh38   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT              true           GRCh38   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA              true           GRCh38   \n",
       "...                                            ...              ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG            true           GRCh38   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT            true           GRCh38   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA            true           GRCh38   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA            true           GRCh38   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC            true           GRCh38   \n",
       "\n",
       "                                   gene_annotation_version alignment_software  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA                Ensembl 93  cell ranger 3.0.2   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA                Ensembl 93  cell ranger 3.0.2   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG                Ensembl 93  cell ranger 3.0.2   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT                Ensembl 93  cell ranger 3.0.2   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA                Ensembl 93  cell ranger 3.0.2   \n",
       "...                                                    ...                ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG              Ensembl 93  cell ranger 3.0.2   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT              Ensembl 93  cell ranger 3.0.2   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA              Ensembl 93  cell ranger 3.0.2   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA              Ensembl 93  cell ranger 3.0.2   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC              Ensembl 93  cell ranger 3.0.2   \n",
       "\n",
       "                                   intron_inclusion disease_ontology_term_id  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA                 no             PATO:0000461   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA                 no             PATO:0000461   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG                 no             PATO:0000461   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT                 no             PATO:0000461   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA                 no             PATO:0000461   \n",
       "...                                             ...                      ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG              yes             PATO:0000461   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT              yes             PATO:0000461   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA              yes             PATO:0000461   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA              yes             PATO:0000461   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC              yes             PATO:0000461   \n",
       "\n",
       "                                   self_reported_ethnicity_ontology_term_id  \\\n",
       "WTDAtest7887999-AAACCTGGTGTGAATA                                    unknown   \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA                                    unknown   \n",
       "WTDAtest7887999-AACCATGGTACAGCAG                                    unknown   \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT                                    unknown   \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA                                    unknown   \n",
       "...                                                                     ...   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG                                  unknown   \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT                                  unknown   \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA                                  unknown   \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA                                  unknown   \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC                                  unknown   \n",
       "\n",
       "                                   development_stage_ontology_term_id  \n",
       "WTDAtest7887999-AAACCTGGTGTGAATA                       HsapDv:0000237  \n",
       "WTDAtest7887999-AAACGGGCAACCGCCA                       HsapDv:0000237  \n",
       "WTDAtest7887999-AACCATGGTACAGCAG                       HsapDv:0000237  \n",
       "WTDAtest7887999-AACGTTGGTGTCCTCT                       HsapDv:0000237  \n",
       "WTDAtest7887999-AACTCCCTCCTAGTGA                       HsapDv:0000237  \n",
       "...                                                               ...  \n",
       "WSSS_A_LNG8757929-TTTGGTTGTAATGCGG                     HsapDv:0000241  \n",
       "WSSS_A_LNG8757929-TTTGGTTGTGCCCTTT                     HsapDv:0000241  \n",
       "WSSS_A_LNG8757929-TTTGGTTTCAAGAGTA                     HsapDv:0000241  \n",
       "WSSS_A_LNG8757929-TTTGTTGAGCGAGTCA                     HsapDv:0000241  \n",
       "WSSS_A_LNG8757929-TTTGTTGTCTTCTGGC                     HsapDv:0000241  \n",
       "\n",
       "[193108 rows x 37 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = adata,\n",
    "    axis = 'obs'\n",
    ")\n",
    "\n",
    "validated_obs = val_workflow.init_workflow()\n",
    "validated_obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add author cell type markers to UNS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['author_cell_type_markers', 'study_PI', 'title', 'unpublished'])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Already in object\n",
    "adata.uns.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>author_cell_type</th>\n",
       "      <th>author_cell_type_markers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1</td>\n",
       "      <td>AGER;RTKN2;CLIC5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT2</td>\n",
       "      <td>SFTPC;SFTPA1;SFTPA2;WIF1;HHIP;CA2;ETV5;WIF1;HHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B_memory</td>\n",
       "      <td>CD27;TNFRSF13B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B_naive</td>\n",
       "      <td>IGHD;FCER2;TCL1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B_plasma_IgA</td>\n",
       "      <td>IGHA1;IGHA2;CCR10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Secretory_Club</td>\n",
       "      <td>SCGB3A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Secretory_Goblet</td>\n",
       "      <td>MUC5AC;TSPAN8;CYP2F1;CEACAM5;VSIG2;FUT6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Suprabasal</td>\n",
       "      <td>LY6D;PLAT;SERPINB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>T_reg</td>\n",
       "      <td>FOXP3;CCR4;CTLA4;IL2RA;TNFRSF4;TIGIT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>gdT</td>\n",
       "      <td>KLRC2;TRDC;TRDV1;TRG-AS1;TRGC1;KIR2DL4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    author_cell_type                          author_cell_type_markers\n",
       "0               AT1                                   AGER;RTKN2;CLIC5\n",
       "1                AT2  SFTPC;SFTPA1;SFTPA2;WIF1;HHIP;CA2;ETV5;WIF1;HHIP\n",
       "2           B_memory                                    CD27;TNFRSF13B\n",
       "3            B_naive                                  IGHD;FCER2;TCL1A\n",
       "4       B_plasma_IgA                                 IGHA1;IGHA2;CCR10\n",
       "..               ...                                               ...\n",
       "75    Secretory_Club                                           SCGB3A2\n",
       "76  Secretory_Goblet           MUC5AC;TSPAN8;CYP2F1;CEACAM5;VSIG2;FUT6\n",
       "77        Suprabasal                                LY6D;PLAT;SERPINB4\n",
       "78             T_reg              FOXP3;CCR4;CTLA4;IL2RA;TNFRSF4;TIGIT\n",
       "79               gdT            KLRC2;TRDC;TRDV1;TRG-AS1;TRGC1;KIR2DL4\n",
       "\n",
       "[80 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.uns[MARKER_GENES]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check author cell type annotations and Cell Ontology IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(adata.uns[MARKER_GENES][AUTHOR_CELL_TYPE]).symmetric_difference(set(adata.obs[AUTHOR_CELL_TYPE]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SFTPC;SFTPA1;SFTPA2;WIF1;HHIP;CA2;ETV5;WIF1;HHIP              18192\n",
       "MARCO;MCEMP1;INHBA;TREM1;ABHD5;PPARG;RETN;CD5L;FABP4          17098\n",
       "SFRP2;PI16;FBLN2;CD248;MFAP5                                  10170\n",
       "CA4;FCN3;SLC6A4;IL7R                                           9348\n",
       "FCER1G;GNLY;KLRF1;KIR2DL1;GZMB;FGFBP2;NKG7                     8116\n",
       "                                                              ...  \n",
       "CXCL13;FDCSP                                                     59\n",
       "FOXI1;CFTR;ASCL3;HEPACAM2;PLCG2;BIK                              54\n",
       "CHGA;CALCA;ASCL1;CHGB;GRP;BEX1                                   44\n",
       "TUBB1;ANK1;PF4;TUBB1;CMTM5;PCSK6;STON2;PRKAR2B;SYTL4;LTBP1       14\n",
       "GLDN;CDH7;DRP2;NFASC;NCMAP;MBP;PRX;MLIP                          11\n",
       "Name: author_cell_type_markers, Length: 79, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[MARKER_GENES] = adata.obs[AUTHOR_CELL_TYPE].map(dict(zip(adata.uns[MARKER_GENES][AUTHOR_CELL_TYPE], adata.uns[MARKER_GENES][MARKER_GENES])))\n",
    "adata.obs[MARKER_GENES].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AT2                          18192\n",
       "Macro_alveolar               17098\n",
       "Fibro_adventitial            10170\n",
       "Endothelia_vascular_Cap_g     9348\n",
       "NK_CD16hi                     8116\n",
       "                             ...  \n",
       "Fibro_immune_recruiting         59\n",
       "Ionocyte_n_Brush                54\n",
       "Neuroendocrine                  44\n",
       "Megakaryocyte                   14\n",
       "Schwann_Myelinating             11\n",
       "Name: author_cell_type, Length: 80, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Already in object\n",
    "adata.obs[AUTHOR_CELL_TYPE].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CL:0002063    18372\n",
       "CL:0000583    17098\n",
       "CL:0002503    10170\n",
       "CL:0000235    10108\n",
       "CL:4028002     9348\n",
       "              ...  \n",
       "CL:0000980       62\n",
       "CL:0005006       54\n",
       "CL:0000165       44\n",
       "CL:0000556       14\n",
       "CL:0000218       11\n",
       "Name: cell_type_ontology_term_id, Length: 65, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Already in object\n",
    "adata.obs[CELL_TYPE_ONTOLOGY_ID].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check whether ENSEMBL IDs in var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_types</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_ids</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <td>AC233755.2</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <td>AC233755.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277475</th>\n",
       "      <td>AC213203.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268674</th>\n",
       "      <td>FAM231C</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33538 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_symbol    feature_types\n",
       "gene_ids                                     \n",
       "ENSG00000243485  MIR1302-2HG  Gene Expression\n",
       "ENSG00000237613      FAM138A  Gene Expression\n",
       "ENSG00000186092        OR4F5  Gene Expression\n",
       "ENSG00000238009   AL627309.1  Gene Expression\n",
       "ENSG00000239945   AL627309.3  Gene Expression\n",
       "...                      ...              ...\n",
       "ENSG00000277856   AC233755.2  Gene Expression\n",
       "ENSG00000275063   AC233755.1  Gene Expression\n",
       "ENSG00000271254   AC240274.1  Gene Expression\n",
       "ENSG00000277475   AC213203.1  Gene Expression\n",
       "ENSG00000268674      FAM231C  Gene Expression\n",
       "\n",
       "[33538 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <td>AC233755.2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000277856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <td>AC233755.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000275063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277475</th>\n",
       "      <td>AC213203.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000277475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268674</th>\n",
       "      <td>FAM231C</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000268674</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33538 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_symbol    feature_types       ensembl_id\n",
       "index                                                         \n",
       "ENSG00000243485  MIR1302-2HG  Gene Expression  ENSG00000243485\n",
       "ENSG00000237613      FAM138A  Gene Expression  ENSG00000237613\n",
       "ENSG00000186092        OR4F5  Gene Expression  ENSG00000186092\n",
       "ENSG00000238009   AL627309.1  Gene Expression  ENSG00000238009\n",
       "ENSG00000239945   AL627309.3  Gene Expression  ENSG00000239945\n",
       "...                      ...              ...              ...\n",
       "ENSG00000277856   AC233755.2  Gene Expression  ENSG00000277856\n",
       "ENSG00000275063   AC233755.1  Gene Expression  ENSG00000275063\n",
       "ENSG00000271254   AC240274.1  Gene Expression  ENSG00000271254\n",
       "ENSG00000277475   AC213203.1  Gene Expression  ENSG00000277475\n",
       "ENSG00000268674      FAM231C  Gene Expression  ENSG00000268674\n",
       "\n",
       "[33538 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var['ensembl_id'] = adata.var.index\n",
    "adata.var.index.name = 'index'\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14469.0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.toarray().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['cell_viability_percentage'] = pd.to_numeric(adata.obs['cell_viability_percentage'], errors='coerce')\n",
    "adata.obs['cell_number_loaded'] = pd.to_numeric(adata.obs['cell_number_loaded'], errors='coerce')\n",
    "adata.obs['sample_collection_year'] = pd.to_numeric(adata.obs['sample_collection_year'], errors='coerce')\n",
    "adata.obs['manner_of_death'] = pd.to_numeric(adata.obs['manner_of_death'], errors='coerce')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.X.astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw = adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<193108x33538 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 282988531 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<193108x33538 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 282988531 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14469"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.toarray().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14469"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.X.toarray().max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validation result\n",
    "\n",
    "### UNS Validation\n",
    "- OK: Tier 1 UNS Google Sheet\n",
    "- OK: Tier 1 UNS AnnData Object\n",
    "\n",
    "### OBS Validation\n",
    "- Validation Error: Tier 1 OBS Google Sheet: 'Frozen' is not among available categories\n",
    "- Validation Error: Tier 1 OBS Anndata Object: 'Frozen' is not among available categories\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Submission Status\n",
    "\n",
    "- CHECK: Raw counts in X and in raw\n",
    "- REVISE: Tier 1 Metadata in OBS\n",
    "- CHECK: Cell Ontology IDs in OBS\n",
    "- CHECK: Author cell type in OBS\n",
    "- CHECK: Marker genes in UNS\n",
    "- CHECK: ENSEMBL IDs in var\n",
    "\n",
    "### Revision:\n",
    "\n",
    "DONE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(join(OUTPUT_PATH_PREREVISION, f\"{DATASET_ID}.h5ad\"), compression='gzip')\n",
    "adata.write_h5ad(join(OUTPUT_PATH_POSTREVISION, f\"{DATASET_ID}.h5ad\"), compression='gzip')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hlcav2-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
