{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import anndata2ri\n",
    "import os\n",
    "from os.path import join\n",
    "import scipy.sparse as sp\n",
    "import numpy as np\n",
    "\n",
    "from hlca_v2.ingestion_utils import get_gspread_df, ValidationWorkflow, AnnDataMerger\n",
    "\n",
    "GSPREAD_JSON = \"/home/icb/raphael.kfuri-rubens/data/hlca_v2/google_sheets_api/hlca-v2-8d5fea10d8f3.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_ID = \"Schiller_unpubl\"\n",
    "H5AD_PATH = f\"/home/icb/raphael.kfuri-rubens/data/hlca_v2/{DATASET_ID}/{DATASET_ID}.h5ad\"\n",
    "MARKER_GENES_PATH = f\"/home/icb/raphael.kfuri-rubens/data/hlca_v2/{DATASET_ID}/marker_genes_cell_ontologies.xlsx\"\n",
    "OUTPUT_PATH_PREREVISION = '/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_prerevision'\n",
    "OUTPUT_PATH_POSTREVISION = '/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_postrevision'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Name constants\n",
    "AUTHOR_CELL_TYPE_L0 = 'author_cell_type_level_0'\n",
    "AUTHOR_CELL_TYPE_L1 = 'author_cell_type_level_1'\n",
    "AUTHOR_CELL_TYPE_L2 = 'author_cell_type_level_2'\n",
    "\n",
    "CELL_TYPE_ONTOLOGY_ID_L0 = 'cell_type_ontology_term_id_level_0'\n",
    "CELL_TYPE_ONTOLOGY_ID_L1 = 'cell_type_ontology_term_id_level_1'\n",
    "\n",
    "CELL_TYPE_ONTOLOGY_LABEL_L0 = 'cell_type_ontology_term_label_level_0'\n",
    "CELL_TYPE_ONTOLOGY_LABEL_L1 = 'cell_type_ontology_term_label_level_1'\n",
    "\n",
    "AUTHOR_CELL_TYPE_DESCRIPTION_L0 = 'author_cell_type_description_level_0'\n",
    "AUTHOR_CELL_TYPE_DESCRIPTION_L1 = 'author_cell_type_description_level_1'\n",
    "\n",
    "MARKER_GENES_L0 = 'author_cell_type_markers_level_0'\n",
    "MARKER_GENES_L1 = 'author_cell_type_markers_level_1'\n",
    "\n",
    "# Finest grained annotation will be generic dataset cell type\n",
    "AUTHOR_CELL_TYPE = 'author_cell_type'\n",
    "CELL_TYPE_ONTOLOGY_ID = 'cell_type_ontology_term_id'\n",
    "CELL_TYPE_ONTOLOGY_LABEL = 'cell_type_ontology_term_label'\n",
    "MARKER_GENES = 'author_cell_type_markers'\n",
    "AUTHOR_CELL_TYPE_DESCRIPTION = 'author_cell_type_description'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(H5AD_PATH)\n",
    "obs = get_gspread_df(GSPREAD_JSON, DATASET_ID, \"tier_1\", \"obs\")\n",
    "uns = get_gspread_df(GSPREAD_JSON, DATASET_ID, \"tier_1\", \"uns\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validate obs and uns from Tier 1 Metadata Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation workflow successful. Returning UNS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation workflow successful. Returning UNS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>study_PI</th>\n",
       "      <th>batch_condition</th>\n",
       "      <th>default_embedding</th>\n",
       "      <th>unpublished</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Schiller_single_nuc</td>\n",
       "      <td>Herbert Schiller</td>\n",
       "      <td>condition, donor_id</td>\n",
       "      <td>X_hlca</td>\n",
       "      <td>protected under embargo</td>\n",
       "      <td>Samples from patients with the same health sta...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 title          study_PI      batch_condition  \\\n",
       "0  Schiller_single_nuc  Herbert Schiller  condition, donor_id   \n",
       "\n",
       "  default_embedding              unpublished  \\\n",
       "0            X_hlca  protected under embargo   \n",
       "\n",
       "                                            comments  \n",
       "0  Samples from patients with the same health sta...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = uns,\n",
    "    axis = 'uns'\n",
    ")\n",
    "\n",
    "validated_uns = val_workflow.init_workflow()\n",
    "validated_uns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation <span style=\"color: #800080; text-decoration-color: #800080; font-weight: bold\">error</span><span style=\"font-weight: bold\">(</span>s<span style=\"font-weight: bold\">)</span> encountered.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation \u001b[1;35merror\u001b[0m\u001b[1m(\u001b[0ms\u001b[1m)\u001b[0m encountered.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">{</span>\n",
       "<span style=\"color: #7fbf7f; text-decoration-color: #7fbf7f\">│   </span><span style=\"color: #008000; text-decoration-color: #008000\">'development_stage_ontology_term_id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"Must start with 'HsapDv:' or be 'na', 'NA', or 'N/A'. Invalid entries found.\"</span>\n",
       "<span style=\"font-weight: bold\">}</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1m{\u001b[0m\n",
       "\u001b[2;32m│   \u001b[0m\u001b[32m'development_stage_ontology_term_id'\u001b[0m: \u001b[32m\"Must start with 'HsapDv:' or be 'na', 'NA', or 'N/A'. Invalid entries found.\"\u001b[0m\n",
       "\u001b[1m}\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Error containing field: development_stage_ontology_term_id\n",
       "Error message: Must start with <span style=\"color: #008000; text-decoration-color: #008000\">'HsapDv:'</span> or be <span style=\"color: #008000; text-decoration-color: #008000\">'na'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'NA'</span>, or <span style=\"color: #008000; text-decoration-color: #008000\">'N/A'</span>. Invalid entries found.\n",
       "Value counts:\n",
       "HsapDv:<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0000242</span>    <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">5</span>\n",
       "HsapDv:<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0000241</span>    <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4</span>\n",
       "unknown           <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4</span>\n",
       "HsapDv:<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0000240</span>    <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3</span>\n",
       "HsapDv:<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0000239</span>    <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1</span>\n",
       "Name: development_stage_ontology_term_id, dtype: int64\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Error containing field: development_stage_ontology_term_id\n",
       "Error message: Must start with \u001b[32m'HsapDv:'\u001b[0m or be \u001b[32m'na'\u001b[0m, \u001b[32m'NA'\u001b[0m, or \u001b[32m'N/A'\u001b[0m. Invalid entries found.\n",
       "Value counts:\n",
       "HsapDv:\u001b[1;36m0000242\u001b[0m    \u001b[1;36m5\u001b[0m\n",
       "HsapDv:\u001b[1;36m0000241\u001b[0m    \u001b[1;36m4\u001b[0m\n",
       "unknown           \u001b[1;36m4\u001b[0m\n",
       "HsapDv:\u001b[1;36m0000240\u001b[0m    \u001b[1;36m3\u001b[0m\n",
       "HsapDv:\u001b[1;36m0000239\u001b[0m    \u001b[1;36m1\u001b[0m\n",
       "Name: development_stage_ontology_term_id, dtype: int64\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Returning error-containing OBS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Returning error-containing OBS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>protocol_URL</th>\n",
       "      <th>institute</th>\n",
       "      <th>sample_collection_site</th>\n",
       "      <th>sample_collection_relative_time_point</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>library_ID_repository</th>\n",
       "      <th>author_batch_notes</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sequenced_fragment</th>\n",
       "      <th>sequencing_platform</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>reference_genome</th>\n",
       "      <th>gene_annotation_version</th>\n",
       "      <th>alignment_software</th>\n",
       "      <th>intron_inclusion</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASK612</td>\n",
       "      <td>ASK612</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>Non-Smoker</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASK620</td>\n",
       "      <td>ASK620</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>Non-Smoker</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASK633</td>\n",
       "      <td>ASK633</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>Non-Smoker</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASK621</td>\n",
       "      <td>ASK621</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>Smoker</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASK627</td>\n",
       "      <td>ASK627</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>Smoker</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ASK634</td>\n",
       "      <td>ASK634</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>Smoker</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ASK543</td>\n",
       "      <td>ASK543</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>COPDII</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ASK611</td>\n",
       "      <td>ASK611</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>COPDII</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ASK631</td>\n",
       "      <td>ASK631</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>COPDII</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>M434WRC</td>\n",
       "      <td>M434WRC</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>LMU Klinikum, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>COPDIV</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MLT220</td>\n",
       "      <td>MLT220</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>LMU Klinikum, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>COPDIV</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MLT251</td>\n",
       "      <td>MLT251</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>LMU Klinikum, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>COPDIV</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MLT255</td>\n",
       "      <td>MLT255</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>LMU Klinikum, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>COPDIV</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>COPDIV</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>LMU Klinikum, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>COPDIV</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>COPDII</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>Smoker</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Non-Smoker</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>same multiplexed protocol, unpublished</td>\n",
       "      <td>CPC/LHI, Helmholtz Center Munich</td>\n",
       "      <td>Asklepios Lungenklinik Gauting, Munich</td>\n",
       "      <td>day1</td>\n",
       "      <td>Non-Smoker</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>same personnel, batch 1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_ID    donor_id                            protocol_URL  \\\n",
       "0       ASK612      ASK612  same multiplexed protocol, unpublished   \n",
       "1       ASK620      ASK620  same multiplexed protocol, unpublished   \n",
       "2       ASK633      ASK633  same multiplexed protocol, unpublished   \n",
       "3       ASK621      ASK621  same multiplexed protocol, unpublished   \n",
       "4       ASK627      ASK627  same multiplexed protocol, unpublished   \n",
       "5       ASK634      ASK634  same multiplexed protocol, unpublished   \n",
       "6       ASK543      ASK543  same multiplexed protocol, unpublished   \n",
       "7       ASK611      ASK611  same multiplexed protocol, unpublished   \n",
       "8       ASK631      ASK631  same multiplexed protocol, unpublished   \n",
       "9      M434WRC     M434WRC  same multiplexed protocol, unpublished   \n",
       "10      MLT220      MLT220  same multiplexed protocol, unpublished   \n",
       "11      MLT251      MLT251  same multiplexed protocol, unpublished   \n",
       "12      MLT255      MLT255  same multiplexed protocol, unpublished   \n",
       "13      COPDIV  unassigned  same multiplexed protocol, unpublished   \n",
       "14      COPDII  unassigned  same multiplexed protocol, unpublished   \n",
       "15      Smoker  unassigned  same multiplexed protocol, unpublished   \n",
       "16  Non-Smoker  unassigned  same multiplexed protocol, unpublished   \n",
       "\n",
       "                           institute                  sample_collection_site  \\\n",
       "0   CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "1   CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "2   CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "3   CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "4   CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "5   CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "6   CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "7   CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "8   CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "9   CPC/LHI, Helmholtz Center Munich                    LMU Klinikum, Munich   \n",
       "10  CPC/LHI, Helmholtz Center Munich                    LMU Klinikum, Munich   \n",
       "11  CPC/LHI, Helmholtz Center Munich                    LMU Klinikum, Munich   \n",
       "12  CPC/LHI, Helmholtz Center Munich                    LMU Klinikum, Munich   \n",
       "13  CPC/LHI, Helmholtz Center Munich                    LMU Klinikum, Munich   \n",
       "14  CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "15  CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "16  CPC/LHI, Helmholtz Center Munich  Asklepios Lungenklinik Gauting, Munich   \n",
       "\n",
       "   sample_collection_relative_time_point  library_ID library_ID_repository  \\\n",
       "0                                   day1  Non-Smoker        not applicable   \n",
       "1                                   day1  Non-Smoker        not applicable   \n",
       "2                                   day1  Non-Smoker        not applicable   \n",
       "3                                   day1      Smoker        not applicable   \n",
       "4                                   day1      Smoker        not applicable   \n",
       "5                                   day1      Smoker        not applicable   \n",
       "6                                   day1      COPDII        not applicable   \n",
       "7                                   day1      COPDII        not applicable   \n",
       "8                                   day1      COPDII        not applicable   \n",
       "9                                   day1      COPDIV        not applicable   \n",
       "10                                  day1      COPDIV        not applicable   \n",
       "11                                  day1      COPDIV        not applicable   \n",
       "12                                  day1      COPDIV        not applicable   \n",
       "13                                  day1      COPDIV        not applicable   \n",
       "14                                  day1      COPDII        not applicable   \n",
       "15                                  day1      Smoker        not applicable   \n",
       "16                                  day1  Non-Smoker        not applicable   \n",
       "\n",
       "         author_batch_notes organism_ontology_term_id  ... sequenced_fragment  \\\n",
       "0   same personnel, batch 1            NCBITaxon:9606  ...        3 prime tag   \n",
       "1   same personnel, batch 1            NCBITaxon:9606  ...        3 prime tag   \n",
       "2   same personnel, batch 1            NCBITaxon:9606  ...        3 prime tag   \n",
       "3   same personnel, batch 1            NCBITaxon:9606  ...        3 prime tag   \n",
       "4   same personnel, batch 1            NCBITaxon:9606  ...        3 prime tag   \n",
       "5   same personnel, batch 1            NCBITaxon:9606  ...        3 prime tag   \n",
       "6   same personnel, batch 2            NCBITaxon:9606  ...        3 prime tag   \n",
       "7   same personnel, batch 2            NCBITaxon:9606  ...        3 prime tag   \n",
       "8   same personnel, batch 2            NCBITaxon:9606  ...        3 prime tag   \n",
       "9   same personnel, batch 2            NCBITaxon:9606  ...        3 prime tag   \n",
       "10  same personnel, batch 2            NCBITaxon:9606  ...        3 prime tag   \n",
       "11  same personnel, batch 2            NCBITaxon:9606  ...        3 prime tag   \n",
       "12  same personnel, batch 2            NCBITaxon:9606  ...        3 prime tag   \n",
       "13  same personnel, batch 2            NCBITaxon:9606  ...        3 prime tag   \n",
       "14  same personnel, batch 2            NCBITaxon:9606  ...        3 prime tag   \n",
       "15  same personnel, batch 1            NCBITaxon:9606  ...        3 prime tag   \n",
       "16  same personnel, batch 1            NCBITaxon:9606  ...        3 prime tag   \n",
       "\n",
       "   sequencing_platform is_primary_data reference_genome  \\\n",
       "0          EFO:0008637            true           GRCh38   \n",
       "1          EFO:0008637            true           GRCh38   \n",
       "2          EFO:0008637            true           GRCh38   \n",
       "3          EFO:0008637            true           GRCh38   \n",
       "4          EFO:0008637            true           GRCh38   \n",
       "5          EFO:0008637            true           GRCh38   \n",
       "6          EFO:0008637            true           GRCh38   \n",
       "7          EFO:0008637            true           GRCh38   \n",
       "8          EFO:0008637            true           GRCh38   \n",
       "9          EFO:0008637            true           GRCh38   \n",
       "10         EFO:0008637            true           GRCh38   \n",
       "11         EFO:0008637            true           GRCh38   \n",
       "12         EFO:0008637            true           GRCh38   \n",
       "13         EFO:0008637            true           GRCh38   \n",
       "14         EFO:0008637            true           GRCh38   \n",
       "15         EFO:0008637            true           GRCh38   \n",
       "16         EFO:0008637            true           GRCh38   \n",
       "\n",
       "   gene_annotation_version  alignment_software intron_inclusion  \\\n",
       "0                     v107  cell ranger 6.1.2.              yes   \n",
       "1                     v107  cell ranger 6.1.2.              yes   \n",
       "2                     v107  cell ranger 6.1.2.              yes   \n",
       "3                     v107  cell ranger 6.1.2.              yes   \n",
       "4                     v107  cell ranger 6.1.2.              yes   \n",
       "5                     v107  cell ranger 6.1.2.              yes   \n",
       "6                     v107  cell ranger 6.1.2.              yes   \n",
       "7                     v107  cell ranger 6.1.2.              yes   \n",
       "8                     v107  cell ranger 6.1.2.              yes   \n",
       "9                     v107  cell ranger 6.1.2.              yes   \n",
       "10                    v107  cell ranger 6.1.2.              yes   \n",
       "11                    v107  cell ranger 6.1.2.              yes   \n",
       "12                    v107  cell ranger 6.1.2.              yes   \n",
       "13                    v107  cell ranger 6.1.2.              yes   \n",
       "14                    v107  cell ranger 6.1.2.              yes   \n",
       "15                    v107  cell ranger 6.1.2.              yes   \n",
       "16                    v107  cell ranger 6.1.2.              yes   \n",
       "\n",
       "   disease_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0              PATO:0000461                                  unknown   \n",
       "1              PATO:0000461                                  unknown   \n",
       "2              PATO:0000461                                  unknown   \n",
       "3              PATO:0000461                                  unknown   \n",
       "4              PATO:0000461                                  unknown   \n",
       "5              PATO:0000461                                  unknown   \n",
       "6             MONDO:0005002                                  unknown   \n",
       "7             MONDO:0005002                                  unknown   \n",
       "8             MONDO:0005002                                  unknown   \n",
       "9             MONDO:0005002                                  unknown   \n",
       "10            MONDO:0005002                                  unknown   \n",
       "11            MONDO:0005002                                  unknown   \n",
       "12            MONDO:0005002                                  unknown   \n",
       "13            MONDO:0005002                                  unknown   \n",
       "14            MONDO:0005002                                  unknown   \n",
       "15             PATO:0000461                                  unknown   \n",
       "16             PATO:0000461                                  unknown   \n",
       "\n",
       "   development_stage_ontology_term_id  \n",
       "0                      HsapDv:0000240  \n",
       "1                      HsapDv:0000242  \n",
       "2                      HsapDv:0000242  \n",
       "3                      HsapDv:0000242  \n",
       "4                      HsapDv:0000242  \n",
       "5                      HsapDv:0000240  \n",
       "6                      HsapDv:0000239  \n",
       "7                      HsapDv:0000241  \n",
       "8                      HsapDv:0000242  \n",
       "9                      HsapDv:0000241  \n",
       "10                     HsapDv:0000241  \n",
       "11                     HsapDv:0000240  \n",
       "12                     HsapDv:0000241  \n",
       "13                            unknown  \n",
       "14                            unknown  \n",
       "15                            unknown  \n",
       "16                            unknown  \n",
       "\n",
       "[17 rows x 37 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = obs,\n",
    "    axis = 'obs'\n",
    ")\n",
    "\n",
    "validated_obs = val_workflow.init_workflow()\n",
    "validated_obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validate obs and uns from adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column donor_id already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column donor_id already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column protocol_URL already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column protocol_URL already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column institute already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column institute already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sample_collection_site already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sample_collection_site already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sample_collection_relative_time_point already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sample_collection_relative_time_point already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column library_ID already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column library_ID already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column library_ID_repository already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column library_ID_repository already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column author_batch_notes already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column author_batch_notes already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column organism_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column organism_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column manner_of_death already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column manner_of_death already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sample_source already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sample_source already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sex_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sex_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sample_collection_method already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sample_collection_method already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column tissue_type already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column tissue_type already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sampled_site_condition already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sampled_site_condition already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column tissue_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column tissue_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column tissue_free_text already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column tissue_free_text already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sample_preservation_method already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sample_preservation_method already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column suspension_type already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column suspension_type already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column cell_enrichment already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column cell_enrichment already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column cell_viability_percentage already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column cell_viability_percentage already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column cell_number_loaded already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column cell_number_loaded already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sample_collection_year already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sample_collection_year already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column assay_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column assay_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column library_preparation_batch already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column library_preparation_batch already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column library_sequencing_run already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column library_sequencing_run already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sequenced_fragment already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sequenced_fragment already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column sequencing_platform already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column sequencing_platform already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column is_primary_data already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column is_primary_data already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column reference_genome already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column reference_genome already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column gene_annotation_version already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column gene_annotation_version already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column alignment_software already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column alignment_software already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column intron_inclusion already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column intron_inclusion already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column disease_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column disease_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column self_reported_ethnicity_ontology_term_id already exists in AnnData OBS. Overwriting with new \n",
       "values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column self_reported_ethnicity_ontology_term_id already exists in AnnData OBS. Overwriting with new \n",
       "values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Warning: Column development_stage_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Warning: Column development_stage_ontology_term_id already exists in AnnData OBS. Overwriting with new values.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>genotype</th>\n",
       "      <th>status</th>\n",
       "      <th>demultiplex_sample</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>...</th>\n",
       "      <th>sequenced_fragment</th>\n",
       "      <th>sequencing_platform</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>reference_genome</th>\n",
       "      <th>gene_annotation_version</th>\n",
       "      <th>alignment_software</th>\n",
       "      <th>intron_inclusion</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGAAACC-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>singlet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>1617</td>\n",
       "      <td>7.388946</td>\n",
       "      <td>2898.0</td>\n",
       "      <td>7.972121</td>\n",
       "      <td>17.563837</td>\n",
       "      <td>26.293996</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAATCTGCA-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>singlet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>2051</td>\n",
       "      <td>7.626570</td>\n",
       "      <td>3268.0</td>\n",
       "      <td>8.092239</td>\n",
       "      <td>12.301102</td>\n",
       "      <td>19.033048</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACACACTTAG-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>singlet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>1958</td>\n",
       "      <td>7.580189</td>\n",
       "      <td>3132.0</td>\n",
       "      <td>8.049747</td>\n",
       "      <td>14.112388</td>\n",
       "      <td>21.328225</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTACCACGC-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>doublet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>2184</td>\n",
       "      <td>7.689371</td>\n",
       "      <td>3527.0</td>\n",
       "      <td>8.168487</td>\n",
       "      <td>11.256025</td>\n",
       "      <td>17.833853</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTGACTGAG-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>singlet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>1456</td>\n",
       "      <td>7.284135</td>\n",
       "      <td>2514.0</td>\n",
       "      <td>7.830028</td>\n",
       "      <td>19.809069</td>\n",
       "      <td>28.758950</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGAGTCGCTAATG-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>704</td>\n",
       "      <td>6.558198</td>\n",
       "      <td>1130.0</td>\n",
       "      <td>7.030858</td>\n",
       "      <td>23.893805</td>\n",
       "      <td>36.460177</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTGTAGCTAAA-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>728</td>\n",
       "      <td>6.591674</td>\n",
       "      <td>1222.0</td>\n",
       "      <td>7.109062</td>\n",
       "      <td>26.595745</td>\n",
       "      <td>38.952537</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCTCATTGT-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>674</td>\n",
       "      <td>6.514713</td>\n",
       "      <td>1089.0</td>\n",
       "      <td>6.993933</td>\n",
       "      <td>28.466483</td>\n",
       "      <td>40.404040</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCCGTATGA-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>813</td>\n",
       "      <td>6.701960</td>\n",
       "      <td>1661.0</td>\n",
       "      <td>7.415777</td>\n",
       "      <td>31.727875</td>\n",
       "      <td>42.865744</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCGCCTTGT-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>809</td>\n",
       "      <td>6.697034</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>7.073270</td>\n",
       "      <td>20.610687</td>\n",
       "      <td>31.382528</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008637</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v107</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48301 rows × 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   condition genotype   status demultiplex_sample  \\\n",
       "AAACCCAAGCGAAACC-1    COPDII        1  singlet           COPDII_1   \n",
       "AAACCCACAATCTGCA-1    COPDII        1  singlet           COPDII_1   \n",
       "AAACCCACACACTTAG-1    COPDII        1  singlet           COPDII_1   \n",
       "AAACCCAGTACCACGC-1    COPDII        1  doublet           COPDII_1   \n",
       "AAACCCAGTGACTGAG-1    COPDII        1  singlet           COPDII_1   \n",
       "...                      ...      ...      ...                ...   \n",
       "TTTGGAGTCGCTAATG-1    Smoker        0  singlet           Smoker_0   \n",
       "TTTGGTTGTAGCTAAA-1    Smoker        0  singlet           Smoker_0   \n",
       "TTTGGTTTCTCATTGT-1    Smoker        0  singlet           Smoker_0   \n",
       "TTTGTTGTCCGTATGA-1    Smoker        0  singlet           Smoker_0   \n",
       "TTTGTTGTCGCCTTGT-1    Smoker        0  singlet           Smoker_0   \n",
       "\n",
       "                    n_genes_by_counts  log1p_n_genes_by_counts  total_counts  \\\n",
       "AAACCCAAGCGAAACC-1               1617                 7.388946        2898.0   \n",
       "AAACCCACAATCTGCA-1               2051                 7.626570        3268.0   \n",
       "AAACCCACACACTTAG-1               1958                 7.580189        3132.0   \n",
       "AAACCCAGTACCACGC-1               2184                 7.689371        3527.0   \n",
       "AAACCCAGTGACTGAG-1               1456                 7.284135        2514.0   \n",
       "...                               ...                      ...           ...   \n",
       "TTTGGAGTCGCTAATG-1                704                 6.558198        1130.0   \n",
       "TTTGGTTGTAGCTAAA-1                728                 6.591674        1222.0   \n",
       "TTTGGTTTCTCATTGT-1                674                 6.514713        1089.0   \n",
       "TTTGTTGTCCGTATGA-1                813                 6.701960        1661.0   \n",
       "TTTGTTGTCGCCTTGT-1                809                 6.697034        1179.0   \n",
       "\n",
       "                    log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "AAACCCAAGCGAAACC-1            7.972121                   17.563837   \n",
       "AAACCCACAATCTGCA-1            8.092239                   12.301102   \n",
       "AAACCCACACACTTAG-1            8.049747                   14.112388   \n",
       "AAACCCAGTACCACGC-1            8.168487                   11.256025   \n",
       "AAACCCAGTGACTGAG-1            7.830028                   19.809069   \n",
       "...                                ...                         ...   \n",
       "TTTGGAGTCGCTAATG-1            7.030858                   23.893805   \n",
       "TTTGGTTGTAGCTAAA-1            7.109062                   26.595745   \n",
       "TTTGGTTTCTCATTGT-1            6.993933                   28.466483   \n",
       "TTTGTTGTCCGTATGA-1            7.415777                   31.727875   \n",
       "TTTGTTGTCGCCTTGT-1            7.073270                   20.610687   \n",
       "\n",
       "                    pct_counts_in_top_100_genes  ...  sequenced_fragment  \\\n",
       "AAACCCAAGCGAAACC-1                    26.293996  ...         3 prime tag   \n",
       "AAACCCACAATCTGCA-1                    19.033048  ...         3 prime tag   \n",
       "AAACCCACACACTTAG-1                    21.328225  ...         3 prime tag   \n",
       "AAACCCAGTACCACGC-1                    17.833853  ...         3 prime tag   \n",
       "AAACCCAGTGACTGAG-1                    28.758950  ...         3 prime tag   \n",
       "...                                         ...  ...                 ...   \n",
       "TTTGGAGTCGCTAATG-1                    36.460177  ...         3 prime tag   \n",
       "TTTGGTTGTAGCTAAA-1                    38.952537  ...         3 prime tag   \n",
       "TTTGGTTTCTCATTGT-1                    40.404040  ...         3 prime tag   \n",
       "TTTGTTGTCCGTATGA-1                    42.865744  ...         3 prime tag   \n",
       "TTTGTTGTCGCCTTGT-1                    31.382528  ...         3 prime tag   \n",
       "\n",
       "                    sequencing_platform  is_primary_data  reference_genome  \\\n",
       "AAACCCAAGCGAAACC-1          EFO:0008637             true            GRCh38   \n",
       "AAACCCACAATCTGCA-1          EFO:0008637             true            GRCh38   \n",
       "AAACCCACACACTTAG-1          EFO:0008637             true            GRCh38   \n",
       "AAACCCAGTACCACGC-1          EFO:0008637             true            GRCh38   \n",
       "AAACCCAGTGACTGAG-1          EFO:0008637             true            GRCh38   \n",
       "...                                 ...              ...               ...   \n",
       "TTTGGAGTCGCTAATG-1          EFO:0008637             true            GRCh38   \n",
       "TTTGGTTGTAGCTAAA-1          EFO:0008637             true            GRCh38   \n",
       "TTTGGTTTCTCATTGT-1          EFO:0008637             true            GRCh38   \n",
       "TTTGTTGTCCGTATGA-1          EFO:0008637             true            GRCh38   \n",
       "TTTGTTGTCGCCTTGT-1          EFO:0008637             true            GRCh38   \n",
       "\n",
       "                    gene_annotation_version  alignment_software  \\\n",
       "AAACCCAAGCGAAACC-1                     v107  cell ranger 6.1.2.   \n",
       "AAACCCACAATCTGCA-1                     v107  cell ranger 6.1.2.   \n",
       "AAACCCACACACTTAG-1                     v107  cell ranger 6.1.2.   \n",
       "AAACCCAGTACCACGC-1                     v107  cell ranger 6.1.2.   \n",
       "AAACCCAGTGACTGAG-1                     v107  cell ranger 6.1.2.   \n",
       "...                                     ...                 ...   \n",
       "TTTGGAGTCGCTAATG-1                     v107  cell ranger 6.1.2.   \n",
       "TTTGGTTGTAGCTAAA-1                     v107  cell ranger 6.1.2.   \n",
       "TTTGGTTTCTCATTGT-1                     v107  cell ranger 6.1.2.   \n",
       "TTTGTTGTCCGTATGA-1                     v107  cell ranger 6.1.2.   \n",
       "TTTGTTGTCGCCTTGT-1                     v107  cell ranger 6.1.2.   \n",
       "\n",
       "                    intron_inclusion  disease_ontology_term_id  \\\n",
       "AAACCCAAGCGAAACC-1               yes             MONDO:0005002   \n",
       "AAACCCACAATCTGCA-1               yes             MONDO:0005002   \n",
       "AAACCCACACACTTAG-1               yes             MONDO:0005002   \n",
       "AAACCCAGTACCACGC-1               yes             MONDO:0005002   \n",
       "AAACCCAGTGACTGAG-1               yes             MONDO:0005002   \n",
       "...                              ...                       ...   \n",
       "TTTGGAGTCGCTAATG-1               yes              PATO:0000461   \n",
       "TTTGGTTGTAGCTAAA-1               yes              PATO:0000461   \n",
       "TTTGGTTTCTCATTGT-1               yes              PATO:0000461   \n",
       "TTTGTTGTCCGTATGA-1               yes              PATO:0000461   \n",
       "TTTGTTGTCGCCTTGT-1               yes              PATO:0000461   \n",
       "\n",
       "                    self_reported_ethnicity_ontology_term_id  \\\n",
       "AAACCCAAGCGAAACC-1                                   unknown   \n",
       "AAACCCACAATCTGCA-1                                   unknown   \n",
       "AAACCCACACACTTAG-1                                   unknown   \n",
       "AAACCCAGTACCACGC-1                                   unknown   \n",
       "AAACCCAGTGACTGAG-1                                   unknown   \n",
       "...                                                      ...   \n",
       "TTTGGAGTCGCTAATG-1                                   unknown   \n",
       "TTTGGTTGTAGCTAAA-1                                   unknown   \n",
       "TTTGGTTTCTCATTGT-1                                   unknown   \n",
       "TTTGTTGTCCGTATGA-1                                   unknown   \n",
       "TTTGTTGTCGCCTTGT-1                                   unknown   \n",
       "\n",
       "                   development_stage_ontology_term_id  \n",
       "AAACCCAAGCGAAACC-1                     HsapDv:0000239  \n",
       "AAACCCACAATCTGCA-1                            unknown  \n",
       "AAACCCACACACTTAG-1                     HsapDv:0000239  \n",
       "AAACCCAGTACCACGC-1                            unknown  \n",
       "AAACCCAGTGACTGAG-1                     HsapDv:0000239  \n",
       "...                                               ...  \n",
       "TTTGGAGTCGCTAATG-1                            unknown  \n",
       "TTTGGTTGTAGCTAAA-1                            unknown  \n",
       "TTTGGTTTCTCATTGT-1                     HsapDv:0000242  \n",
       "TTTGTTGTCCGTATGA-1                            unknown  \n",
       "TTTGTTGTCGCCTTGT-1                     HsapDv:0000242  \n",
       "\n",
       "[48301 rows x 77 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge obs and uns\n",
    "merger = AnnDataMerger(\n",
    "    adata = adata,\n",
    "    obs_df = obs\n",
    ")\n",
    "\n",
    "adata = merger.add_obs_metadata(\n",
    "    adata_col = 'sample_ID',\n",
    "    df_col = 'sample_ID',\n",
    "    skip = None,\n",
    ")\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>author_cell_type</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>author_cell_type_markers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT2</td>\n",
       "      <td>CL:0002063</td>\n",
       "      <td>RTKN2; AGER; NCKAP5; CLIC5; SCEL;  CAV2; COL4A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1</td>\n",
       "      <td>CL:0002062</td>\n",
       "      <td>SFTPC; SFTPA1; NAPSA; SFTPA2; WIF1; ACOXL; AGB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Basal cell</td>\n",
       "      <td>CL:0000646</td>\n",
       "      <td>KRT17; TP63; CACHD1; KRT5; KRT15; AQP3;  FGFR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Multiciliated</td>\n",
       "      <td>CL:4030034</td>\n",
       "      <td>DNAH12;CCDC78; CAPS;  ZBBX; HYDIN; CFAP157;ERI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Goblet</td>\n",
       "      <td>CL:0000160</td>\n",
       "      <td>BPIFB1;MUC5B;CP; PLEKHS1; BMPR1B; ERN2; MUC4; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Club</td>\n",
       "      <td>CL:0000158</td>\n",
       "      <td>MET; STEAP4; BMP6; SCGB3A2; SFTPB;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Alveolar macrophages</td>\n",
       "      <td>CL:0000583</td>\n",
       "      <td>MARCO; INHBA; TREM1; ABHD5; PPARG; FABP4;SLC11A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Macrophage CHIT1+</td>\n",
       "      <td>CL:0000253</td>\n",
       "      <td>SLC1A3; SDC2; CHIT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Macrophage interstitial</td>\n",
       "      <td>CL:4033043</td>\n",
       "      <td>F13A1; STAB1; MRC1; RBPJ; F13A1; CD163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Macrophage MERTK+</td>\n",
       "      <td>CL:0000253</td>\n",
       "      <td>MERTK; SLC9A9; RGL1;HK2; CTSL; SAT1; TNS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DC1</td>\n",
       "      <td>CL:0000990</td>\n",
       "      <td>CLEC9A; CLNK; CPVL; WDFY4; HDAC9;MCOLN2; LRRC18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DC2</td>\n",
       "      <td>CL:0001056</td>\n",
       "      <td>CD86; CLEC7A; MOB3B; PLAUR;CLEC10A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Plasmacytoid DCs</td>\n",
       "      <td>CL:0001058</td>\n",
       "      <td>CLEC4C; PLXNA4; RHEX; NIBAN3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DC activated</td>\n",
       "      <td>CL:0001056</td>\n",
       "      <td>CCR7; BIRC3; RASSF4; TRAF1; SLCO5A1; DAPP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Monocytes</td>\n",
       "      <td>CL:0000860</td>\n",
       "      <td>CD300E; FCN1;  VCAN; ADGRE2; FGR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Monocytes CSF3R+/CD16-</td>\n",
       "      <td>CL:0000875</td>\n",
       "      <td>CSF3R; NAMPT; XPO6; SORL1; IL1R2; ADGRG3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>B cells</td>\n",
       "      <td>CL:0000236</td>\n",
       "      <td>CD19; MS4A1;BANK1; BLK; PAX5; FCRL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Plasma cells</td>\n",
       "      <td>CL:0000786</td>\n",
       "      <td>MZB1; POU2AF1; TENT5C; FCRL5; NUGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CD4 T cells naive</td>\n",
       "      <td>CL:0000895</td>\n",
       "      <td>LEF1; BCL11B; IL7R;  CD69; CD28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CD4 T cells TRM</td>\n",
       "      <td>CL:4033038</td>\n",
       "      <td>ITGA1;ITGAE; CD2; CD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CD8_EM/EMRA</td>\n",
       "      <td>CL:0000913</td>\n",
       "      <td>GZMH;KLRG1;PZP;GNLY;KLRD1; CD8A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CD8_TRM/EM</td>\n",
       "      <td>CL:0000909</td>\n",
       "      <td>GZMK;EOMES;DTHD1;CRTAM; TNIP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CD8 T cells TRM</td>\n",
       "      <td>CL:4033039</td>\n",
       "      <td>CCL5;ITGA1; ITGAE;ZNF683;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Regulatory T cell</td>\n",
       "      <td>CL:0000815</td>\n",
       "      <td>FOXP3;CTLA4;IL2RA;CD28;TIGIT;ICOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NK cells</td>\n",
       "      <td>CL:0000623</td>\n",
       "      <td>GNLY;KLRF1;MYBL1;APBA2; NCR1;FCGR3A; NKG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Mast cells</td>\n",
       "      <td>CL:0000097</td>\n",
       "      <td>TPSAB1; TPSB2; CPA3; HPGDS; MS4A2;SLC18A2; RGS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Lipofibroblasts</td>\n",
       "      <td>CL:0002553</td>\n",
       "      <td>ITGA8;MACF1;PIEZO2; SVEP1; CDH11; NCAM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Adventitial fibroblasts</td>\n",
       "      <td>CL:0002553</td>\n",
       "      <td>COL3A1; CCDC80; ABCA10; FBLN1; FBN1; CFH; PODN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SMCs</td>\n",
       "      <td>CL:0000192</td>\n",
       "      <td>LGR6; ACTA2; MYH11;  LMOD1; MYOCD; IRAG1; TAGLN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Pericytes</td>\n",
       "      <td>CL:0009089</td>\n",
       "      <td>GUCY1A2; PDGFRB; ADARB2; LAMC3; AFF2; ADCY3; TBX5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>EC general capillary</td>\n",
       "      <td>CL:0002144</td>\n",
       "      <td>CA4; EDN1; NOSTRIN; FCN3; SLC6A4; BTNL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>EC aerocyte capillary</td>\n",
       "      <td>CL:0002145</td>\n",
       "      <td>EDNRB;  HPGD; EMCN; PDE1C; CYP3A5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>EC venous pulmonary</td>\n",
       "      <td>CL:4033008</td>\n",
       "      <td>CPE; PKHD1L1; TMTC1; ADGRG6; MYRIP;SELP; ACKR1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>EC venous systemic</td>\n",
       "      <td>CL:0002543</td>\n",
       "      <td>COL15A1; PLVAP; VWA1; KCNE3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>EC arterial</td>\n",
       "      <td>CL:1000413</td>\n",
       "      <td>BMX; GJA5; DKK2;  EFNB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>LECs</td>\n",
       "      <td>CL:0009086</td>\n",
       "      <td>PROX1; MMRN1; PDPN;RELN; CCL21; TBX1; FLT4; SE...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           author_cell_type cell_type_ontology_term_id  \\\n",
       "0                       AT2               CL:0002063     \n",
       "1                       AT1               CL:0002062     \n",
       "2                Basal cell               CL:0000646     \n",
       "3             Multiciliated               CL:4030034     \n",
       "4                    Goblet               CL:0000160     \n",
       "5                      Club               CL:0000158     \n",
       "7      Alveolar macrophages                 CL:0000583   \n",
       "8         Macrophage CHIT1+                 CL:0000253   \n",
       "9   Macrophage interstitial               CL:4033043     \n",
       "10        Macrophage MERTK+                 CL:0000253   \n",
       "11                      DC1                 CL:0000990   \n",
       "12                      DC2                 CL:0001056   \n",
       "13         Plasmacytoid DCs                 CL:0001058   \n",
       "14             DC activated                 CL:0001056   \n",
       "15                Monocytes                 CL:0000860   \n",
       "16   Monocytes CSF3R+/CD16-                 CL:0000875   \n",
       "18                  B cells                 CL:0000236   \n",
       "19             Plasma cells                 CL:0000786   \n",
       "20        CD4 T cells naive                 CL:0000895   \n",
       "21          CD4 T cells TRM                 CL:4033038   \n",
       "22              CD8_EM/EMRA                 CL:0000913   \n",
       "23               CD8_TRM/EM                 CL:0000909   \n",
       "24          CD8 T cells TRM                 CL:4033039   \n",
       "25        Regulatory T cell                 CL:0000815   \n",
       "26                 NK cells                 CL:0000623   \n",
       "28               Mast cells               CL:0000097     \n",
       "30          Lipofibroblasts               CL:0002553     \n",
       "31  Adventitial fibroblasts               CL:0002553     \n",
       "32                     SMCs                CL:0000192    \n",
       "33                Pericytes                 CL:0009089   \n",
       "35     EC general capillary                 CL:0002144   \n",
       "36    EC aerocyte capillary                 CL:0002145   \n",
       "37      EC venous pulmonary                 CL:4033008   \n",
       "38       EC venous systemic               CL:0002543     \n",
       "39              EC arterial                 CL:1000413   \n",
       "40                     LECs                 CL:0009086   \n",
       "\n",
       "                             author_cell_type_markers  \n",
       "0   RTKN2; AGER; NCKAP5; CLIC5; SCEL;  CAV2; COL4A...  \n",
       "1   SFTPC; SFTPA1; NAPSA; SFTPA2; WIF1; ACOXL; AGB...  \n",
       "2      KRT17; TP63; CACHD1; KRT5; KRT15; AQP3;  FGFR3  \n",
       "3   DNAH12;CCDC78; CAPS;  ZBBX; HYDIN; CFAP157;ERI...  \n",
       "4   BPIFB1;MUC5B;CP; PLEKHS1; BMPR1B; ERN2; MUC4; ...  \n",
       "5                  MET; STEAP4; BMP6; SCGB3A2; SFTPB;  \n",
       "7    MARCO; INHBA; TREM1; ABHD5; PPARG; FABP4;SLC11A1  \n",
       "8                                 SLC1A3; SDC2; CHIT1  \n",
       "9              F13A1; STAB1; MRC1; RBPJ; F13A1; CD163  \n",
       "10          MERTK; SLC9A9; RGL1;HK2; CTSL; SAT1; TNS3  \n",
       "11    CLEC9A; CLNK; CPVL; WDFY4; HDAC9;MCOLN2; LRRC18  \n",
       "12                 CD86; CLEC7A; MOB3B; PLAUR;CLEC10A  \n",
       "13                       CLEC4C; PLXNA4; RHEX; NIBAN3  \n",
       "14         CCR7; BIRC3; RASSF4; TRAF1; SLCO5A1; DAPP1  \n",
       "15                   CD300E; FCN1;  VCAN; ADGRE2; FGR  \n",
       "16           CSF3R; NAMPT; XPO6; SORL1; IL1R2; ADGRG3  \n",
       "18                CD19; MS4A1;BANK1; BLK; PAX5; FCRL1  \n",
       "19                MZB1; POU2AF1; TENT5C; FCRL5; NUGGC  \n",
       "20                    LEF1; BCL11B; IL7R;  CD69; CD28  \n",
       "21                              ITGA1;ITGAE; CD2; CD6  \n",
       "22                    GZMH;KLRG1;PZP;GNLY;KLRD1; CD8A  \n",
       "23                      GZMK;EOMES;DTHD1;CRTAM; TNIP3  \n",
       "24                          CCL5;ITGA1; ITGAE;ZNF683;  \n",
       "25                  FOXP3;CTLA4;IL2RA;CD28;TIGIT;ICOS  \n",
       "26          GNLY;KLRF1;MYBL1;APBA2; NCR1;FCGR3A; NKG7  \n",
       "28  TPSAB1; TPSB2; CPA3; HPGDS; MS4A2;SLC18A2; RGS...  \n",
       "30            ITGA8;MACF1;PIEZO2; SVEP1; CDH11; NCAM2  \n",
       "31     COL3A1; CCDC80; ABCA10; FBLN1; FBN1; CFH; PODN  \n",
       "32    LGR6; ACTA2; MYH11;  LMOD1; MYOCD; IRAG1; TAGLN  \n",
       "33  GUCY1A2; PDGFRB; ADARB2; LAMC3; AFF2; ADCY3; TBX5  \n",
       "35            CA4; EDN1; NOSTRIN; FCN3; SLC6A4; BTNL9  \n",
       "36                  EDNRB;  HPGD; EMCN; PDE1C; CYP3A5  \n",
       "37  CPE; PKHD1L1; TMTC1; ADGRG6; MYRIP;SELP; ACKR1...  \n",
       "38                        COL15A1; PLVAP; VWA1; KCNE3  \n",
       "39                            BMX; GJA5; DKK2;  EFNB2  \n",
       "40  PROX1; MMRN1; PDPN;RELN; CCL21; TBX1; FLT4; SE...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marker_genes_df = pd.read_excel(MARKER_GENES_PATH, sheet_name='Sheet1')\n",
    "\n",
    "# remove rows with all NaNs\n",
    "marker_genes_df = marker_genes_df.dropna(how='all')\n",
    "\n",
    "# add to all values in 'ontology' column the prefix 'CL:'\n",
    "marker_genes_df['ontology'] = 'CL:' + marker_genes_df['ontology'].astype(str)\n",
    "\n",
    "marker_genes_df.columns = [AUTHOR_CELL_TYPE, CELL_TYPE_ONTOLOGY_ID, MARKER_GENES]\n",
    "\n",
    "\n",
    "marker_genes_df[AUTHOR_CELL_TYPE] = marker_genes_df[AUTHOR_CELL_TYPE].replace('Macrophae MERTK+', 'Macrophage MERTK+')\n",
    "\n",
    "marker_genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Michael's object with cell type and ontology annotations\n",
    "adata_new = sc.read_h5ad(\"/home/icb/raphael.kfuri-rubens/data/hlca_v2/Schiller_unpubl/COPD_29_04_2024.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>genotype</th>\n",
       "      <th>status</th>\n",
       "      <th>demultiplex_sample</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>...</th>\n",
       "      <th>alignment_software</th>\n",
       "      <th>intron_inclusion</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>Level_3</th>\n",
       "      <th>Level_2</th>\n",
       "      <th>Level_1</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGAAACC-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>singlet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>1617</td>\n",
       "      <td>7.388946</td>\n",
       "      <td>2898.0</td>\n",
       "      <td>7.972121</td>\n",
       "      <td>17.563837</td>\n",
       "      <td>26.293996</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "      <td>Lipofibroblast</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>Fibroblast lineage</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>outside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAATCTGCA-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>singlet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>2051</td>\n",
       "      <td>7.626570</td>\n",
       "      <td>3268.0</td>\n",
       "      <td>8.092239</td>\n",
       "      <td>12.301102</td>\n",
       "      <td>19.033048</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Alveolar macrophages</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Immune</td>\n",
       "      <td>outside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTACCACGC-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>doublet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>2184</td>\n",
       "      <td>7.689371</td>\n",
       "      <td>3527.0</td>\n",
       "      <td>8.168487</td>\n",
       "      <td>11.256025</td>\n",
       "      <td>17.833853</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Lymphatic EC</td>\n",
       "      <td>Lymphatic EC</td>\n",
       "      <td>Lymphatic EC</td>\n",
       "      <td>Endothelial</td>\n",
       "      <td>within</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTGACTGAG-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>singlet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>1456</td>\n",
       "      <td>7.284135</td>\n",
       "      <td>2514.0</td>\n",
       "      <td>7.830028</td>\n",
       "      <td>19.809069</td>\n",
       "      <td>28.758950</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "      <td>EC arterial</td>\n",
       "      <td>EC arterial</td>\n",
       "      <td>Blood vessels</td>\n",
       "      <td>Endothelial</td>\n",
       "      <td>outside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGATCCCGC-1</th>\n",
       "      <td>COPDII</td>\n",
       "      <td>1</td>\n",
       "      <td>singlet</td>\n",
       "      <td>COPDII_1</td>\n",
       "      <td>1469</td>\n",
       "      <td>7.293018</td>\n",
       "      <td>2574.0</td>\n",
       "      <td>7.853605</td>\n",
       "      <td>17.055167</td>\n",
       "      <td>26.107226</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>MONDO:0005002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>AT1</td>\n",
       "      <td>AT1</td>\n",
       "      <td>Alveolar epithelium</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>outside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGAGTCGCTAATG-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>704</td>\n",
       "      <td>6.558198</td>\n",
       "      <td>1130.0</td>\n",
       "      <td>7.030858</td>\n",
       "      <td>23.893805</td>\n",
       "      <td>36.460177</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EC aerocyte capillary</td>\n",
       "      <td>EC capillary</td>\n",
       "      <td>Blood vessels</td>\n",
       "      <td>Endothelial</td>\n",
       "      <td>outside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTGTAGCTAAA-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>728</td>\n",
       "      <td>6.591674</td>\n",
       "      <td>1222.0</td>\n",
       "      <td>7.109062</td>\n",
       "      <td>26.595745</td>\n",
       "      <td>38.952537</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Adventitial fibroblasts</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>Fibroblast lineage</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>outside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCTCATTGT-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>674</td>\n",
       "      <td>6.514713</td>\n",
       "      <td>1089.0</td>\n",
       "      <td>6.993933</td>\n",
       "      <td>28.466483</td>\n",
       "      <td>40.404040</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "      <td>Lipofibroblast</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>Fibroblast lineage</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>outside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCCGTATGA-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>813</td>\n",
       "      <td>6.701960</td>\n",
       "      <td>1661.0</td>\n",
       "      <td>7.415777</td>\n",
       "      <td>31.727875</td>\n",
       "      <td>42.865744</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EC general capillary</td>\n",
       "      <td>EC capillary</td>\n",
       "      <td>Blood vessels</td>\n",
       "      <td>Endothelial</td>\n",
       "      <td>outside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCGCCTTGT-1</th>\n",
       "      <td>Smoker</td>\n",
       "      <td>0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>Smoker_0</td>\n",
       "      <td>809</td>\n",
       "      <td>6.697034</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>7.073270</td>\n",
       "      <td>20.610687</td>\n",
       "      <td>31.382528</td>\n",
       "      <td>...</td>\n",
       "      <td>cell ranger 6.1.2.</td>\n",
       "      <td>yes</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "      <td>Lipofibroblast</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>Fibroblast lineage</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>outside</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42241 rows × 82 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   condition genotype   status demultiplex_sample  \\\n",
       "AAACCCAAGCGAAACC-1    COPDII        1  singlet           COPDII_1   \n",
       "AAACCCACAATCTGCA-1    COPDII        1  singlet           COPDII_1   \n",
       "AAACCCAGTACCACGC-1    COPDII        1  doublet           COPDII_1   \n",
       "AAACCCAGTGACTGAG-1    COPDII        1  singlet           COPDII_1   \n",
       "AAACGAAAGATCCCGC-1    COPDII        1  singlet           COPDII_1   \n",
       "...                      ...      ...      ...                ...   \n",
       "TTTGGAGTCGCTAATG-1    Smoker        0  singlet           Smoker_0   \n",
       "TTTGGTTGTAGCTAAA-1    Smoker        0  singlet           Smoker_0   \n",
       "TTTGGTTTCTCATTGT-1    Smoker        0  singlet           Smoker_0   \n",
       "TTTGTTGTCCGTATGA-1    Smoker        0  singlet           Smoker_0   \n",
       "TTTGTTGTCGCCTTGT-1    Smoker        0  singlet           Smoker_0   \n",
       "\n",
       "                    n_genes_by_counts  log1p_n_genes_by_counts  total_counts  \\\n",
       "AAACCCAAGCGAAACC-1               1617                 7.388946        2898.0   \n",
       "AAACCCACAATCTGCA-1               2051                 7.626570        3268.0   \n",
       "AAACCCAGTACCACGC-1               2184                 7.689371        3527.0   \n",
       "AAACCCAGTGACTGAG-1               1456                 7.284135        2514.0   \n",
       "AAACGAAAGATCCCGC-1               1469                 7.293018        2574.0   \n",
       "...                               ...                      ...           ...   \n",
       "TTTGGAGTCGCTAATG-1                704                 6.558198        1130.0   \n",
       "TTTGGTTGTAGCTAAA-1                728                 6.591674        1222.0   \n",
       "TTTGGTTTCTCATTGT-1                674                 6.514713        1089.0   \n",
       "TTTGTTGTCCGTATGA-1                813                 6.701960        1661.0   \n",
       "TTTGTTGTCGCCTTGT-1                809                 6.697034        1179.0   \n",
       "\n",
       "                    log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "AAACCCAAGCGAAACC-1            7.972121                   17.563837   \n",
       "AAACCCACAATCTGCA-1            8.092239                   12.301102   \n",
       "AAACCCAGTACCACGC-1            8.168487                   11.256025   \n",
       "AAACCCAGTGACTGAG-1            7.830028                   19.809069   \n",
       "AAACGAAAGATCCCGC-1            7.853605                   17.055167   \n",
       "...                                ...                         ...   \n",
       "TTTGGAGTCGCTAATG-1            7.030858                   23.893805   \n",
       "TTTGGTTGTAGCTAAA-1            7.109062                   26.595745   \n",
       "TTTGGTTTCTCATTGT-1            6.993933                   28.466483   \n",
       "TTTGTTGTCCGTATGA-1            7.415777                   31.727875   \n",
       "TTTGTTGTCGCCTTGT-1            7.073270                   20.610687   \n",
       "\n",
       "                    pct_counts_in_top_100_genes  ...  alignment_software  \\\n",
       "AAACCCAAGCGAAACC-1                    26.293996  ...  cell ranger 6.1.2.   \n",
       "AAACCCACAATCTGCA-1                    19.033048  ...  cell ranger 6.1.2.   \n",
       "AAACCCAGTACCACGC-1                    17.833853  ...  cell ranger 6.1.2.   \n",
       "AAACCCAGTGACTGAG-1                    28.758950  ...  cell ranger 6.1.2.   \n",
       "AAACGAAAGATCCCGC-1                    26.107226  ...  cell ranger 6.1.2.   \n",
       "...                                         ...  ...                 ...   \n",
       "TTTGGAGTCGCTAATG-1                    36.460177  ...  cell ranger 6.1.2.   \n",
       "TTTGGTTGTAGCTAAA-1                    38.952537  ...  cell ranger 6.1.2.   \n",
       "TTTGGTTTCTCATTGT-1                    40.404040  ...  cell ranger 6.1.2.   \n",
       "TTTGTTGTCCGTATGA-1                    42.865744  ...  cell ranger 6.1.2.   \n",
       "TTTGTTGTCGCCTTGT-1                    31.382528  ...  cell ranger 6.1.2.   \n",
       "\n",
       "                    intron_inclusion  disease_ontology_term_id  \\\n",
       "AAACCCAAGCGAAACC-1               yes             MONDO:0005002   \n",
       "AAACCCACAATCTGCA-1               yes             MONDO:0005002   \n",
       "AAACCCAGTACCACGC-1               yes             MONDO:0005002   \n",
       "AAACCCAGTGACTGAG-1               yes             MONDO:0005002   \n",
       "AAACGAAAGATCCCGC-1               yes             MONDO:0005002   \n",
       "...                              ...                       ...   \n",
       "TTTGGAGTCGCTAATG-1               yes              PATO:0000461   \n",
       "TTTGGTTGTAGCTAAA-1               yes              PATO:0000461   \n",
       "TTTGGTTTCTCATTGT-1               yes              PATO:0000461   \n",
       "TTTGTTGTCCGTATGA-1               yes              PATO:0000461   \n",
       "TTTGTTGTCGCCTTGT-1               yes              PATO:0000461   \n",
       "\n",
       "                    self_reported_ethnicity_ontology_term_id  \\\n",
       "AAACCCAAGCGAAACC-1                                   unknown   \n",
       "AAACCCACAATCTGCA-1                                   unknown   \n",
       "AAACCCAGTACCACGC-1                                   unknown   \n",
       "AAACCCAGTGACTGAG-1                                   unknown   \n",
       "AAACGAAAGATCCCGC-1                                   unknown   \n",
       "...                                                      ...   \n",
       "TTTGGAGTCGCTAATG-1                                   unknown   \n",
       "TTTGGTTGTAGCTAAA-1                                   unknown   \n",
       "TTTGGTTTCTCATTGT-1                                   unknown   \n",
       "TTTGTTGTCCGTATGA-1                                   unknown   \n",
       "TTTGTTGTCGCCTTGT-1                                   unknown   \n",
       "\n",
       "                    development_stage_ontology_term_id  \\\n",
       "AAACCCAAGCGAAACC-1                      HsapDv:0000239   \n",
       "AAACCCACAATCTGCA-1                             unknown   \n",
       "AAACCCAGTACCACGC-1                             unknown   \n",
       "AAACCCAGTGACTGAG-1                      HsapDv:0000239   \n",
       "AAACGAAAGATCCCGC-1                             unknown   \n",
       "...                                                ...   \n",
       "TTTGGAGTCGCTAATG-1                             unknown   \n",
       "TTTGGTTGTAGCTAAA-1                             unknown   \n",
       "TTTGGTTTCTCATTGT-1                      HsapDv:0000242   \n",
       "TTTGTTGTCCGTATGA-1                             unknown   \n",
       "TTTGTTGTCGCCTTGT-1                      HsapDv:0000242   \n",
       "\n",
       "                                  cell_type       Level_3  \\\n",
       "AAACCCAAGCGAAACC-1           Lipofibroblast   Fibroblasts   \n",
       "AAACCCACAATCTGCA-1     Alveolar macrophages   Macrophages   \n",
       "AAACCCAGTACCACGC-1             Lymphatic EC  Lymphatic EC   \n",
       "AAACCCAGTGACTGAG-1              EC arterial   EC arterial   \n",
       "AAACGAAAGATCCCGC-1                      AT1           AT1   \n",
       "...                                     ...           ...   \n",
       "TTTGGAGTCGCTAATG-1    EC aerocyte capillary  EC capillary   \n",
       "TTTGGTTGTAGCTAAA-1  Adventitial fibroblasts   Fibroblasts   \n",
       "TTTGGTTTCTCATTGT-1           Lipofibroblast   Fibroblasts   \n",
       "TTTGTTGTCCGTATGA-1     EC general capillary  EC capillary   \n",
       "TTTGTTGTCGCCTTGT-1           Lipofibroblast   Fibroblasts   \n",
       "\n",
       "                                Level_2      Level_1    group  \n",
       "AAACCCAAGCGAAACC-1   Fibroblast lineage       Stroma  outside  \n",
       "AAACCCACAATCTGCA-1              Myeloid       Immune  outside  \n",
       "AAACCCAGTACCACGC-1         Lymphatic EC  Endothelial   within  \n",
       "AAACCCAGTGACTGAG-1        Blood vessels  Endothelial  outside  \n",
       "AAACGAAAGATCCCGC-1  Alveolar epithelium   Epithelial  outside  \n",
       "...                                 ...          ...      ...  \n",
       "TTTGGAGTCGCTAATG-1        Blood vessels  Endothelial  outside  \n",
       "TTTGGTTGTAGCTAAA-1   Fibroblast lineage       Stroma  outside  \n",
       "TTTGGTTTCTCATTGT-1   Fibroblast lineage       Stroma  outside  \n",
       "TTTGTTGTCCGTATGA-1        Blood vessels  Endothelial  outside  \n",
       "TTTGTTGTCGCCTTGT-1   Fibroblast lineage       Stroma  outside  \n",
       "\n",
       "[42241 rows x 82 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = adata.obs.merge(adata_new.obs[['cell_type', 'Level_3', 'Level_2', 'Level_1', 'group']], left_index=True, right_index=True)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 42241 × 20669\n",
       "    obs: 'condition', 'genotype', 'status', 'demultiplex_sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mito', 'log1p_total_counts_mito', 'pct_counts_mito', 'sample', 'n_counts', 'n_genes', 'doublet_scores', 'scDblFinder_class', 'soupx_groups', 'scrublet', 'Level_1_transfered_label', 'Level_1_transfer_uncert', 'Level_2_transfered_label', 'Level_2_transfer_uncert', 'Level_3_transfered_label', 'Level_3_transfer_uncert', 'Level_4_transfered_label', 'Level_4_transfer_uncert', 'Level_5_transfered_label', 'Level_5_transfer_uncert', 'condition_coarse', 'leiden_0.2', 'leiden_0.3', 'leiden_0.4', 'leiden_0.5', 'leiden_1.0', 'hlca_manual_lv1', 'hlca_manual_lv2', 'sample_ID', 'donor_id', 'protocol_URL', 'institute', 'sample_collection_site', 'sample_collection_relative_time_point', 'library_ID', 'library_ID_repository', 'author_batch_notes', 'organism_ontology_term_id', 'manner_of_death', 'sample_source', 'sex_ontology_term_id', 'sample_collection_method', 'tissue_type', 'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text', 'sample_preservation_method', 'suspension_type', 'cell_enrichment', 'cell_viability_percentage', 'cell_number_loaded', 'sample_collection_year', 'assay_ontology_term_id', 'library_preparation_batch', 'library_sequencing_run', 'sequenced_fragment', 'sequencing_platform', 'is_primary_data', 'reference_genome', 'gene_annotation_version', 'alignment_software', 'intron_inclusion', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id'\n",
       "    var: 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection', 'ensembl_ID', 'gene_name'\n",
       "    uns: 'Level_1_transfered_label_colors', 'Level_2_transfered_label_colors', 'Level_3_transfered_label_colors', 'Level_4_transfered_label_colors', 'Level_5_transfered_label_colors', 'batch_condition', 'comments', 'condition_colors', 'default_embedding', 'hlca_manual_lv1_colors', 'hlca_manual_lv2_colors', 'hvg', 'leiden', 'leiden_0.2_colors', 'leiden_0.3_colors', 'leiden_0.4_colors', 'leiden_0.5_colors', 'leiden_1.0_colors', 'log1p', 'neighbors', 'pca', 'sample_colors', 'scrublet_colors', 'study_PI', 'title', 'umap', 'unpublished'\n",
       "    obsm: 'X_hlca', 'X_hlca_umap', 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'soupX_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = adata[adata.obs.index.isin(adata_new.obs.index)].copy()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['condition', 'genotype', 'status', 'demultiplex_sample',\n",
       "       'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts',\n",
       "       'log1p_total_counts', 'pct_counts_in_top_50_genes',\n",
       "       'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes',\n",
       "       'pct_counts_in_top_500_genes', 'total_counts_mito',\n",
       "       'log1p_total_counts_mito', 'pct_counts_mito', 'sample', 'n_counts',\n",
       "       'n_genes', 'doublet_scores', 'scDblFinder_class', 'soupx_groups',\n",
       "       'scrublet', 'Level_1_transfered_label', 'Level_1_transfer_uncert',\n",
       "       'Level_2_transfered_label', 'Level_2_transfer_uncert',\n",
       "       'Level_3_transfered_label', 'Level_3_transfer_uncert',\n",
       "       'Level_4_transfered_label', 'Level_4_transfer_uncert',\n",
       "       'Level_5_transfered_label', 'Level_5_transfer_uncert',\n",
       "       'condition_coarse', 'leiden_0.2', 'leiden_0.3', 'leiden_0.4',\n",
       "       'leiden_0.5', 'leiden_1.0', 'hlca_manual_lv1', 'hlca_manual_lv2',\n",
       "       'sample_ID', 'donor_id', 'protocol_URL', 'institute',\n",
       "       'sample_collection_site', 'sample_collection_relative_time_point',\n",
       "       'library_ID', 'library_ID_repository', 'author_batch_notes',\n",
       "       'organism_ontology_term_id', 'manner_of_death', 'sample_source',\n",
       "       'sex_ontology_term_id', 'sample_collection_method', 'tissue_type',\n",
       "       'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text',\n",
       "       'sample_preservation_method', 'suspension_type', 'cell_enrichment',\n",
       "       'cell_viability_percentage', 'cell_number_loaded',\n",
       "       'sample_collection_year', 'assay_ontology_term_id',\n",
       "       'library_preparation_batch', 'library_sequencing_run',\n",
       "       'sequenced_fragment', 'sequencing_platform', 'is_primary_data',\n",
       "       'reference_genome', 'gene_annotation_version', 'alignment_software',\n",
       "       'intron_inclusion', 'disease_ontology_term_id',\n",
       "       'self_reported_ethnicity_ontology_term_id',\n",
       "       'development_stage_ontology_term_id', 'cell_type', 'Level_3', 'Level_2',\n",
       "       'Level_1', 'group'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs = df\n",
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Basal',\n",
       " 'Interstitial macrophage',\n",
       " 'Lipofibroblast',\n",
       " 'Lymphatic EC',\n",
       " 'Regulatory T cells',\n",
       " 'Smooth muscle'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check non-overlap of cell types with AUTHOR_CELL_TYPE in marker_genes_df\n",
    "\n",
    "cell_type_adata = set(adata.obs['cell_type'].unique())\n",
    "cell_type_marker_genes = set(marker_genes_df[AUTHOR_CELL_TYPE].unique())\n",
    "\n",
    "cell_type_adata - cell_type_marker_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Basal cell',\n",
       " 'LECs',\n",
       " 'Lipofibroblasts',\n",
       " 'Macrophage interstitial',\n",
       " 'Regulatory T cell',\n",
       " 'SMCs'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_type_marker_genes - cell_type_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create mapping dict for cell types\n",
    "cell_type_mapping = {\n",
    "    'Basal': 'Basal cell',\n",
    "    'Interstitial macrophage': 'Macrophage interstitial',\n",
    "    'Lipofibroblast': 'Lipofibroblasts',\n",
    "    'Lymphatic EC': 'LECs',\n",
    "    'Regulatory T cells': 'Regulatory T cell',\n",
    "    'Smooth muscle': 'SMCs'\n",
    "}\n",
    "\n",
    "adata.obs[AUTHOR_CELL_TYPE] = adata.obs['cell_type'].replace(cell_type_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# map AUTHOR_CELL_TYPE and CELL_TYPE_ONTOLOGY_ID to cell type annotations\n",
    "adata.obs[CELL_TYPE_ONTOLOGY_ID] = adata.obs[AUTHOR_CELL_TYPE].map(dict(zip(marker_genes_df[AUTHOR_CELL_TYPE], marker_genes_df[CELL_TYPE_ONTOLOGY_ID])))\n",
    "adata.obs[MARKER_GENES] = adata.obs[AUTHOR_CELL_TYPE].map(dict(zip(marker_genes_df[AUTHOR_CELL_TYPE], marker_genes_df[MARKER_GENES])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RTKN2; AGER; NCKAP5; CLIC5; SCEL;  CAV2; COL4A3; LAMA3                     9254\n",
       "ITGA8;MACF1;PIEZO2; SVEP1; CDH11; NCAM2                                    5411\n",
       "SFTPC; SFTPA1; NAPSA; SFTPA2; WIF1; ACOXL; AGBL1; ABCA3; LAMP3             4953\n",
       "CA4; EDN1; NOSTRIN; FCN3; SLC6A4; BTNL9                                    3158\n",
       "MARCO; INHBA; TREM1; ABHD5; PPARG; FABP4;SLC11A1                           2046\n",
       "COL15A1; PLVAP; VWA1; KCNE3                                                1863\n",
       "COL3A1; CCDC80; ABCA10; FBLN1; FBN1; CFH; PODN                             1603\n",
       "EDNRB;  HPGD; EMCN; PDE1C; CYP3A5                                          1481\n",
       "ITGA1;ITGAE; CD2; CD6                                                      1445\n",
       "LGR6; ACTA2; MYH11;  LMOD1; MYOCD; IRAG1; TAGLN                            1253\n",
       "GUCY1A2; PDGFRB; ADARB2; LAMC3; AFF2; ADCY3; TBX5                           869\n",
       "DNAH12;CCDC78; CAPS;  ZBBX; HYDIN; CFAP157;ERICH3;CFAP47; CFAP73; CDHR4     718\n",
       "F13A1; STAB1; MRC1; RBPJ; F13A1; CD163                                      678\n",
       "BMX; GJA5; DKK2;  EFNB2                                                     670\n",
       "PROX1; MMRN1; PDPN;RELN; CCL21; TBX1; FLT4; SEMA3D; FLRT2                   628\n",
       "CD300E; FCN1;  VCAN; ADGRE2; FGR                                            608\n",
       "GZMH;KLRG1;PZP;GNLY;KLRD1; CD8A                                             590\n",
       "GNLY;KLRF1;MYBL1;APBA2; NCR1;FCGR3A; NKG7                                   572\n",
       "MZB1; POU2AF1; TENT5C; FCRL5; NUGGC                                         551\n",
       "LEF1; BCL11B; IL7R;  CD69; CD28                                             455\n",
       "MET; STEAP4; BMP6; SCGB3A2; SFTPB;                                          442\n",
       "CCL5;ITGA1; ITGAE;ZNF683;                                                   368\n",
       "SLC1A3; SDC2; CHIT1                                                         328\n",
       "GZMK;EOMES;DTHD1;CRTAM; TNIP3                                               323\n",
       "CD86; CLEC7A; MOB3B; PLAUR;CLEC10A                                          282\n",
       "MERTK; SLC9A9; RGL1;HK2; CTSL; SAT1; TNS3                                   272\n",
       "CPE; PKHD1L1; TMTC1; ADGRG6; MYRIP;SELP; ACKR1; PLA1A; IL33                 261\n",
       "TPSAB1; TPSB2; CPA3; HPGDS; MS4A2;SLC18A2; RGS13; HDC; CDK15                251\n",
       "CD19; MS4A1;BANK1; BLK; PAX5; FCRL1                                         242\n",
       "CSF3R; NAMPT; XPO6; SORL1; IL1R2; ADGRG3                                    156\n",
       "FOXP3;CTLA4;IL2RA;CD28;TIGIT;ICOS                                           155\n",
       "KRT17; TP63; CACHD1; KRT5; KRT15; AQP3;  FGFR3                              136\n",
       "BPIFB1;MUC5B;CP; PLEKHS1; BMPR1B; ERN2; MUC4; MUC16                          90\n",
       "CLEC9A; CLNK; CPVL; WDFY4; HDAC9;MCOLN2; LRRC18                              56\n",
       "CLEC4C; PLXNA4; RHEX; NIBAN3                                                 47\n",
       "CCR7; BIRC3; RASSF4; TRAF1; SLCO5A1; DAPP1                                   26\n",
       "Name: author_cell_type_markers, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[MARKER_GENES].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 42241 × 20669\n",
       "    obs: 'condition', 'genotype', 'status', 'demultiplex_sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mito', 'log1p_total_counts_mito', 'pct_counts_mito', 'sample', 'n_counts', 'n_genes', 'doublet_scores', 'scDblFinder_class', 'soupx_groups', 'scrublet', 'Level_1_transfered_label', 'Level_1_transfer_uncert', 'Level_2_transfered_label', 'Level_2_transfer_uncert', 'Level_3_transfered_label', 'Level_3_transfer_uncert', 'Level_4_transfered_label', 'Level_4_transfer_uncert', 'Level_5_transfered_label', 'Level_5_transfer_uncert', 'condition_coarse', 'leiden_0.2', 'leiden_0.3', 'leiden_0.4', 'leiden_0.5', 'leiden_1.0', 'hlca_manual_lv1', 'hlca_manual_lv2', 'sample_ID', 'donor_id', 'protocol_URL', 'institute', 'sample_collection_site', 'sample_collection_relative_time_point', 'library_ID', 'library_ID_repository', 'author_batch_notes', 'organism_ontology_term_id', 'manner_of_death', 'sample_source', 'sex_ontology_term_id', 'sample_collection_method', 'tissue_type', 'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text', 'sample_preservation_method', 'suspension_type', 'cell_enrichment', 'cell_viability_percentage', 'cell_number_loaded', 'sample_collection_year', 'assay_ontology_term_id', 'library_preparation_batch', 'library_sequencing_run', 'sequenced_fragment', 'sequencing_platform', 'is_primary_data', 'reference_genome', 'gene_annotation_version', 'alignment_software', 'intron_inclusion', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'cell_type', 'Level_3', 'Level_2', 'Level_1', 'group', 'author_cell_type', 'cell_type_ontology_term_id', 'author_cell_type_markers'\n",
       "    var: 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection', 'ensembl_ID', 'gene_name'\n",
       "    uns: 'Level_1_transfered_label_colors', 'Level_2_transfered_label_colors', 'Level_3_transfered_label_colors', 'Level_4_transfered_label_colors', 'Level_5_transfered_label_colors', 'batch_condition', 'comments', 'condition_colors', 'default_embedding', 'hlca_manual_lv1_colors', 'hlca_manual_lv2_colors', 'hvg', 'leiden', 'leiden_0.2_colors', 'leiden_0.3_colors', 'leiden_0.4_colors', 'leiden_0.5_colors', 'leiden_1.0_colors', 'log1p', 'neighbors', 'pca', 'sample_colors', 'scrublet_colors', 'study_PI', 'title', 'umap', 'unpublished'\n",
       "    obsm: 'X_hlca', 'X_hlca_umap', 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'soupX_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add author cell type markers to UNS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns[MARKER_GENES] = marker_genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check author cell type annotations and Cell Ontology IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['condition', 'genotype', 'status', 'demultiplex_sample',\n",
       "       'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts',\n",
       "       'log1p_total_counts', 'pct_counts_in_top_50_genes',\n",
       "       'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes',\n",
       "       'pct_counts_in_top_500_genes', 'total_counts_mito',\n",
       "       'log1p_total_counts_mito', 'pct_counts_mito', 'sample', 'n_counts',\n",
       "       'n_genes', 'doublet_scores', 'scDblFinder_class', 'soupx_groups',\n",
       "       'scrublet', 'Level_1_transfered_label', 'Level_1_transfer_uncert',\n",
       "       'Level_2_transfered_label', 'Level_2_transfer_uncert',\n",
       "       'Level_3_transfered_label', 'Level_3_transfer_uncert',\n",
       "       'Level_4_transfered_label', 'Level_4_transfer_uncert',\n",
       "       'Level_5_transfered_label', 'Level_5_transfer_uncert',\n",
       "       'condition_coarse', 'leiden_0.2', 'leiden_0.3', 'leiden_0.4',\n",
       "       'leiden_0.5', 'leiden_1.0', 'hlca_manual_lv1', 'hlca_manual_lv2',\n",
       "       'sample_ID', 'donor_id', 'protocol_URL', 'institute',\n",
       "       'sample_collection_site', 'sample_collection_relative_time_point',\n",
       "       'library_ID', 'library_ID_repository', 'author_batch_notes',\n",
       "       'organism_ontology_term_id', 'manner_of_death', 'sample_source',\n",
       "       'sex_ontology_term_id', 'sample_collection_method', 'tissue_type',\n",
       "       'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text',\n",
       "       'sample_preservation_method', 'suspension_type', 'cell_enrichment',\n",
       "       'cell_viability_percentage', 'cell_number_loaded',\n",
       "       'sample_collection_year', 'assay_ontology_term_id',\n",
       "       'library_preparation_batch', 'library_sequencing_run',\n",
       "       'sequenced_fragment', 'sequencing_platform', 'is_primary_data',\n",
       "       'reference_genome', 'gene_annotation_version', 'alignment_software',\n",
       "       'intron_inclusion', 'disease_ontology_term_id',\n",
       "       'self_reported_ethnicity_ontology_term_id',\n",
       "       'development_stage_ontology_term_id', 'cell_type', 'Level_3', 'Level_2',\n",
       "       'Level_1', 'group', 'author_cell_type', 'cell_type_ontology_term_id',\n",
       "       'author_cell_type_markers'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AT2                        9254\n",
       "Lipofibroblasts            5411\n",
       "AT1                        4953\n",
       "EC general capillary       3158\n",
       "Alveolar macrophages       2046\n",
       "EC venous systemic         1863\n",
       "Adventitial fibroblasts    1603\n",
       "EC aerocyte capillary      1481\n",
       "CD4 T cells TRM            1445\n",
       "SMCs                       1253\n",
       "Pericytes                   869\n",
       "Multiciliated               718\n",
       "Macrophage interstitial     678\n",
       "EC arterial                 670\n",
       "LECs                        628\n",
       "Monocytes                   608\n",
       "CD8_EM/EMRA                 590\n",
       "NK cells                    572\n",
       "Plasma cells                551\n",
       "CD4 T cells naive           455\n",
       "Club                        442\n",
       "CD8 T cells TRM             368\n",
       "Macrophage CHIT1+           328\n",
       "CD8_TRM/EM                  323\n",
       "DC2                         282\n",
       "Macrophage MERTK+           272\n",
       "EC venous pulmonary         261\n",
       "Mast cells                  251\n",
       "B cells                     242\n",
       "Monocytes CSF3R+/CD16-      156\n",
       "Regulatory T cell           155\n",
       "Basal cell                  136\n",
       "Goblet                       90\n",
       "DC1                          56\n",
       "Plasmacytoid DCs             47\n",
       "DC activated                 26\n",
       "Name: author_cell_type, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[AUTHOR_CELL_TYPE].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Epithelial     15593\n",
       "Immune          9451\n",
       "Stroma          9136\n",
       "Endothelial     8061\n",
       "Name: Level_1, dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['Level_1'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Alveolar epithelium    14207\n",
       "Fibroblast lineage      7883\n",
       "Blood vessels           7433\n",
       "Myeloid                 4750\n",
       "Lymphoid                4701\n",
       "Airway epithelium       1386\n",
       "Smooth muscle           1253\n",
       "Lymphatic EC             628\n",
       "Name: Level_2, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['Level_2'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AT2                        9254\n",
       "Fibroblasts                7883\n",
       "AT1                        4953\n",
       "EC capillary               4639\n",
       "T cell lineage             3336\n",
       "Macrophages                3324\n",
       "EC venous                  2124\n",
       "Smooth muscle              1253\n",
       "B cell lineage              793\n",
       "Monocytes                   764\n",
       "Multiciliated               718\n",
       "EC arterial                 670\n",
       "Lymphatic EC                628\n",
       "Innate lymphoid cell NK     572\n",
       "Secretory                   532\n",
       "Dendritic cells             411\n",
       "Mast cells                  251\n",
       "Basal                       136\n",
       "Name: Level_3, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['Level_3'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[AUTHOR_CELL_TYPE_L0] = adata.obs['Level_1']\n",
    "adata.obs[AUTHOR_CELL_TYPE_L1] = adata.obs['Level_2']\n",
    "adata.obs[AUTHOR_CELL_TYPE_L2] = adata.obs['Level_3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CL:0002063      9254\n",
       "CL:0002553      7014\n",
       "CL:0002062      4953\n",
       "CL:0002144      3158\n",
       "CL:0000583      2046\n",
       "CL:0002543      1863\n",
       "CL:0002145      1481\n",
       "CL:4033038      1445\n",
       "CL:0000192      1253\n",
       "CL:0009089       869\n",
       "CL:4030034       718\n",
       "CL:4033043       678\n",
       "CL:1000413       670\n",
       "CL:0009086       628\n",
       "CL:0000860       608\n",
       "CL:0000253       600\n",
       "CL:0000913       590\n",
       "CL:0000623       572\n",
       "CL:0000786       551\n",
       "CL:0000895       455\n",
       "CL:0000158       442\n",
       "CL:4033039       368\n",
       "CL:0000909       323\n",
       "CL:0001056       308\n",
       "CL:4033008       261\n",
       "CL:0000097       251\n",
       "CL:0000236       242\n",
       "CL:0000875       156\n",
       "CL:0000815       155\n",
       "CL:0000646       136\n",
       "CL:0000160        90\n",
       "CL:0000990        56\n",
       "CL:0001058        47\n",
       "Name: cell_type_ontology_term_id, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[CELL_TYPE_ONTOLOGY_ID].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check whether ENSEMBL IDs in var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>highly_variable_nbatches</th>\n",
       "      <th>highly_variable_intersection</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ATAD3B</th>\n",
       "      <td>False</td>\n",
       "      <td>0.036800</td>\n",
       "      <td>0.242076</td>\n",
       "      <td>-0.357507</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>ATAD3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRDM16</th>\n",
       "      <td>False</td>\n",
       "      <td>0.168010</td>\n",
       "      <td>0.664629</td>\n",
       "      <td>0.025617</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>PRDM16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PEX10</th>\n",
       "      <td>False</td>\n",
       "      <td>0.010376</td>\n",
       "      <td>0.226664</td>\n",
       "      <td>-0.387448</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>PEX10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SKI</th>\n",
       "      <td>False</td>\n",
       "      <td>0.261257</td>\n",
       "      <td>0.610046</td>\n",
       "      <td>-0.378022</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "      <td>SKI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PEX14</th>\n",
       "      <td>False</td>\n",
       "      <td>0.200704</td>\n",
       "      <td>0.592536</td>\n",
       "      <td>-0.165523</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000142655</td>\n",
       "      <td>PEX14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <td>False</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.632660</td>\n",
       "      <td>0.991525</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>ENSG00000275063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <td>False</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>-0.458416</td>\n",
       "      <td>-2.481308</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>ENSG00000277856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>False</td>\n",
       "      <td>0.009453</td>\n",
       "      <td>0.267841</td>\n",
       "      <td>-0.261455</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268674</th>\n",
       "      <td>False</td>\n",
       "      <td>0.000401</td>\n",
       "      <td>0.486345</td>\n",
       "      <td>0.468138</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>ENSG00000268674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277475</th>\n",
       "      <td>False</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.235231</td>\n",
       "      <td>-0.307846</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>ENSG00000277475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20669 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 highly_variable     means  dispersions  dispersions_norm  \\\n",
       "index                                                                       \n",
       "ATAD3B                     False  0.036800     0.242076         -0.357507   \n",
       "PRDM16                     False  0.168010     0.664629          0.025617   \n",
       "PEX10                      False  0.010376     0.226664         -0.387448   \n",
       "SKI                        False  0.261257     0.610046         -0.378022   \n",
       "PEX14                      False  0.200704     0.592536         -0.165523   \n",
       "...                          ...       ...          ...               ...   \n",
       "ENSG00000275063            False  0.000176     0.632660          0.991525   \n",
       "ENSG00000277856            False  0.000064    -0.458416         -2.481308   \n",
       "ENSG00000271254            False  0.009453     0.267841         -0.261455   \n",
       "ENSG00000268674            False  0.000401     0.486345          0.468138   \n",
       "ENSG00000277475            False  0.000123     0.235231         -0.307846   \n",
       "\n",
       "                 highly_variable_nbatches  highly_variable_intersection  \\\n",
       "index                                                                     \n",
       "ATAD3B                                  0                         False   \n",
       "PRDM16                                  0                         False   \n",
       "PEX10                                   0                         False   \n",
       "SKI                                     0                         False   \n",
       "PEX14                                   0                         False   \n",
       "...                                   ...                           ...   \n",
       "ENSG00000275063                         0                         False   \n",
       "ENSG00000277856                         0                         False   \n",
       "ENSG00000271254                         0                         False   \n",
       "ENSG00000268674                         0                         False   \n",
       "ENSG00000277475                         0                         False   \n",
       "\n",
       "                      ensembl_id      gene_symbol  \n",
       "index                                              \n",
       "ATAD3B           ENSG00000160072           ATAD3B  \n",
       "PRDM16           ENSG00000142611           PRDM16  \n",
       "PEX10            ENSG00000157911            PEX10  \n",
       "SKI              ENSG00000157933              SKI  \n",
       "PEX14            ENSG00000142655            PEX14  \n",
       "...                          ...              ...  \n",
       "ENSG00000275063  ENSG00000275063  ENSG00000275063  \n",
       "ENSG00000277856  ENSG00000277856  ENSG00000277856  \n",
       "ENSG00000271254  ENSG00000271254  ENSG00000271254  \n",
       "ENSG00000268674  ENSG00000268674  ENSG00000268674  \n",
       "ENSG00000277475  ENSG00000277475  ENSG00000277475  \n",
       "\n",
       "[20669 rows x 8 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>highly_variable_nbatches</th>\n",
       "      <th>highly_variable_intersection</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ATAD3B</th>\n",
       "      <td>False</td>\n",
       "      <td>0.036800</td>\n",
       "      <td>0.242076</td>\n",
       "      <td>-0.357507</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>ATAD3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRDM16</th>\n",
       "      <td>False</td>\n",
       "      <td>0.168010</td>\n",
       "      <td>0.664629</td>\n",
       "      <td>0.025617</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>PRDM16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PEX10</th>\n",
       "      <td>False</td>\n",
       "      <td>0.010376</td>\n",
       "      <td>0.226664</td>\n",
       "      <td>-0.387448</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>PEX10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SKI</th>\n",
       "      <td>False</td>\n",
       "      <td>0.261257</td>\n",
       "      <td>0.610046</td>\n",
       "      <td>-0.378022</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "      <td>SKI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PEX14</th>\n",
       "      <td>False</td>\n",
       "      <td>0.200704</td>\n",
       "      <td>0.592536</td>\n",
       "      <td>-0.165523</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000142655</td>\n",
       "      <td>PEX14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <td>False</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.632660</td>\n",
       "      <td>0.991525</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>ENSG00000275063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <td>False</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>-0.458416</td>\n",
       "      <td>-2.481308</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>ENSG00000277856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>False</td>\n",
       "      <td>0.009453</td>\n",
       "      <td>0.267841</td>\n",
       "      <td>-0.261455</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268674</th>\n",
       "      <td>False</td>\n",
       "      <td>0.000401</td>\n",
       "      <td>0.486345</td>\n",
       "      <td>0.468138</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>ENSG00000268674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277475</th>\n",
       "      <td>False</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.235231</td>\n",
       "      <td>-0.307846</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>ENSG00000277475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20669 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 highly_variable     means  dispersions  dispersions_norm  \\\n",
       "index                                                                       \n",
       "ATAD3B                     False  0.036800     0.242076         -0.357507   \n",
       "PRDM16                     False  0.168010     0.664629          0.025617   \n",
       "PEX10                      False  0.010376     0.226664         -0.387448   \n",
       "SKI                        False  0.261257     0.610046         -0.378022   \n",
       "PEX14                      False  0.200704     0.592536         -0.165523   \n",
       "...                          ...       ...          ...               ...   \n",
       "ENSG00000275063            False  0.000176     0.632660          0.991525   \n",
       "ENSG00000277856            False  0.000064    -0.458416         -2.481308   \n",
       "ENSG00000271254            False  0.009453     0.267841         -0.261455   \n",
       "ENSG00000268674            False  0.000401     0.486345          0.468138   \n",
       "ENSG00000277475            False  0.000123     0.235231         -0.307846   \n",
       "\n",
       "                 highly_variable_nbatches  highly_variable_intersection  \\\n",
       "index                                                                     \n",
       "ATAD3B                                  0                         False   \n",
       "PRDM16                                  0                         False   \n",
       "PEX10                                   0                         False   \n",
       "SKI                                     0                         False   \n",
       "PEX14                                   0                         False   \n",
       "...                                   ...                           ...   \n",
       "ENSG00000275063                         0                         False   \n",
       "ENSG00000277856                         0                         False   \n",
       "ENSG00000271254                         0                         False   \n",
       "ENSG00000268674                         0                         False   \n",
       "ENSG00000277475                         0                         False   \n",
       "\n",
       "                      ensembl_id      gene_symbol  \n",
       "index                                              \n",
       "ATAD3B           ENSG00000160072           ATAD3B  \n",
       "PRDM16           ENSG00000142611           PRDM16  \n",
       "PEX10            ENSG00000157911            PEX10  \n",
       "SKI              ENSG00000157933              SKI  \n",
       "PEX14            ENSG00000142655            PEX14  \n",
       "...                          ...              ...  \n",
       "ENSG00000275063  ENSG00000275063  ENSG00000275063  \n",
       "ENSG00000277856  ENSG00000277856  ENSG00000277856  \n",
       "ENSG00000271254  ENSG00000271254  ENSG00000271254  \n",
       "ENSG00000268674  ENSG00000268674  ENSG00000268674  \n",
       "ENSG00000277475  ENSG00000277475  ENSG00000277475  \n",
       "\n",
       "[20669 rows x 8 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.index.name = 'index'\n",
    "adata.var.rename(columns={'gene_name': 'gene_symbol', 'ensembl_ID': 'ensembl_id'}, inplace=True)\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['gene_symbol'] = adata.var['gene_symbol'].astype(str)\n",
    "adata.var['ensembl_id'] = adata.var['ensembl_id'].astype(str)\n",
    "adata.var.index = adata.var.index.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "highly_variable                    bool\n",
       "means                           float64\n",
       "dispersions                     float64\n",
       "dispersions_norm                float32\n",
       "highly_variable_nbatches          int64\n",
       "highly_variable_intersection       bool\n",
       "ensembl_id                       object\n",
       "gene_symbol                      object\n",
       "dtype: object"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3026.0"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.layers['soupX_counts'].toarray().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers['soupX_counts']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.X.astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "highly_variable                    bool\n",
       "means                           float64\n",
       "dispersions                     float64\n",
       "dispersions_norm                float32\n",
       "highly_variable_nbatches          int64\n",
       "highly_variable_intersection       bool\n",
       "ensembl_id                       object\n",
       "gene_symbol                      object\n",
       "dtype: object"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw = adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "highly_variable                    bool\n",
       "means                           float64\n",
       "dispersions                     float64\n",
       "dispersions_norm                float32\n",
       "highly_variable_nbatches          int64\n",
       "highly_variable_intersection       bool\n",
       "ensembl_id                       object\n",
       "gene_symbol                      object\n",
       "dtype: object"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.var.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<42241x20669 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 69689379 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<42241x20669 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 69689379 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3026"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.toarray().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3026"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.X.toarray().max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validation result\n",
    "\n",
    "### UNS Validation\n",
    "- OK: Tier 1 UNS Google Sheet\n",
    "- OK: Tier 1 UNS AnnData Object  same as Google Sheet\n",
    "\n",
    "### OBS Validation\n",
    "- OK: Tier 1 OBS Google Sheet\n",
    "- OK: Tier 1 OBS AnnData Object same as Google Sheet\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Submission Status\n",
    "\n",
    "- CHECK: Raw counts in X and in raw\n",
    "- CHECK: Tier 1 Metadata in OBS\n",
    "- CHECK: Cell Ontology IDs in OBS\n",
    "- CHECK: Author cell type in OBS\n",
    "- CHECK: Marker genes in UNS\n",
    "- CHECK: ENSEMBL IDs and gene symbols in var\n",
    "\n",
    "# Revisions:\n",
    "\n",
    "DONE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(join(OUTPUT_PATH_PREREVISION, f\"{DATASET_ID}.h5ad\"), compression='gzip')\n",
    "adata.write_h5ad(join(OUTPUT_PATH_POSTREVISION, f\"{DATASET_ID}.h5ad\"), compression='gzip')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hlcav2-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
