{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "package ‘methods’ was built under R version 4.3.3 \n",
      "During startup - Warning messages:\n",
      "1: package ‘datasets’ was built under R version 4.3.3 \n",
      "2: package ‘utils’ was built under R version 4.3.3 \n",
      "3: package ‘grDevices’ was built under R version 4.3.3 \n",
      "4: package ‘graphics’ was built under R version 4.3.3 \n",
      "5: package ‘stats’ was built under R version 4.3.3 \n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import anndata2ri\n",
    "import os\n",
    "from os.path import join\n",
    "import scipy.sparse as sp\n",
    "import numpy as np\n",
    "\n",
    "from hlca_v2.ingestion_utils import get_gspread_df, ValidationWorkflow, AnnDataMerger\n",
    "\n",
    "GSPREAD_JSON = \"/home/icb/raphael.kfuri-rubens/data/hlca_v2/google_sheets_api/hlca-v2-8d5fea10d8f3.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_ID = \"Niklason_Kaminski_publ\"\n",
    "RDS_PATH = f\"/home/icb/raphael.kfuri-rubens/data/hlca_v2/{DATASET_ID}/{DATASET_ID}.rds\"\n",
    "OUTPUT_PATH_PREREVISION = '/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_prerevision'\n",
    "OUTPUT_PATH_POSTREVISION = '/home/icb/raphael.kfuri-rubens/data/hlca_v2/HLCA_V2_CORE/adata_postrevision'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Name constants\n",
    "AUTHOR_CELL_TYPE_L0 = 'author_cell_type_level_0'\n",
    "AUTHOR_CELL_TYPE_L1 = 'author_cell_type_level_1'\n",
    "\n",
    "CELL_TYPE_ONTOLOGY_ID_L0 = 'cell_type_ontology_term_id_level_0'\n",
    "CELL_TYPE_ONTOLOGY_ID_L1 = 'cell_type_ontology_term_id_level_1'\n",
    "\n",
    "CELL_TYPE_ONTOLOGY_LABEL_L0 = 'cell_type_ontology_term_label_level_0'\n",
    "CELL_TYPE_ONTOLOGY_LABEL_L1 = 'cell_type_ontology_term_label_level_1'\n",
    "\n",
    "AUTHOR_CELL_TYPE_DESCRIPTION_L0 = 'author_cell_type_description_level_0'\n",
    "AUTHOR_CELL_TYPE_DESCRIPTION_L1 = 'author_cell_type_description_level_1'\n",
    "\n",
    "MARKER_GENES_L0 = 'author_cell_type_markers_level_0'\n",
    "MARKER_GENES_L1 = 'author_cell_type_markers_level_1'\n",
    "\n",
    "# Finest grained annotation will be generic dataset cell type\n",
    "AUTHOR_CELL_TYPE = 'author_cell_type'\n",
    "CELL_TYPE_ONTOLOGY_ID = 'cell_type_ontology_term_id'\n",
    "CELL_TYPE_ONTOLOGY_LABEL = 'cell_type_ontology_term_label'\n",
    "MARKER_GENES = 'author_cell_type_markers'\n",
    "AUTHOR_CELL_TYPE_DESCRIPTION = 'author_cell_type_description'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_4183780/2547887543.py:1: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  anndata2ri.activate()\n"
     ]
    }
   ],
   "source": [
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: package ‘tools’ was built under R version 4.3.3 \n",
       "2: In (function (package, help, pos = 2, lib.loc = NULL, character.only = FALSE,  :\n",
       "  library ‘/ictstr01/home/icb/raphael.kfuri-rubens/R/x86_64-redhat-linux-gnu-library/4.3’ contains no packages\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i RDS_PATH\n",
    "\n",
    "suppressPackageStartupMessages(library(Seurat))\n",
    "rds <- readRDS(file=RDS_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 46204 17867 \n",
      "metadata(0):\n",
      "assays(3): counts logcounts scaledata\n",
      "rownames(46204): DPM1 SCYL3 ... ENSG00000277705 ENSG00000280139\n",
      "rowData names(0):\n",
      "colnames(17867): NDRI_1b__AAACCTGAGGATATAC NDRI_1b__AAAGATGGTCTGGAGA\n",
      "  ... 388C_GTGCGGTCACCTATCC 92C_GAATGAATCTCTTATG\n",
      "colData names(22): nGene nUMI ... Cell_Ontology_Class_ID ident\n",
      "reducedDimNames(3): CCA CCA.ALIGNED TSNE\n",
      "mainExpName: RNA\n",
      "altExpNames(0):\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/icb/raphael.kfuri-rubens/miniconda3/envs/hlcav2-env/lib/python3.10/site-packages/anndata2ri/_r2py.py:132: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  return AnnData(exprs, obs, var, uns, obsm, layers=layers)\n"
     ]
    }
   ],
   "source": [
    "%%R -o adata\n",
    "\n",
    "adata <- as.SingleCellExperiment(rds)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nGene</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>percent.mito</th>\n",
       "      <th>exp</th>\n",
       "      <th>species</th>\n",
       "      <th>class</th>\n",
       "      <th>library.ident</th>\n",
       "      <th>disease.ident</th>\n",
       "      <th>subject.ident</th>\n",
       "      <th>...</th>\n",
       "      <th>cell.cat.ident</th>\n",
       "      <th>name_initial</th>\n",
       "      <th>var.ratio.pca</th>\n",
       "      <th>hum.stash</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>stash</th>\n",
       "      <th>Annotation_for_Publication</th>\n",
       "      <th>Cell_Ontology_Class_ID</th>\n",
       "      <th>ident</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAACCTGAGGATATAC</th>\n",
       "      <td>2607</td>\n",
       "      <td>8348.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.026007</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.950244</td>\n",
       "      <td>Mo</td>\n",
       "      <td>8344.0</td>\n",
       "      <td>2603</td>\n",
       "      <td>Mo</td>\n",
       "      <td>Mo</td>\n",
       "      <td>CL:0000576</td>\n",
       "      <td>CL:0000576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAAGATGGTCTGGAGA</th>\n",
       "      <td>1496</td>\n",
       "      <td>3905.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.050192</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.057564</td>\n",
       "      <td>Mo</td>\n",
       "      <td>3905.0</td>\n",
       "      <td>1496</td>\n",
       "      <td>Mo</td>\n",
       "      <td>Mo</td>\n",
       "      <td>CL:0000576</td>\n",
       "      <td>CL:0000576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAAGTAGGTGTGGTTT</th>\n",
       "      <td>2776</td>\n",
       "      <td>9400.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.044371</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.067565</td>\n",
       "      <td>Mo</td>\n",
       "      <td>9398.0</td>\n",
       "      <td>2774</td>\n",
       "      <td>Mo</td>\n",
       "      <td>Mo</td>\n",
       "      <td>CL:0000576</td>\n",
       "      <td>CL:0000576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAGGAGCAGTCTTGCA</th>\n",
       "      <td>1261</td>\n",
       "      <td>3436.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.036671</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.033633</td>\n",
       "      <td>Mo</td>\n",
       "      <td>3436.0</td>\n",
       "      <td>1261</td>\n",
       "      <td>Mo</td>\n",
       "      <td>Mo</td>\n",
       "      <td>CL:0000576</td>\n",
       "      <td>CL:0000576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAGTCTGCAATGAAAC</th>\n",
       "      <td>2507</td>\n",
       "      <td>8426.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.026110</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.989638</td>\n",
       "      <td>Mo</td>\n",
       "      <td>8426.0</td>\n",
       "      <td>2507</td>\n",
       "      <td>Mo</td>\n",
       "      <td>Mo</td>\n",
       "      <td>CL:0000576</td>\n",
       "      <td>CL:0000576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222C_GGCTGGTAGCTGATAA</th>\n",
       "      <td>1211</td>\n",
       "      <td>2697.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.041898</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>222C</td>\n",
       "      <td>Control</td>\n",
       "      <td>222C</td>\n",
       "      <td>...</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Mast</td>\n",
       "      <td>0.985601</td>\n",
       "      <td>Mast</td>\n",
       "      <td>2697.0</td>\n",
       "      <td>1211</td>\n",
       "      <td>Mast</td>\n",
       "      <td>Mast</td>\n",
       "      <td>CL:0000097</td>\n",
       "      <td>CL:0000097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388C_GCTGGGTCAGACACTT</th>\n",
       "      <td>1217</td>\n",
       "      <td>2522.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.055908</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>388C</td>\n",
       "      <td>Control</td>\n",
       "      <td>388C</td>\n",
       "      <td>...</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Mast</td>\n",
       "      <td>0.985205</td>\n",
       "      <td>Mast</td>\n",
       "      <td>2522.0</td>\n",
       "      <td>1217</td>\n",
       "      <td>Mast</td>\n",
       "      <td>Mast</td>\n",
       "      <td>CL:0000097</td>\n",
       "      <td>CL:0000097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388C_GGACGTCAGCAGGTCA</th>\n",
       "      <td>922</td>\n",
       "      <td>1851.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.060508</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>388C</td>\n",
       "      <td>Control</td>\n",
       "      <td>388C</td>\n",
       "      <td>...</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Mast</td>\n",
       "      <td>0.947162</td>\n",
       "      <td>Mast</td>\n",
       "      <td>1851.0</td>\n",
       "      <td>922</td>\n",
       "      <td>Mast</td>\n",
       "      <td>Mast</td>\n",
       "      <td>CL:0000097</td>\n",
       "      <td>CL:0000097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388C_GTGCGGTCACCTATCC</th>\n",
       "      <td>1074</td>\n",
       "      <td>2025.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.070123</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>388C</td>\n",
       "      <td>Control</td>\n",
       "      <td>388C</td>\n",
       "      <td>...</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Mast</td>\n",
       "      <td>0.986631</td>\n",
       "      <td>Mast</td>\n",
       "      <td>2025.0</td>\n",
       "      <td>1074</td>\n",
       "      <td>Mast</td>\n",
       "      <td>Mast</td>\n",
       "      <td>CL:0000097</td>\n",
       "      <td>CL:0000097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92C_GAATGAATCTCTTATG</th>\n",
       "      <td>2493</td>\n",
       "      <td>8247.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.029950</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>92C</td>\n",
       "      <td>Control</td>\n",
       "      <td>192C</td>\n",
       "      <td>...</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Mast</td>\n",
       "      <td>0.954736</td>\n",
       "      <td>Mast</td>\n",
       "      <td>8247.0</td>\n",
       "      <td>2493</td>\n",
       "      <td>Mast</td>\n",
       "      <td>Mast</td>\n",
       "      <td>CL:0000097</td>\n",
       "      <td>CL:0000097</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17867 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           nGene    nUMI orig.ident  percent.mito   exp  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC   2607  8348.0    NDRI_1b      0.026007  Hum2   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA   1496  3905.0    NDRI_1b      0.050192  Hum2   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT   2776  9400.0    NDRI_1b      0.044371  Hum2   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA   1261  3436.0    NDRI_1b      0.036671  Hum2   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC   2507  8426.0    NDRI_1b      0.026110  Hum2   \n",
       "...                          ...     ...        ...           ...   ...   \n",
       "222C_GGCTGGTAGCTGATAA       1211  2697.0   CellSoup      0.041898  None   \n",
       "388C_GCTGGGTCAGACACTT       1217  2522.0   CellSoup      0.055908  None   \n",
       "388C_GGACGTCAGCAGGTCA        922  1851.0   CellSoup      0.060508  None   \n",
       "388C_GTGCGGTCACCTATCC       1074  2025.0   CellSoup      0.070123  None   \n",
       "92C_GAATGAATCTCTTATG        2493  8247.0   CellSoup      0.029950  None   \n",
       "\n",
       "                          species           class library.ident disease.ident  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC   Human    Original CCA          None          None   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA   Human    Original CCA          None          None   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT   Human    Original CCA          None          None   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA   Human    Original CCA          None          None   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC   Human    Original CCA          None          None   \n",
       "...                           ...             ...           ...           ...   \n",
       "222C_GGCTGGTAGCTGATAA        None  Human Controls          222C       Control   \n",
       "388C_GCTGGGTCAGACACTT        None  Human Controls          388C       Control   \n",
       "388C_GGACGTCAGCAGGTCA        None  Human Controls          388C       Control   \n",
       "388C_GTGCGGTCACCTATCC        None  Human Controls          388C       Control   \n",
       "92C_GAATGAATCTCTTATG         None  Human Controls           92C       Control   \n",
       "\n",
       "                          subject.ident  ... cell.cat.ident name_initial  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC          Hum2  ...           None         None   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA          Hum2  ...           None         None   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT          Hum2  ...           None         None   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA          Hum2  ...           None         None   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC          Hum2  ...           None         None   \n",
       "...                                 ...  ...            ...          ...   \n",
       "222C_GGCTGGTAGCTGATAA              222C  ...        Myeloid         Mast   \n",
       "388C_GCTGGGTCAGACACTT              388C  ...        Myeloid         Mast   \n",
       "388C_GGACGTCAGCAGGTCA              388C  ...        Myeloid         Mast   \n",
       "388C_GTGCGGTCACCTATCC              388C  ...        Myeloid         Mast   \n",
       "92C_GAATGAATCTCTTATG               192C  ...        Myeloid         Mast   \n",
       "\n",
       "                          var.ratio.pca hum.stash  nCount_RNA nFeature_RNA  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC      0.950244        Mo      8344.0         2603   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA      1.057564        Mo      3905.0         1496   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT      1.067565        Mo      9398.0         2774   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA      1.033633        Mo      3436.0         1261   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC      0.989638        Mo      8426.0         2507   \n",
       "...                                 ...       ...         ...          ...   \n",
       "222C_GGCTGGTAGCTGATAA          0.985601      Mast      2697.0         1211   \n",
       "388C_GCTGGGTCAGACACTT          0.985205      Mast      2522.0         1217   \n",
       "388C_GGACGTCAGCAGGTCA          0.947162      Mast      1851.0          922   \n",
       "388C_GTGCGGTCACCTATCC          0.986631      Mast      2025.0         1074   \n",
       "92C_GAATGAATCTCTTATG           0.954736      Mast      8247.0         2493   \n",
       "\n",
       "                           stash  Annotation_for_Publication  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC     Mo                          Mo   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA     Mo                          Mo   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT     Mo                          Mo   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA     Mo                          Mo   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC     Mo                          Mo   \n",
       "...                          ...                         ...   \n",
       "222C_GGCTGGTAGCTGATAA       Mast                        Mast   \n",
       "388C_GCTGGGTCAGACACTT       Mast                        Mast   \n",
       "388C_GGACGTCAGCAGGTCA       Mast                        Mast   \n",
       "388C_GTGCGGTCACCTATCC       Mast                        Mast   \n",
       "92C_GAATGAATCTCTTATG        Mast                        Mast   \n",
       "\n",
       "                          Cell_Ontology_Class_ID       ident  \n",
       "NDRI_1b__AAACCTGAGGATATAC             CL:0000576  CL:0000576  \n",
       "NDRI_1b__AAAGATGGTCTGGAGA             CL:0000576  CL:0000576  \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT             CL:0000576  CL:0000576  \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA             CL:0000576  CL:0000576  \n",
       "NDRI_1b__AAGTCTGCAATGAAAC             CL:0000576  CL:0000576  \n",
       "...                                          ...         ...  \n",
       "222C_GGCTGGTAGCTGATAA                 CL:0000097  CL:0000097  \n",
       "388C_GCTGGGTCAGACACTT                 CL:0000097  CL:0000097  \n",
       "388C_GGACGTCAGCAGGTCA                 CL:0000097  CL:0000097  \n",
       "388C_GTGCGGTCACCTATCC                 CL:0000097  CL:0000097  \n",
       "92C_GAATGAATCTCTTATG                  CL:0000097  CL:0000097  \n",
       "\n",
       "[17867 rows x 22 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs = get_gspread_df(GSPREAD_JSON, DATASET_ID, \"tier_1\", \"obs\")\n",
    "uns = get_gspread_df(GSPREAD_JSON, DATASET_ID, \"tier_1\", \"uns\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation workflow successful. Returning UNS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation workflow successful. Returning UNS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>study_PI</th>\n",
       "      <th>batch_condition</th>\n",
       "      <th>default_embedding</th>\n",
       "      <th>unpublished</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>All - Single Cell Connectomic Analysis of Adul...</td>\n",
       "      <td>Laura,E,Niklason, Naftali,Kaminski</td>\n",
       "      <td>subject.ident,library.ident,exp</td>\n",
       "      <td>hum_GEN2@reductions$tsne</td>\n",
       "      <td>consented for release</td>\n",
       "      <td>HC1 and NDRI_1b were dissociated by Dr. Rared...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               title  \\\n",
       "0  All - Single Cell Connectomic Analysis of Adul...   \n",
       "\n",
       "                             study_PI                  batch_condition  \\\n",
       "0  Laura,E,Niklason, Naftali,Kaminski  subject.ident,library.ident,exp   \n",
       "\n",
       "          default_embedding            unpublished  \\\n",
       "0  hum_GEN2@reductions$tsne  consented for release   \n",
       "\n",
       "                                            comments  \n",
       "0   HC1 and NDRI_1b were dissociated by Dr. Rared...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = uns,\n",
    "    axis = 'uns'\n",
    ")\n",
    "\n",
    "validated_uns = val_workflow.init_workflow()\n",
    "validated_uns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Validation <span style=\"color: #800080; text-decoration-color: #800080; font-weight: bold\">error</span><span style=\"font-weight: bold\">(</span>s<span style=\"font-weight: bold\">)</span> encountered.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Validation \u001b[1;35merror\u001b[0m\u001b[1m(\u001b[0ms\u001b[1m)\u001b[0m encountered.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">{</span>\n",
       "<span style=\"color: #7fbf7f; text-decoration-color: #7fbf7f\">│   </span><span style=\"color: #008000; text-decoration-color: #008000\">'sample_source'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"Must be one of {'surgical donor', 'postmortem donor', 'organ donor'}. Invalid entries found.\"</span>,\n",
       "<span style=\"color: #7fbf7f; text-decoration-color: #7fbf7f\">│   </span><span style=\"color: #008000; text-decoration-color: #008000\">'sample_collection_method'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"Must be one of {'biopsy', 'brush', 'surgical resection'}. Invalid entries found.\"</span>\n",
       "<span style=\"font-weight: bold\">}</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1m{\u001b[0m\n",
       "\u001b[2;32m│   \u001b[0m\u001b[32m'sample_source'\u001b[0m: \u001b[32m\"Must be one of \u001b[0m\u001b[32m{\u001b[0m\u001b[32m'surgical donor', 'postmortem donor', 'organ donor'\u001b[0m\u001b[32m}\u001b[0m\u001b[32m. Invalid entries found.\"\u001b[0m,\n",
       "\u001b[2;32m│   \u001b[0m\u001b[32m'sample_collection_method'\u001b[0m: \u001b[32m\"Must be one of \u001b[0m\u001b[32m{\u001b[0m\u001b[32m'biopsy', 'brush', 'surgical resection'\u001b[0m\u001b[32m}\u001b[0m\u001b[32m. Invalid entries found.\"\u001b[0m\n",
       "\u001b[1m}\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Error containing field: sample_source\n",
       "Error message: Must be one of <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'surgical donor'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'postmortem donor'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'organ donor'</span><span style=\"font-weight: bold\">}</span>. Invalid entries found.\n",
       "Value counts:\n",
       "N/A                 <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>\n",
       "surgical donor       <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1</span>\n",
       "postmortem donor     <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1</span>\n",
       "Name: sample_source, dtype: int64\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Error containing field: sample_source\n",
       "Error message: Must be one of \u001b[1m{\u001b[0m\u001b[32m'surgical donor'\u001b[0m, \u001b[32m'postmortem donor'\u001b[0m, \u001b[32m'organ donor'\u001b[0m\u001b[1m}\u001b[0m. Invalid entries found.\n",
       "Value counts:\n",
       "N/A                 \u001b[1;36m12\u001b[0m\n",
       "surgical donor       \u001b[1;36m1\u001b[0m\n",
       "postmortem donor     \u001b[1;36m1\u001b[0m\n",
       "Name: sample_source, dtype: int64\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Error containing field: sample_collection_method\n",
       "Error message: Must be one of <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'biopsy'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'brush'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'surgical resection'</span><span style=\"font-weight: bold\">}</span>. Invalid entries found.\n",
       "Value counts:\n",
       "N/A                   <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>\n",
       "surgical resection     <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>\n",
       "Name: sample_collection_method, dtype: int64\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Error containing field: sample_collection_method\n",
       "Error message: Must be one of \u001b[1m{\u001b[0m\u001b[32m'biopsy'\u001b[0m, \u001b[32m'brush'\u001b[0m, \u001b[32m'surgical resection'\u001b[0m\u001b[1m}\u001b[0m. Invalid entries found.\n",
       "Value counts:\n",
       "N/A                   \u001b[1;36m12\u001b[0m\n",
       "surgical resection     \u001b[1;36m2\u001b[0m\n",
       "Name: sample_collection_method, dtype: int64\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Returning error-containing OBS dataframe.\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Returning error-containing OBS dataframe.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>protocol_URL</th>\n",
       "      <th>institute</th>\n",
       "      <th>sample_collection_site</th>\n",
       "      <th>sample_collection_relative_time_point</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>library_ID_repository</th>\n",
       "      <th>author_batch_notes</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sequenced_fragment</th>\n",
       "      <th>sequencing_platform</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>reference_genome</th>\n",
       "      <th>gene_annotation_version</th>\n",
       "      <th>alignment_software</th>\n",
       "      <th>intron_inclusion</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hum1</td>\n",
       "      <td>SA_donor_1</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_1_YSM</td>\n",
       "      <td>N/A</td>\n",
       "      <td>HC1</td>\n",
       "      <td>GSM3926545</td>\n",
       "      <td>Dissociated by MSBR and sequenced fresh (separ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hum2</td>\n",
       "      <td>SA_donor_2</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_1_YSM</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>GSM3926546</td>\n",
       "      <td>Dissociated by MSBR and sequenced fresh (separ...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>001C</td>\n",
       "      <td>SA_donor_3</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>001C</td>\n",
       "      <td>GSM4050097</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>048C</td>\n",
       "      <td>SA_donor_4</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>048C</td>\n",
       "      <td>GSM4050099</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>098C</td>\n",
       "      <td>SA_donor_5</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>098C</td>\n",
       "      <td>GSM4050101</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>133C</td>\n",
       "      <td>SA_donor_6</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>133C</td>\n",
       "      <td>GSM4050103</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1372C</td>\n",
       "      <td>SA_donor_7</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>1372C</td>\n",
       "      <td>GSM4050105</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>137C</td>\n",
       "      <td>SA_donor_8</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>137C</td>\n",
       "      <td>GSM4050107</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>160C</td>\n",
       "      <td>SA_donor_9</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>160C</td>\n",
       "      <td>GSM4050109</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>192C</td>\n",
       "      <td>SA_donor_10</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>192C</td>\n",
       "      <td>GSM4050111</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>222C</td>\n",
       "      <td>SA_donor_11</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>222C</td>\n",
       "      <td>GSM4050112</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>244C</td>\n",
       "      <td>SA_donor_12</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>244C</td>\n",
       "      <td>GSM4050113</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>388C</td>\n",
       "      <td>SA_donor_13</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>388C</td>\n",
       "      <td>GSM4050114</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>454C</td>\n",
       "      <td>SA_donor_14</td>\n",
       "      <td>https://www.science.org/doi/10.1126/sciadv.aaw...</td>\n",
       "      <td>Yale School of Medicine</td>\n",
       "      <td>Site_2_BWH</td>\n",
       "      <td>N/A</td>\n",
       "      <td>454C</td>\n",
       "      <td>GSM4050115</td>\n",
       "      <td>Dissociated by members of Rosas lab, frozen in...</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>...</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_ID     donor_id                                       protocol_URL  \\\n",
       "0       Hum1   SA_donor_1  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "1       Hum2   SA_donor_2  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "2       001C   SA_donor_3  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "3       048C   SA_donor_4  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "4       098C   SA_donor_5  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "5       133C   SA_donor_6  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "6      1372C   SA_donor_7  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "7       137C   SA_donor_8  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "8       160C   SA_donor_9  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "9       192C  SA_donor_10  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "10      222C  SA_donor_11  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "11      244C  SA_donor_12  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "12      388C  SA_donor_13  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "13      454C  SA_donor_14  https://www.science.org/doi/10.1126/sciadv.aaw...   \n",
       "\n",
       "                  institute sample_collection_site  \\\n",
       "0   Yale School of Medicine             Site_1_YSM   \n",
       "1   Yale School of Medicine             Site_1_YSM   \n",
       "2   Yale School of Medicine             Site_2_BWH   \n",
       "3   Yale School of Medicine             Site_2_BWH   \n",
       "4   Yale School of Medicine             Site_2_BWH   \n",
       "5   Yale School of Medicine             Site_2_BWH   \n",
       "6   Yale School of Medicine             Site_2_BWH   \n",
       "7   Yale School of Medicine             Site_2_BWH   \n",
       "8   Yale School of Medicine             Site_2_BWH   \n",
       "9   Yale School of Medicine             Site_2_BWH   \n",
       "10  Yale School of Medicine             Site_2_BWH   \n",
       "11  Yale School of Medicine             Site_2_BWH   \n",
       "12  Yale School of Medicine             Site_2_BWH   \n",
       "13  Yale School of Medicine             Site_2_BWH   \n",
       "\n",
       "   sample_collection_relative_time_point library_ID library_ID_repository  \\\n",
       "0                                    N/A        HC1            GSM3926545   \n",
       "1                                    N/A    NDRI_1b            GSM3926546   \n",
       "2                                    N/A       001C            GSM4050097   \n",
       "3                                    N/A       048C            GSM4050099   \n",
       "4                                    N/A       098C            GSM4050101   \n",
       "5                                    N/A       133C            GSM4050103   \n",
       "6                                    N/A      1372C            GSM4050105   \n",
       "7                                    N/A       137C            GSM4050107   \n",
       "8                                    N/A       160C            GSM4050109   \n",
       "9                                    N/A       192C            GSM4050111   \n",
       "10                                   N/A       222C            GSM4050112   \n",
       "11                                   N/A       244C            GSM4050113   \n",
       "12                                   N/A       388C            GSM4050114   \n",
       "13                                   N/A       454C            GSM4050115   \n",
       "\n",
       "                                   author_batch_notes  \\\n",
       "0   Dissociated by MSBR and sequenced fresh (separ...   \n",
       "1   Dissociated by MSBR and sequenced fresh (separ...   \n",
       "2   Dissociated by members of Rosas lab, frozen in...   \n",
       "3   Dissociated by members of Rosas lab, frozen in...   \n",
       "4   Dissociated by members of Rosas lab, frozen in...   \n",
       "5   Dissociated by members of Rosas lab, frozen in...   \n",
       "6   Dissociated by members of Rosas lab, frozen in...   \n",
       "7   Dissociated by members of Rosas lab, frozen in...   \n",
       "8   Dissociated by members of Rosas lab, frozen in...   \n",
       "9   Dissociated by members of Rosas lab, frozen in...   \n",
       "10  Dissociated by members of Rosas lab, frozen in...   \n",
       "11  Dissociated by members of Rosas lab, frozen in...   \n",
       "12  Dissociated by members of Rosas lab, frozen in...   \n",
       "13  Dissociated by members of Rosas lab, frozen in...   \n",
       "\n",
       "   organism_ontology_term_id  ... sequenced_fragment sequencing_platform  \\\n",
       "0             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "1             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "2             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "3             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "4             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "5             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "6             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "7             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "8             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "9             NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "10            NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "11            NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "12            NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "13            NCBITaxon:9606  ...        3 prime tag         EFO:0008563   \n",
       "\n",
       "   is_primary_data reference_genome gene_annotation_version  \\\n",
       "0             true          GRCh38                     v91    \n",
       "1             true          GRCh38                     v91    \n",
       "2            false          GRCh38                     v91    \n",
       "3            false          GRCh38                     v91    \n",
       "4            false          GRCh38                     v91    \n",
       "5            false          GRCh38                     v91    \n",
       "6            false          GRCh38                     v91    \n",
       "7            false          GRCh38                     v91    \n",
       "8            false          GRCh38                     v91    \n",
       "9            false          GRCh38                     v91    \n",
       "10           false          GRCh38                     v91    \n",
       "11           false          GRCh38                     v91    \n",
       "12           false          GRCh38                     v91    \n",
       "13           false          GRCh38                     v91    \n",
       "\n",
       "   alignment_software intron_inclusion disease_ontology_term_id  \\\n",
       "0         zUMIs v0.4                no             PATO:0000461   \n",
       "1         zUMIs v0.4                no             PATO:0000461   \n",
       "2         zUMIs v0.4                no             PATO:0000461   \n",
       "3         zUMIs v0.4                no             PATO:0000461   \n",
       "4         zUMIs v0.4                no             PATO:0000461   \n",
       "5         zUMIs v0.4                no             PATO:0000461   \n",
       "6         zUMIs v0.4                no             PATO:0000461   \n",
       "7         zUMIs v0.4                no             PATO:0000461   \n",
       "8         zUMIs v0.4                no             PATO:0000461   \n",
       "9         zUMIs v0.4                no             PATO:0000461   \n",
       "10        zUMIs v0.4                no             PATO:0000461   \n",
       "11        zUMIs v0.4                no             PATO:0000461   \n",
       "12        zUMIs v0.4                no             PATO:0000461   \n",
       "13        zUMIs v0.4                no             PATO:0000461   \n",
       "\n",
       "   self_reported_ethnicity_ontology_term_id development_stage_ontology_term_id  \n",
       "0                                   unknown                     HsapDv:0000242  \n",
       "1                                   unknown                     HsapDv:0000243  \n",
       "2                                   unknown                     HsapDv:0000237  \n",
       "3                                   unknown                     HsapDv:0000238  \n",
       "4                                   unknown                     HsapDv:0000239  \n",
       "5                                   unknown                     HsapDv:0000238  \n",
       "6                                   unknown                     HsapDv:0000237  \n",
       "7                                   unknown                     HsapDv:0000240  \n",
       "8                                   unknown                     HsapDv:0000241  \n",
       "9                                   unknown                     HsapDv:0000241  \n",
       "10                                  unknown                     HsapDv:0000241  \n",
       "11                                  unknown                     HsapDv:0000240  \n",
       "12                                  unknown                     HsapDv:0000241  \n",
       "13                                  unknown                     HsapDv:0000239  \n",
       "\n",
       "[14 rows x 37 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_workflow = ValidationWorkflow(\n",
    "    input = obs,\n",
    "    axis = 'obs'\n",
    ")\n",
    "\n",
    "validated_obs = val_workflow.init_workflow()\n",
    "validated_obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validate obs and uns from adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['nGene', 'nUMI', 'orig.ident', 'percent.mito', 'exp', 'species',\n",
       "       'class', 'library.ident', 'disease.ident', 'subject.ident',\n",
       "       'cell.type.desc.ident', 'cell.type.ident', 'cell.cat.ident',\n",
       "       'name_initial', 'var.ratio.pca', 'hum.stash', 'nCount_RNA',\n",
       "       'nFeature_RNA', 'stash', 'Annotation_for_Publication',\n",
       "       'Cell_Ontology_Class_ID', 'ident'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hum1     1\n",
       "Hum2     1\n",
       "001C     1\n",
       "048C     1\n",
       "098C     1\n",
       "133C     1\n",
       "1372C    1\n",
       "137C     1\n",
       "160C     1\n",
       "192C     1\n",
       "222C     1\n",
       "244C     1\n",
       "388C     1\n",
       "454C     1\n",
       "Name: sample_ID, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs['sample_ID'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hum2     2997\n",
       "098C     2849\n",
       "222C     2825\n",
       "Hum1     2193\n",
       "133C     2074\n",
       "1372C    1291\n",
       "192C      823\n",
       "048C      583\n",
       "160C      574\n",
       "388C      502\n",
       "001C      466\n",
       "454C      363\n",
       "137C      211\n",
       "244C      116\n",
       "Name: subject.ident, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['subject.ident'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check non-intersecting values of obs['library_ID'] and adata.obs['subject.ident']\n",
    "\n",
    "sub_id = adata.obs['subject.ident'].value_counts(dropna=False).index.tolist()\n",
    "lib_id = obs['sample_ID'].value_counts().index.tolist()\n",
    "\n",
    "# show non-intersecting values \n",
    "set(sub_id).symmetric_difference(set(lib_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 17867 × 46204\n",
       "    obs: 'nGene', 'nUMI', 'orig.ident', 'percent.mito', 'exp', 'species', 'class', 'library.ident', 'disease.ident', 'subject.ident', 'cell.type.desc.ident', 'cell.type.ident', 'cell.cat.ident', 'name_initial', 'var.ratio.pca', 'hum.stash', 'nCount_RNA', 'nFeature_RNA', 'stash', 'Annotation_for_Publication', 'Cell_Ontology_Class_ID', 'ident'\n",
       "    uns: 'title', 'study_PI', 'batch_condition', 'default_embedding', 'unpublished', 'comments'\n",
       "    obsm: 'CCA', 'CCA.ALIGNED', 'X_tsne'\n",
       "    layers: 'logcounts', 'scaledata'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merger = AnnDataMerger(\n",
    "    adata = adata,\n",
    "    uns_df = uns\n",
    ")\n",
    "\n",
    "adata = merger.add_uns_metadata()\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nGene</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>percent.mito</th>\n",
       "      <th>exp</th>\n",
       "      <th>species</th>\n",
       "      <th>class</th>\n",
       "      <th>library.ident</th>\n",
       "      <th>disease.ident</th>\n",
       "      <th>subject.ident</th>\n",
       "      <th>...</th>\n",
       "      <th>sequencing_platform</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>reference_genome</th>\n",
       "      <th>gene_annotation_version</th>\n",
       "      <th>alignment_software</th>\n",
       "      <th>intron_inclusion</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "      <th>sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAACCTGAGGATATAC</th>\n",
       "      <td>2607</td>\n",
       "      <td>8348.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.026007</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000243</td>\n",
       "      <td>Hum2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAAGATGGTCTGGAGA</th>\n",
       "      <td>1496</td>\n",
       "      <td>3905.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.050192</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000243</td>\n",
       "      <td>Hum2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAAGTAGGTGTGGTTT</th>\n",
       "      <td>2776</td>\n",
       "      <td>9400.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.044371</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000243</td>\n",
       "      <td>Hum2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAGGAGCAGTCTTGCA</th>\n",
       "      <td>1261</td>\n",
       "      <td>3436.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.036671</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000243</td>\n",
       "      <td>Hum2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDRI_1b__AAGTCTGCAATGAAAC</th>\n",
       "      <td>2507</td>\n",
       "      <td>8426.0</td>\n",
       "      <td>NDRI_1b</td>\n",
       "      <td>0.026110</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Original CCA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Hum2</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>true</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000243</td>\n",
       "      <td>Hum2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222C_GGCTGGTAGCTGATAA</th>\n",
       "      <td>1211</td>\n",
       "      <td>2697.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.041898</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>222C</td>\n",
       "      <td>Control</td>\n",
       "      <td>222C</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "      <td>222C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388C_GCTGGGTCAGACACTT</th>\n",
       "      <td>1217</td>\n",
       "      <td>2522.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.055908</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>388C</td>\n",
       "      <td>Control</td>\n",
       "      <td>388C</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "      <td>388C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388C_GGACGTCAGCAGGTCA</th>\n",
       "      <td>922</td>\n",
       "      <td>1851.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.060508</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>388C</td>\n",
       "      <td>Control</td>\n",
       "      <td>388C</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "      <td>388C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388C_GTGCGGTCACCTATCC</th>\n",
       "      <td>1074</td>\n",
       "      <td>2025.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.070123</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>388C</td>\n",
       "      <td>Control</td>\n",
       "      <td>388C</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "      <td>388C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92C_GAATGAATCTCTTATG</th>\n",
       "      <td>2493</td>\n",
       "      <td>8247.0</td>\n",
       "      <td>CellSoup</td>\n",
       "      <td>0.029950</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Controls</td>\n",
       "      <td>92C</td>\n",
       "      <td>Control</td>\n",
       "      <td>192C</td>\n",
       "      <td>...</td>\n",
       "      <td>EFO:0008563</td>\n",
       "      <td>false</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>v91</td>\n",
       "      <td>zUMIs v0.4</td>\n",
       "      <td>no</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>HsapDv:0000241</td>\n",
       "      <td>192C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17867 rows × 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           nGene    nUMI orig.ident  percent.mito   exp  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC   2607  8348.0    NDRI_1b      0.026007  Hum2   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA   1496  3905.0    NDRI_1b      0.050192  Hum2   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT   2776  9400.0    NDRI_1b      0.044371  Hum2   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA   1261  3436.0    NDRI_1b      0.036671  Hum2   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC   2507  8426.0    NDRI_1b      0.026110  Hum2   \n",
       "...                          ...     ...        ...           ...   ...   \n",
       "222C_GGCTGGTAGCTGATAA       1211  2697.0   CellSoup      0.041898  None   \n",
       "388C_GCTGGGTCAGACACTT       1217  2522.0   CellSoup      0.055908  None   \n",
       "388C_GGACGTCAGCAGGTCA        922  1851.0   CellSoup      0.060508  None   \n",
       "388C_GTGCGGTCACCTATCC       1074  2025.0   CellSoup      0.070123  None   \n",
       "92C_GAATGAATCTCTTATG        2493  8247.0   CellSoup      0.029950  None   \n",
       "\n",
       "                          species           class library.ident disease.ident  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC   Human    Original CCA          None          None   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA   Human    Original CCA          None          None   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT   Human    Original CCA          None          None   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA   Human    Original CCA          None          None   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC   Human    Original CCA          None          None   \n",
       "...                           ...             ...           ...           ...   \n",
       "222C_GGCTGGTAGCTGATAA        None  Human Controls          222C       Control   \n",
       "388C_GCTGGGTCAGACACTT        None  Human Controls          388C       Control   \n",
       "388C_GGACGTCAGCAGGTCA        None  Human Controls          388C       Control   \n",
       "388C_GTGCGGTCACCTATCC        None  Human Controls          388C       Control   \n",
       "92C_GAATGAATCTCTTATG         None  Human Controls           92C       Control   \n",
       "\n",
       "                          subject.ident  ... sequencing_platform  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC          Hum2  ...         EFO:0008563   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA          Hum2  ...         EFO:0008563   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT          Hum2  ...         EFO:0008563   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA          Hum2  ...         EFO:0008563   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC          Hum2  ...         EFO:0008563   \n",
       "...                                 ...  ...                 ...   \n",
       "222C_GGCTGGTAGCTGATAA              222C  ...         EFO:0008563   \n",
       "388C_GCTGGGTCAGACACTT              388C  ...         EFO:0008563   \n",
       "388C_GGACGTCAGCAGGTCA              388C  ...         EFO:0008563   \n",
       "388C_GTGCGGTCACCTATCC              388C  ...         EFO:0008563   \n",
       "92C_GAATGAATCTCTTATG               192C  ...         EFO:0008563   \n",
       "\n",
       "                          is_primary_data reference_genome  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC            true          GRCh38    \n",
       "NDRI_1b__AAAGATGGTCTGGAGA            true          GRCh38    \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT            true          GRCh38    \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA            true          GRCh38    \n",
       "NDRI_1b__AAGTCTGCAATGAAAC            true          GRCh38    \n",
       "...                                   ...              ...   \n",
       "222C_GGCTGGTAGCTGATAA               false          GRCh38    \n",
       "388C_GCTGGGTCAGACACTT               false          GRCh38    \n",
       "388C_GGACGTCAGCAGGTCA               false          GRCh38    \n",
       "388C_GTGCGGTCACCTATCC               false          GRCh38    \n",
       "92C_GAATGAATCTCTTATG                false          GRCh38    \n",
       "\n",
       "                          gene_annotation_version  alignment_software  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC                    v91          zUMIs v0.4    \n",
       "NDRI_1b__AAAGATGGTCTGGAGA                    v91          zUMIs v0.4    \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT                    v91          zUMIs v0.4    \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA                    v91          zUMIs v0.4    \n",
       "NDRI_1b__AAGTCTGCAATGAAAC                    v91          zUMIs v0.4    \n",
       "...                                           ...                 ...   \n",
       "222C_GGCTGGTAGCTGATAA                        v91          zUMIs v0.4    \n",
       "388C_GCTGGGTCAGACACTT                        v91          zUMIs v0.4    \n",
       "388C_GGACGTCAGCAGGTCA                        v91          zUMIs v0.4    \n",
       "388C_GTGCGGTCACCTATCC                        v91          zUMIs v0.4    \n",
       "92C_GAATGAATCTCTTATG                         v91          zUMIs v0.4    \n",
       "\n",
       "                          intron_inclusion  disease_ontology_term_id  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC               no              PATO:0000461   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA               no              PATO:0000461   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT               no              PATO:0000461   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA               no              PATO:0000461   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC               no              PATO:0000461   \n",
       "...                                    ...                       ...   \n",
       "222C_GGCTGGTAGCTGATAA                   no              PATO:0000461   \n",
       "388C_GCTGGGTCAGACACTT                   no              PATO:0000461   \n",
       "388C_GGACGTCAGCAGGTCA                   no              PATO:0000461   \n",
       "388C_GTGCGGTCACCTATCC                   no              PATO:0000461   \n",
       "92C_GAATGAATCTCTTATG                    no              PATO:0000461   \n",
       "\n",
       "                           self_reported_ethnicity_ontology_term_id  \\\n",
       "NDRI_1b__AAACCTGAGGATATAC                                   unknown   \n",
       "NDRI_1b__AAAGATGGTCTGGAGA                                   unknown   \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT                                   unknown   \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA                                   unknown   \n",
       "NDRI_1b__AAGTCTGCAATGAAAC                                   unknown   \n",
       "...                                                             ...   \n",
       "222C_GGCTGGTAGCTGATAA                                       unknown   \n",
       "388C_GCTGGGTCAGACACTT                                       unknown   \n",
       "388C_GGACGTCAGCAGGTCA                                       unknown   \n",
       "388C_GTGCGGTCACCTATCC                                       unknown   \n",
       "92C_GAATGAATCTCTTATG                                        unknown   \n",
       "\n",
       "                          development_stage_ontology_term_id sample_ID  \n",
       "NDRI_1b__AAACCTGAGGATATAC                     HsapDv:0000243      Hum2  \n",
       "NDRI_1b__AAAGATGGTCTGGAGA                     HsapDv:0000243      Hum2  \n",
       "NDRI_1b__AAAGTAGGTGTGGTTT                     HsapDv:0000243      Hum2  \n",
       "NDRI_1b__AAGGAGCAGTCTTGCA                     HsapDv:0000243      Hum2  \n",
       "NDRI_1b__AAGTCTGCAATGAAAC                     HsapDv:0000243      Hum2  \n",
       "...                                                      ...       ...  \n",
       "222C_GGCTGGTAGCTGATAA                         HsapDv:0000241      222C  \n",
       "388C_GCTGGGTCAGACACTT                         HsapDv:0000241      388C  \n",
       "388C_GGACGTCAGCAGGTCA                         HsapDv:0000241      388C  \n",
       "388C_GTGCGGTCACCTATCC                         HsapDv:0000241      388C  \n",
       "92C_GAATGAATCTCTTATG                          HsapDv:0000241      192C  \n",
       "\n",
       "[17867 rows x 59 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merger = AnnDataMerger(\n",
    "    adata = adata,\n",
    "    obs_df = obs\n",
    ")\n",
    "\n",
    "adata = merger.add_obs_metadata(\n",
    "    adata_col = 'subject.ident',\n",
    "    df_col = 'sample_ID',\n",
    "    skip = None\n",
    ")\n",
    "\n",
    "adata.obs['sample_ID'] = adata.obs['subject.ident']\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add author cell type markers to UNS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>author_cell_type_markers</th>\n",
       "      <th>author_cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGER</td>\n",
       "      <td>ATI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SFTPC</td>\n",
       "      <td>ATII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CD79A</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JCHAIN</td>\n",
       "      <td>B_Jchain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KRT5</td>\n",
       "      <td>Basal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CCDC153</td>\n",
       "      <td>Ciliated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CD86</td>\n",
       "      <td>DC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CA4</td>\n",
       "      <td>EC_cap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PROX1</td>\n",
       "      <td>EC_lymph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>VWF</td>\n",
       "      <td>EC_vasc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ITGA8</td>\n",
       "      <td>Fib_Col13a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FBLN2</td>\n",
       "      <td>Fib_Col14a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>INHBA</td>\n",
       "      <td>Mac_alv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CD14</td>\n",
       "      <td>Mac_inter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CPA3</td>\n",
       "      <td>Mast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FCN1</td>\n",
       "      <td>Mo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NKG7</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SCGB1A1</td>\n",
       "      <td>Secretory</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ACTA2</td>\n",
       "      <td>SMCs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CD8A</td>\n",
       "      <td>T_CD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CD3D</td>\n",
       "      <td>T_cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   author_cell_type_markers author_cell_type\n",
       "0                      AGER              ATI\n",
       "1                     SFTPC             ATII\n",
       "2                     CD79A          B_cells\n",
       "3                    JCHAIN         B_Jchain\n",
       "4                      KRT5            Basal\n",
       "5                   CCDC153         Ciliated\n",
       "6                      CD86               DC\n",
       "7                       CA4           EC_cap\n",
       "8                     PROX1         EC_lymph\n",
       "9                       VWF          EC_vasc\n",
       "10                    ITGA8      Fib_Col13a1\n",
       "11                    FBLN2      Fib_Col14a1\n",
       "12                    INHBA          Mac_alv\n",
       "13                     CD14        Mac_inter\n",
       "14                     CPA3             Mast\n",
       "15                     FCN1               Mo\n",
       "16                     NKG7               NK\n",
       "17                  SCGB1A1        Secretory\n",
       "18                    ACTA2             SMCs\n",
       "19                     CD8A            T_CD8\n",
       "20                     CD3D          T_cells"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marker_list = pd.read_csv(\"/ictstr01/home/icb/raphael.kfuri-rubens/data/hlca_v2/Niklason_Kaminski_publ/curated.marker.list.2024-05-21.txt\", sep=\";\").T\n",
    "marker_list[AUTHOR_CELL_TYPE] = marker_list.index\n",
    "marker_list.rename(columns={0:MARKER_GENES}, inplace=True)\n",
    "marker_list.reset_index(drop=True, inplace=True)\n",
    "\n",
    "marker_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check author cell type annotations and Cell Ontology IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['nGene', 'nUMI', 'orig.ident', 'percent.mito', 'exp', 'species',\n",
       "       'class', 'library.ident', 'disease.ident', 'subject.ident',\n",
       "       'cell.type.desc.ident', 'cell.type.ident', 'cell.cat.ident',\n",
       "       'name_initial', 'var.ratio.pca', 'hum.stash', 'nCount_RNA',\n",
       "       'nFeature_RNA', 'stash', 'Annotation_for_Publication',\n",
       "       'Cell_Ontology_Class_ID', 'ident', 'donor_id', 'protocol_URL',\n",
       "       'institute', 'sample_collection_site',\n",
       "       'sample_collection_relative_time_point', 'library_ID',\n",
       "       'library_ID_repository', 'author_batch_notes',\n",
       "       'organism_ontology_term_id', 'manner_of_death', 'sample_source',\n",
       "       'sex_ontology_term_id', 'sample_collection_method', 'tissue_type',\n",
       "       'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text',\n",
       "       'sample_preservation_method', 'suspension_type', 'cell_enrichment',\n",
       "       'cell_viability_percentage', 'cell_number_loaded',\n",
       "       'sample_collection_year', 'assay_ontology_term_id',\n",
       "       'library_preparation_batch', 'library_sequencing_run',\n",
       "       'sequenced_fragment', 'sequencing_platform', 'is_primary_data',\n",
       "       'reference_genome', 'gene_annotation_version', 'alignment_software',\n",
       "       'intron_inclusion', 'disease_ontology_term_id',\n",
       "       'self_reported_ethnicity_ontology_term_id',\n",
       "       'development_stage_ontology_term_id', 'sample_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "T_cells         2018\n",
       "ATII            1797\n",
       "Mo              1718\n",
       "Mac_alv         1710\n",
       "DC              1343\n",
       "T_CD8+          1292\n",
       "NK              1189\n",
       "Mac_inter       1116\n",
       "Ciliated         966\n",
       "Fib_Col14a1+     724\n",
       "EC_lymph         687\n",
       "ATI              672\n",
       "EC_vasc          632\n",
       "B_cells          617\n",
       "EC_cap           309\n",
       "Mast             272\n",
       "Secretory        215\n",
       "Fib_Col13a1+     196\n",
       "B_Jchain+        171\n",
       "Basal            122\n",
       "SMCs             101\n",
       "Name: author_cell_type, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[AUTHOR_CELL_TYPE] = adata.obs['Annotation_for_Publication']\n",
    "adata.obs[AUTHOR_CELL_TYPE].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CL:0000084    2018\n",
       "CL:0002063    1797\n",
       "CL:0000576    1718\n",
       "CL:0000583    1710\n",
       "CL:0000451    1343\n",
       "CL:0000795    1292\n",
       "CL:0000623    1189\n",
       "CL:4033043    1116\n",
       "CL:0000064     966\n",
       "CL:4028006     724\n",
       "CL:0002138     687\n",
       "CL:0002062     672\n",
       "CL:4028002     632\n",
       "CL:0000236     617\n",
       "CL:4028003     309\n",
       "CL:0000097     272\n",
       "CL:0000151     215\n",
       "CL:4028004     196\n",
       "CL:0000786     171\n",
       "CL:0002329     122\n",
       "CL:0000192     101\n",
       "Name: cell_type_ontology_term_id, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[CELL_TYPE_ONTOLOGY_ID] = adata.obs['Cell_Ontology_Class_ID']\n",
    "adata.obs[CELL_TYPE_ONTOLOGY_ID].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>author_cell_type_markers</th>\n",
       "      <th>author_cell_type</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGER</td>\n",
       "      <td>ATI</td>\n",
       "      <td>CL:0002062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SFTPC</td>\n",
       "      <td>ATII</td>\n",
       "      <td>CL:0002063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CD79A</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>CL:0000236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JCHAIN</td>\n",
       "      <td>B_Jchain</td>\n",
       "      <td>CL:0000236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KRT5</td>\n",
       "      <td>Basal</td>\n",
       "      <td>CL:0002329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CCDC153</td>\n",
       "      <td>Ciliated</td>\n",
       "      <td>CL:0000064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CD86</td>\n",
       "      <td>DC</td>\n",
       "      <td>CL:0000451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CA4</td>\n",
       "      <td>EC_cap</td>\n",
       "      <td>CL:4028003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PROX1</td>\n",
       "      <td>EC_lymph</td>\n",
       "      <td>CL:0002138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>VWF</td>\n",
       "      <td>EC_vasc</td>\n",
       "      <td>CL:4028002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ITGA8</td>\n",
       "      <td>Fib_Col13a1</td>\n",
       "      <td>CL:0000057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FBLN2</td>\n",
       "      <td>Fib_Col14a1</td>\n",
       "      <td>CL:0000057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>INHBA</td>\n",
       "      <td>Mac_alv</td>\n",
       "      <td>CL:0000583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CD14</td>\n",
       "      <td>Mac_inter</td>\n",
       "      <td>CL:4033043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CPA3</td>\n",
       "      <td>Mast</td>\n",
       "      <td>CL:0000097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FCN1</td>\n",
       "      <td>Mo</td>\n",
       "      <td>CL:0000576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NKG7</td>\n",
       "      <td>NK</td>\n",
       "      <td>CL:0000623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SCGB1A1</td>\n",
       "      <td>Secretory</td>\n",
       "      <td>CL:0000151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ACTA2</td>\n",
       "      <td>SMCs</td>\n",
       "      <td>CL:0000192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CD8A</td>\n",
       "      <td>T_CD8</td>\n",
       "      <td>CL:0000625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CD3D</td>\n",
       "      <td>T_cells</td>\n",
       "      <td>CL:0000084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   author_cell_type_markers author_cell_type cell_type_ontology_term_id\n",
       "0                      AGER              ATI                 CL:0002062\n",
       "1                     SFTPC             ATII                 CL:0002063\n",
       "2                     CD79A          B_cells                 CL:0000236\n",
       "3                    JCHAIN         B_Jchain                 CL:0000236\n",
       "4                      KRT5            Basal                 CL:0002329\n",
       "5                   CCDC153         Ciliated                 CL:0000064\n",
       "6                      CD86               DC                 CL:0000451\n",
       "7                       CA4           EC_cap                 CL:4028003\n",
       "8                     PROX1         EC_lymph                 CL:0002138\n",
       "9                       VWF          EC_vasc                 CL:4028002\n",
       "10                    ITGA8      Fib_Col13a1                 CL:0000057\n",
       "11                    FBLN2      Fib_Col14a1                 CL:0000057\n",
       "12                    INHBA          Mac_alv                 CL:0000583\n",
       "13                     CD14        Mac_inter                 CL:4033043\n",
       "14                     CPA3             Mast                 CL:0000097\n",
       "15                     FCN1               Mo                 CL:0000576\n",
       "16                     NKG7               NK                 CL:0000623\n",
       "17                  SCGB1A1        Secretory                 CL:0000151\n",
       "18                    ACTA2             SMCs                 CL:0000192\n",
       "19                     CD8A            T_CD8                 CL:0000625\n",
       "20                     CD3D          T_cells                 CL:0000084"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_type_ontology_mapping = adata.obs.set_index(AUTHOR_CELL_TYPE)[CELL_TYPE_ONTOLOGY_ID].to_dict()\n",
    "\n",
    "marker_list[CELL_TYPE_ONTOLOGY_ID] = marker_list[AUTHOR_CELL_TYPE].map(cell_type_ontology_mapping)\n",
    "\n",
    "missing_ontologies = {\n",
    "    'B_Jchain': 'CL:0000236',\n",
    "    'Fib_Col13a1': 'CL:0000057',\n",
    "    'Fib_Col14a1': 'CL:0000057',\n",
    "    'T_CD8': 'CL:0000625'\n",
    "}\n",
    "\n",
    "marker_list[CELL_TYPE_ONTOLOGY_ID] = marker_list[AUTHOR_CELL_TYPE].map(missing_ontologies).fillna(marker_list[CELL_TYPE_ONTOLOGY_ID])\n",
    "\n",
    "marker_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 17867 × 46204\n",
       "    obs: 'nGene', 'nUMI', 'orig.ident', 'percent.mito', 'exp', 'species', 'class', 'library.ident', 'disease.ident', 'subject.ident', 'cell.type.desc.ident', 'cell.type.ident', 'cell.cat.ident', 'name_initial', 'var.ratio.pca', 'hum.stash', 'nCount_RNA', 'nFeature_RNA', 'stash', 'Annotation_for_Publication', 'Cell_Ontology_Class_ID', 'ident', 'donor_id', 'protocol_URL', 'institute', 'sample_collection_site', 'sample_collection_relative_time_point', 'library_ID', 'library_ID_repository', 'author_batch_notes', 'organism_ontology_term_id', 'manner_of_death', 'sample_source', 'sex_ontology_term_id', 'sample_collection_method', 'tissue_type', 'sampled_site_condition', 'tissue_ontology_term_id', 'tissue_free_text', 'sample_preservation_method', 'suspension_type', 'cell_enrichment', 'cell_viability_percentage', 'cell_number_loaded', 'sample_collection_year', 'assay_ontology_term_id', 'library_preparation_batch', 'library_sequencing_run', 'sequenced_fragment', 'sequencing_platform', 'is_primary_data', 'reference_genome', 'gene_annotation_version', 'alignment_software', 'intron_inclusion', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sample_ID', 'author_cell_type', 'cell_type_ontology_term_id'\n",
       "    uns: 'title', 'study_PI', 'batch_condition', 'default_embedding', 'unpublished', 'comments', 'author_cell_type_markers'\n",
       "    obsm: 'CCA', 'CCA.ALIGNED', 'X_tsne'\n",
       "    layers: 'logcounts', 'scaledata'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.uns[MARKER_GENES] = marker_list\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check whether ENSEMBL IDs in var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FGR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFH</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227029</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RN7SL782P</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274532</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277705</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000280139</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46204 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [DPM1, SCYL3, C1orf112, FGR, CFH, FUCA2, GCLC, NFYA, STPG1, NIPAL3, LAS1L, ENPP4, SEMA3F, CFTR, ANKIB1, CYP51A1, KRIT1, RAD52, MYH16, BAD, LAP3, CD99, HECW1, MAD1L1, LASP1, SNX11, TMEM176A, M6PR, KLHL13, ICA1, ALS2, CASP10, CFLAR, TFPI, NDUFAF7, RBM5, MTMR7, SLC7A2, ARF5, POLDIP2, PLXND1, AK2, CD38, FKBP4, KDM1A, RBM6, CAMKK1, RECQL, VPS50, NDUFAB1, PDK4, ABCB5, SLC25A13, ST7, CDC27, SLC4A1, CALCR, HCCS, DVL2, UPF1, SKAP2, SLC25A5, MCUB, POLR2J, DHX33, MEOX1, THSD7A, LIG3, RPAP3, ACSM3, CIAPIN1, SPPL2B, FAM214B, PRKAR2B, MSL3, CREBBP, TSPOAP1, GCFC2, WDR54, CROT, KMT2E, RHBDD2, IBTK, ZNF195, MYCBP2, FBXL3, ITGAL, ITGA3, ZFX, LAMP2, GDE1, REX1BD, CRLF1, OSBPL7, MAP3K14, ABCC8, AP2B1, ZNF263, SPATA20, TNFRSF12A, ...]\n",
       "\n",
       "[46204 rows x 0 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "4689 input query terms found dup hits:\t[('BORCS8-MEF2B', 2), ('PMS2P4', 2), ('PMS2P1', 2), ('CRYBB2P1', 2), ('LRP5L', 2), ('TPTEP1', 2), ('\n",
      "1133 input query terms found no hit:\t['C1orf112', 'FAM214B', 'KIAA0100', 'TMEM159', 'CXorf56', 'ARNTL2', 'SARS', 'ADSS', 'FAM214A', 'KIAA\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>C1orf112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FGR</th>\n",
       "      <td>ENSG00000000938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFH</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227029</th>\n",
       "      <td>ENSG00000227029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RN7SL782P</th>\n",
       "      <td>ENSG00000239708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274532</th>\n",
       "      <td>ENSG00000274532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277705</th>\n",
       "      <td>ENSG00000277705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000280139</th>\n",
       "      <td>ENSG00000280139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46204 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ensembl_id\n",
       "DPM1             ENSG00000000419\n",
       "SCYL3            ENSG00000000457\n",
       "C1orf112                C1orf112\n",
       "FGR              ENSG00000000938\n",
       "CFH              ENSG00000000971\n",
       "...                          ...\n",
       "ENSG00000227029  ENSG00000227029\n",
       "RN7SL782P        ENSG00000239708\n",
       "ENSG00000274532  ENSG00000274532\n",
       "ENSG00000277705  ENSG00000277705\n",
       "ENSG00000280139  ENSG00000280139\n",
       "\n",
       "[46204 rows x 1 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import mygene\n",
    "\n",
    "def map_gene_names_to_ensembl(mixed_list):\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    gene_names = [item for item in mixed_list if not item.startswith(\"ENS\")]\n",
    "    gene_info = mg.querymany(gene_names, scopes='symbol', fields='ensembl.gene', species='human')\n",
    "\n",
    "    gene_to_ensembl = {}\n",
    "    for gene in gene_info:\n",
    "        if 'notfound' not in gene and 'ensembl' in gene:\n",
    "            ensembl_id = gene['ensembl']['gene'] if isinstance(gene['ensembl'], dict) else gene['ensembl'][0]['gene']\n",
    "            gene_to_ensembl[gene['query']] = ensembl_id\n",
    "\n",
    "    mapped_list = [gene_to_ensembl.get(item, item) for item in mixed_list]\n",
    "    return mapped_list\n",
    "\n",
    "\n",
    "adata.var['ensembl_id'] = map_gene_names_to_ensembl(adata.var.index.tolist())\n",
    "\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>C1orf112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM214B</th>\n",
       "      <td>FAM214B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIAA0100</th>\n",
       "      <td>KIAA0100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM159</th>\n",
       "      <td>TMEM159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CXorf56</th>\n",
       "      <td>CXorf56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR4419B</th>\n",
       "      <td>MIR4419B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM197Y10</th>\n",
       "      <td>FAM197Y10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNMTL1P1</th>\n",
       "      <td>RNMTL1P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNU6-785P</th>\n",
       "      <td>RNU6-785P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNMTL1P2</th>\n",
       "      <td>RNMTL1P2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1254 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ensembl_id\n",
       "C1orf112    C1orf112\n",
       "FAM214B      FAM214B\n",
       "KIAA0100    KIAA0100\n",
       "TMEM159      TMEM159\n",
       "CXorf56      CXorf56\n",
       "...              ...\n",
       "MIR4419B    MIR4419B\n",
       "FAM197Y10  FAM197Y10\n",
       "RNMTL1P1    RNMTL1P1\n",
       "RNU6-785P  RNU6-785P\n",
       "RNMTL1P2    RNMTL1P2\n",
       "\n",
       "[1254 rows x 1 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1218 gene names without ensembl IDs\n",
    "adata.var[~adata.var['ensembl_id'].str.startswith('ENSG')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FGR</th>\n",
       "      <td>ENSG00000000938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFH</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227029</th>\n",
       "      <td>ENSG00000227029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RN7SL782P</th>\n",
       "      <td>ENSG00000239708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274532</th>\n",
       "      <td>ENSG00000274532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277705</th>\n",
       "      <td>ENSG00000277705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000280139</th>\n",
       "      <td>ENSG00000280139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46204 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ensembl_id\n",
       "DPM1             ENSG00000000419\n",
       "SCYL3            ENSG00000000457\n",
       "C1orf112                     NaN\n",
       "FGR              ENSG00000000938\n",
       "CFH              ENSG00000000971\n",
       "...                          ...\n",
       "ENSG00000227029  ENSG00000227029\n",
       "RN7SL782P        ENSG00000239708\n",
       "ENSG00000274532  ENSG00000274532\n",
       "ENSG00000277705  ENSG00000277705\n",
       "ENSG00000280139  ENSG00000280139\n",
       "\n",
       "[46204 rows x 1 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var['ensembl_id'] = adata.var['ensembl_id'].apply(lambda x: x if str(x).startswith('ENSG') else np.nan)\n",
    "\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>NaN</td>\n",
       "      <td>C1orf112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>ENSG00000000938</td>\n",
       "      <td>FGR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>CFH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227029</th>\n",
       "      <td>ENSG00000227029</td>\n",
       "      <td>ENSG00000227029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239708</th>\n",
       "      <td>ENSG00000239708</td>\n",
       "      <td>RN7SL782P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274532</th>\n",
       "      <td>ENSG00000274532</td>\n",
       "      <td>ENSG00000274532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277705</th>\n",
       "      <td>ENSG00000277705</td>\n",
       "      <td>ENSG00000277705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000280139</th>\n",
       "      <td>ENSG00000280139</td>\n",
       "      <td>ENSG00000280139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46204 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ensembl_id      gene_symbol\n",
       "index                                            \n",
       "ENSG00000000419  ENSG00000000419             DPM1\n",
       "ENSG00000000457  ENSG00000000457            SCYL3\n",
       "NaN                          NaN         C1orf112\n",
       "ENSG00000000938  ENSG00000000938              FGR\n",
       "ENSG00000000971  ENSG00000000971              CFH\n",
       "...                          ...              ...\n",
       "ENSG00000227029  ENSG00000227029  ENSG00000227029\n",
       "ENSG00000239708  ENSG00000239708        RN7SL782P\n",
       "ENSG00000274532  ENSG00000274532  ENSG00000274532\n",
       "ENSG00000277705  ENSG00000277705  ENSG00000277705\n",
       "ENSG00000280139  ENSG00000280139  ENSG00000280139\n",
       "\n",
       "[46204 rows x 2 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var['gene_symbol'] = adata.var.index\n",
    "adata.var.index = adata.var['ensembl_id']\n",
    "adata.var.index.name = 'index'\n",
    "\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nan</th>\n",
       "      <td>nan</td>\n",
       "      <td>C1orf112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>ENSG00000000938</td>\n",
       "      <td>FGR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>CFH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227029</th>\n",
       "      <td>ENSG00000227029</td>\n",
       "      <td>ENSG00000227029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239708</th>\n",
       "      <td>ENSG00000239708</td>\n",
       "      <td>RN7SL782P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274532</th>\n",
       "      <td>ENSG00000274532</td>\n",
       "      <td>ENSG00000274532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277705</th>\n",
       "      <td>ENSG00000277705</td>\n",
       "      <td>ENSG00000277705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000280139</th>\n",
       "      <td>ENSG00000280139</td>\n",
       "      <td>ENSG00000280139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46204 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ensembl_id      gene_symbol\n",
       "index                                            \n",
       "ENSG00000000419  ENSG00000000419             DPM1\n",
       "ENSG00000000457  ENSG00000000457            SCYL3\n",
       "nan                          nan         C1orf112\n",
       "ENSG00000000938  ENSG00000000938              FGR\n",
       "ENSG00000000971  ENSG00000000971              CFH\n",
       "...                          ...              ...\n",
       "ENSG00000227029  ENSG00000227029  ENSG00000227029\n",
       "ENSG00000239708  ENSG00000239708        RN7SL782P\n",
       "ENSG00000274532  ENSG00000274532  ENSG00000274532\n",
       "ENSG00000277705  ENSG00000277705  ENSG00000277705\n",
       "ENSG00000280139  ENSG00000280139  ENSG00000280139\n",
       "\n",
       "[46204 rows x 2 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var['gene_symbol'] = adata.var['gene_symbol'].astype(str)\n",
    "adata.var['ensembl_id'] =adata.var['ensembl_id'].astype(str)\n",
    "\n",
    "adata.var.index = adata.var.index.astype(str)\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ensembl_id     object\n",
       "gene_symbol    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.index = adata.var.index.astype('str')\n",
    "adata.var.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22619.0"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.toarray().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make checks for adata before saving raw\n",
    "adata.obs = adata.obs.astype(str)\n",
    "\n",
    "# Make specific columns numeric\n",
    "adata.obs['cell_viability_percentage'] = pd.to_numeric(adata.obs['cell_viability_percentage'], errors='coerce')\n",
    "adata.obs['cell_number_loaded'] = pd.to_numeric(adata.obs['cell_number_loaded'], errors='coerce')\n",
    "adata.obs['sample_collection_year'] = pd.to_numeric(adata.obs['sample_collection_year'], errors='coerce')\n",
    "\n",
    "\n",
    "numeric_cols = ['nGene', 'nUMI', 'percent.mito']\n",
    "for col in numeric_cols:\n",
    "    adata.obs[col] = pd.to_numeric(adata.obs[col], errors='coerce')\n",
    "\n",
    "# Make df indices str\n",
    "adata.obs.index = adata.obs.index.astype('str')\n",
    "adata.var.index = adata.var.index.astype('str')\n",
    "adata.var_names = adata.var_names.astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.X.astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw = adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<17867x46204 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 39438833 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<17867x46204 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 39438833 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22619"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.toarray().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22619"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.X.toarray().max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validation result\n",
    "\n",
    "### UNS Validation\n",
    "- OK: Tier 1 UNS Google Sheet\n",
    "- OK: Tier 1 UNS AnnData Object  same as Google Sheet\n",
    "\n",
    "### OBS Validation\n",
    "- OK: Tier 1 OBS Google Sheet\n",
    "- OK: Tier 1 OBS AnnData Object same as Google Sheet\n",
    "\n",
    "\n",
    "## Not ideal:\n",
    "\n",
    "sample_source, sample_collection_method, sequenced_fragment, and intron_inclusion\n",
    "\n",
    "contain NAs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Submission Status\n",
    "\n",
    "- CHECK: Raw counts in X and in raw\n",
    "- CHECK: Tier 1 Metadata in OBS\n",
    "- CHECK: Cell Ontology IDs in OBS\n",
    "- CHECK: Author cell type in OBS\n",
    "- CHECK: Marker genes in UNS\n",
    "- CHECK: ENSEMBL IDs and gene symbols in var\n",
    "\n",
    "# Revisions:\n",
    "\n",
    "- Find out NAs about sample_source, sample_collection_method, sequenced_fragment, and intron_inclusion\n",
    "\n",
    "- ENSEMBL IDs have been queryied -> Maybe ask at some point again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(join(OUTPUT_PATH_PREREVISION, f\"{DATASET_ID}.h5ad\"))\n",
    "adata.write_h5ad(join(OUTPUT_PATH_POSTREVISION, f\"{DATASET_ID}.h5ad\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hlcav2-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
